// @svar-ui/react-grid/dist/index.es.js
import { jsxs as q, jsx as c, Fragment as xe } from "react/jsx-runtime";
import Ut, { createContext as jt, useContext as Z, useMemo as S, useRef as A, useEffect as z, useCallback as W, useState as K, forwardRef as Bt, useImperativeHandle as Gt } from "react";
import { Text as qt, RichSelect as Qt, DatePicker as Zt, SuggestDropdown as nt, Dropdown as Jt, Calendar as en, context as ot, Locale as tn, Material as pt, Willow as gt, WillowDark as wt } from "@svar-ui/react-core";
import { en as rt } from "@svar-ui/grid-locales";
import { EventBusRouter as nn } from "@svar-ui/lib-state";
import { getRenderValue as Pe, getValue as on, isCommunity as rn, getPrintFilterValue as ln, getPrintCellStyle as St, getHeaderFooterPrintColumns as xt, getPrintColumns as sn, hotkeys as cn, defaultHotkeys as an, scrollTo as dn, DataStore as un, assignChecks as Rt, defaultMenuOptions as fn, handleAction as Nt, defaultToolbarButtons as hn } from "@svar-ui/grid-store";
import { defaultMenuOptions as vo, defaultToolbarButtons as bo, getEditorConfig as Co } from "@svar-ui/grid-store";
import { useWritableProp as pn, useStore as Y, styleObject as Dt, useStoreWithCounter as gn, delegateClick as wn, writable as xn, useStoreLater as we } from "@svar-ui/lib-react";
import { locate as et, getID as We, setID as Ee, clickOutside as Le, locateID as mt, locale as Tt, setEnv as mn, env as yn } from "@svar-ui/lib-dom";
import { registerMenuItem as vn, ContextMenu as kt } from "@svar-ui/react-menu";
import { Toolbar as bn } from "@svar-ui/react-toolbar";
const ne = jt(null);
function Cn(t, e) {
  const n = new ResizeObserver((r) => {
    requestAnimationFrame(() => e(r[0].contentRect));
  });
  return n.observe(t.parentNode), {
    destroy() {
      n.disconnect();
    }
  };
}
const yt = 5, Sn = 700;
function Ke(t) {
  const e = t.getBoundingClientRect(), n = document.body, r = e.top + n.scrollTop - n.clientTop || 0, p = e.left + n.scrollLeft - n.clientLeft || 0;
  return {
    y: Math.round(r),
    x: Math.round(p),
    width: t.offsetWidth,
    height: t.offsetHeight
  };
}
function tt(t, e) {
  const n = Ke(e);
  return { x: t.clientX - n.x, y: t.clientY - n.y };
}
function Rn(t, e) {
  const n = e.current;
  let r = null, p, x, s = !1, v = !1;
  const f = document.createElement("DIV");
  f.className = "wx-drag-zone", f.setAttribute("tabindex", -1);
  function u() {
    clearTimeout(p), p = null;
  }
  function g(l) {
    const H = et(l);
    H && (r = {
      container: f,
      sourceNode: l.target,
      from: We(H),
      pos: tt(l, t)
    }, x = r.pos, a(l));
  }
  function a(l) {
    if (!r) return;
    const H = r.pos = tt(l, t);
    if (!s) {
      if (!v && !l?.target?.getAttribute("draggable-data") && Math.abs(x.x - H.x) < yt && Math.abs(x.y - H.y) < yt)
        return;
      if (T(l) === !1) return N();
    }
    if (v) {
      const P = window.scrollX || document.documentElement.scrollLeft || document.body.scrollLeft, G = window.scrollY || document.documentElement.scrollTop || document.body.scrollTop;
      r.targetNode = document.elementFromPoint(
        l.pageX - P,
        l.pageY - G
      );
    } else r.targetNode = l.target;
    n.move && n.move(l, r), f.style.left = -(r.offset ? r.offset.x : 0) + "px", f.style.top = r.pos.y + (r.offset ? r.offset.y : 0) + "px";
  }
  function b(l) {
    f.parentNode && f.parentNode.removeChild(f), f.innerHTML = "", s && n.end && n.end(l, r), r = x = null, N();
  }
  function D(l) {
    n.getReorder && !n.getReorder() || l.button === 0 && (R(l), window.addEventListener("mousemove", C), window.addEventListener("mouseup", d), g(l));
  }
  function C(l) {
    a(l);
  }
  function d(l) {
    b(l);
  }
  function m(l) {
    if (n.getReorder && !n.getReorder()) return;
    p = setTimeout(() => {
      v = !0, g(l.touches[0]);
    }, Sn), R(l);
    function H() {
      p && u(), l.target.removeEventListener("touchmove", h), l.target.removeEventListener("touchend", H), b(l);
    }
    l.target.addEventListener("touchmove", h), l.target.addEventListener("touchend", H), t.addEventListener("contextmenu", y);
  }
  function h(l) {
    s ? (l.preventDefault(), a(l.touches[0])) : p && u();
  }
  function y(l) {
    if (s || p)
      return l.preventDefault(), !1;
  }
  function i(l) {
    l.preventDefault();
  }
  function R(l) {
    if (!n.getDraggableInfo) return;
    const { hasDraggable: H } = n.getDraggableInfo();
    (!H || l.target.getAttribute("draggable-data")) && (document.body.style.userSelect = "none", document.body.style.webkitUserSelect = "none");
  }
  function T(l) {
    if (s = !0, n.start) {
      if (n.start(l, r) === !1) return !1;
      t.appendChild(f), document.body.style.cursor = "move";
    }
  }
  function N(l) {
    s = v = !1, document.body.style.cursor = "", document.body.style.userSelect = "", document.body.style.webkitUserSelect = "", window.removeEventListener("mousemove", C), window.removeEventListener("mouseup", d), l && (t.removeEventListener("mousedown", D), t.removeEventListener("touchstart", m), t.removeEventListener("dragstart", i));
  }
  return t.addEventListener("mousedown", D), t.addEventListener("touchstart", m), t.addEventListener("dragstart", i), {
    destroy() {
      N(!0);
    }
  };
}
const Nn = 4e-3;
function Dn() {
  return {
    dirX: 0,
    dirY: 0,
    scrollSpeedFactor: 1
  };
}
function Tn(t, e, n, r) {
  const { node: p, left: x, top: s, bottom: v, sense: f, xScroll: u, yScroll: g } = r, a = tt(t, p);
  n.scrollState || (n.scrollState = Dn());
  let b = 0, D = 0;
  a.x < x + f ? b = -1 : a.x > e.width - f && (b = 1), a.y < s + Math.round(f / 2) ? D = -1 : a.y > e.height - v - Math.round(f / 2) && (D = 1), (n.scrollState.dirX !== b || n.scrollState.dirY !== D) && (Et(n), n.scrollState.dirX = b, n.scrollState.dirY = D), (u && n.scrollState.dirX !== 0 || g && n.scrollState.dirY !== 0) && kn(n, r, {
    x: n.scrollState.dirX,
    y: n.scrollState.dirY
  });
}
function kn(t, e, n) {
  t.autoScrollTimer || (t.autoScrollTimer = setTimeout(() => {
    t.activeAutoScroll = setInterval(
      En,
      15,
      t,
      e,
      n
    );
  }, 250));
}
function Et(t) {
  t.scrollSpeedFactor = 1, t.autoScrollTimer && (t.autoScrollTimer = clearTimeout(t.autoScrollTimer), t.activeAutoScroll = clearInterval(t.activeAutoScroll));
}
function En(t, e, n) {
  const { x: r, y: p } = n;
  t.scrollSpeedFactor += Nn, r !== 0 && Hn(t, e, r), p !== 0 && Ln(t, e, p);
}
function Ln(t, e, n) {
  const r = e.node.scrollTop;
  Lt(
    r + Math.round(e.sense / 3) * t.scrollSpeedFactor * n,
    "scrollTop",
    e
  );
}
function Hn(t, e, n) {
  const r = e.node.scrollLeft;
  Lt(
    r + Math.round(e.sense / 3) * t.scrollSpeedFactor * n,
    "scrollLeft",
    e
  );
}
function Lt(t, e, n) {
  n.node[e] = t;
}
function Ye(t, e, n, r, p, x) {
  const s = {};
  return t && (s.width = `${t}px`, s.minWidth = `${t}px`), e && (s.flexGrow = e), x && (s.height = `${x}px`), n && (s.position = "sticky", n.left && (s.left = `${r}px`), n.right && (s.right = `${p}px`)), s;
}
function Ht(t, e, n) {
  let r = "";
  if (t.fixed)
    for (const p in t.fixed) {
      let x = t.fixed[p] === -1;
      !x && t.fixed.leftSize && e.colspan && (x = e.colspan + t._colindex - 1 === t.fixed.leftSize), r += x ? "wx-shadow " : "wx-fixed ";
    }
  return r += e.rowspan > 1 ? "wx-rowspan " : "", r += e.colspan > 1 ? "wx-colspan " : "", r += e.vertical ? "wx-vertical " : "", r += n ? n(t) + " " : "", r;
}
function $n(t) {
  const {
    row: e,
    column: n,
    cellStyle: r = null,
    columnStyle: p = null,
    children: x
  } = t, [s, v] = pn(t.focusable), f = Z(ne), u = Y(f, "focusCell"), g = Y(f, "search"), a = Y(f, "reorder"), b = S(
    () => g?.rows[e.id] && g.rows[e.id][n.id],
    [g, e.id, n.id]
  ), D = S(
    () => Ye(
      n.width,
      n.flexgrow,
      n.fixed,
      n.left,
      n.right
    ),
    [n.width, n.flexgrow, n.fixed, n.left, n.right]
  );
  function C(N, l) {
    let H = "wx-cell";
    return H += n.fixed ? " " + (n.fixed === -1 ? "wx-shadow" : "wx-fixed") : "", H += N ? " " + N(n) : "", H += l ? " " + l(e, n) : "", H += n.treetoggle ? " wx-tree-cell" : "", H;
  }
  const d = S(
    () => C(p, r),
    [p, r, n, e]
  ), m = S(() => typeof n.draggable == "function" ? n.draggable(e, n) !== !1 : n.draggable, [n, e]), h = A(null);
  z(() => {
    h.current && s && u?.row === e.id && u?.column === n.id && h.current.focus();
  }, [u, s, e.id, n.id]);
  const y = W(() => {
    s && f.exec("focus-cell", {
      row: e.id,
      column: n.id,
      eventSource: "focus"
    });
  }, [f, s, e.id, n.id]);
  z(() => () => {
    s && u && (f.exec("focus-cell", { eventSource: "destroy" }), v(!1));
  }, [f, v]);
  function i(N) {
    const l = new RegExp(`(${g.value.trim()})`, "gi");
    return String(N).split(l).map((P) => ({ text: P, highlight: l.test(P) }));
  }
  const R = S(() => {
    const N = n.fixed && n.fixed.left === -1 || n.fixed.right === -1, l = n.fixed && n.fixed.right;
    return [
      d,
      N ? "wx-shadow" : "",
      l ? "wx-fixed-right" : ""
    ].filter(Boolean).join(" ");
  }, [d, n]), T = n.cell;
  return /* @__PURE__ */ q(
    "div",
    {
      className: "wx-TSCaXsGV " + R,
      ref: h,
      onFocus: y,
      style: D,
      "data-row-id": Ee(e.id),
      "data-col-id": Ee(n.id),
      tabIndex: s ? "0" : "-1",
      role: "gridcell",
      "aria-colindex": n._colindex,
      "aria-readonly": n.editor ? void 0 : !0,
      children: [
        a && n.draggable ? m ? /* @__PURE__ */ c(
          "i",
          {
            "draggable-data": "true",
            className: "wx-TSCaXsGV wx-draggable wxi-drag"
          }
        ) : /* @__PURE__ */ c("i", { className: "wx-TSCaXsGV wx-draggable-stub" }) : null,
        n.treetoggle ? /* @__PURE__ */ q(xe, { children: [
          /* @__PURE__ */ c("span", { style: { marginLeft: `${e.$level * 28}px` } }),
          e.$count ? /* @__PURE__ */ c(
            "i",
            {
              "data-action": "toggle-row",
              className: `wx-TSCaXsGV wx-table-tree-toggle wxi-menu-${e.open !== !1 ? "down" : "right"}`
            }
          ) : null
        ] }) : null,
        T ? /* @__PURE__ */ c(
          T,
          {
            api: f,
            row: e,
            column: n,
            onAction: ({ action: N, data: l }) => f.exec(N, l)
          }
        ) : x ? x() : b ? /* @__PURE__ */ c("span", { children: i(Pe(e, n)).map(
          ({ highlight: N, text: l }, H) => N ? /* @__PURE__ */ c("mark", { className: "wx-TSCaXsGV wx-search", children: l }, H) : /* @__PURE__ */ c("span", { children: l }, H)
        ) }) : Pe(e, n)
      ]
    }
  );
}
function vt(t, e) {
  let n, r;
  function p(v) {
    n = v.clientX, t.style.opacity = 1, document.body.style.cursor = "ew-resize", document.body.style.userSelect = "none", window.addEventListener("mousemove", x), window.addEventListener("mouseup", s), e && e.down && e.down(t);
  }
  function x(v) {
    r = v.clientX - n, e && e.move && e.move(r);
  }
  function s() {
    t.style.opacity = "", document.body.style.cursor = "", document.body.style.userSelect = "", e && e.up && e.up(r), window.removeEventListener("mousemove", x), window.removeEventListener("mouseup", s);
  }
  return t.addEventListener("mousedown", p), {
    destroy() {
      t.removeEventListener("mousedown", p);
    }
  };
}
function Mn({ filter: t, column: e, action: n, filterValue: r }) {
  function p({ value: x }) {
    n({ value: x, key: e.id });
  }
  return /* @__PURE__ */ c(
    qt,
    {
      ...t.config ?? {},
      value: r,
      onChange: p
    }
  );
}
function An({ filter: t, column: e, action: n, filterValue: r }) {
  const p = Z(ne), x = Y(p, "flatData"), s = S(
    () => t?.config?.options || e?.options || f(),
    [t, e, x]
  ), v = S(() => t?.config?.template, [t]);
  function f() {
    const a = [];
    return x.forEach((b) => {
      const D = on(b, e);
      a.includes(D) || a.push(D);
    }), a.map((b) => ({ id: b, label: b }));
  }
  function u({ value: a }) {
    n({ value: a, key: e.id });
  }
  function g(a) {
    a.key !== "Tab" && a.preventDefault();
  }
  return /* @__PURE__ */ c("div", { style: { width: "100%" }, onKeyDown: g, children: /* @__PURE__ */ c(
    Qt,
    {
      placeholder: "",
      clear: !0,
      ...t?.config ?? {},
      options: s,
      value: r,
      onChange: u,
      children: (a) => v ? v(a) : a.label
    }
  ) });
}
function In({ filter: t, column: e, action: n, filterValue: r }) {
  function p({ value: x }) {
    n({ value: x, key: e.id });
  }
  return /* @__PURE__ */ c("div", { className: "wx-aaaK5VYO", style: { width: "100%" }, children: /* @__PURE__ */ c(
    Zt,
    {
      placeholder: "",
      clear: !0,
      ...t?.config ?? {},
      value: r,
      onChange: p
    }
  ) });
}
const zn = {
  text: Mn,
  richselect: An,
  datepicker: In
};
function Vn({ filter: t, column: e }) {
  const n = Z(ne), r = Y(n, "filterValues");
  function p(s) {
    n.exec("filter-rows", s);
  }
  const x = S(() => zn[t.type], [t.type]);
  return /* @__PURE__ */ c(
    x,
    {
      filter: t,
      column: e,
      action: p,
      filterValue: r[e.id]
    }
  );
}
function On(t) {
  const {
    cell: e,
    column: n,
    row: r,
    lastRow: p,
    sortRow: x,
    columnStyle: s,
    bodyHeight: v,
    hasSplit: f
  } = t, u = Z(ne), g = Y(u, "sortMarks"), a = S(() => g ? g[n.id] : void 0, [g, n.id]), b = A(), D = W(
    (E) => {
      b.current = e.flexgrow ? E.parentNode.clientWidth : e.width;
    },
    [e.flexgrow, e.width]
  ), C = W(
    (E, $) => {
      u.exec("resize-column", {
        id: e.id,
        width: Math.max(1, (b.current || 0) + E),
        inProgress: $
      });
    },
    [u, e.id]
  ), d = W((E) => C(E, !0), [C]), m = W((E) => C(E, !1), [C]), h = W(
    (E) => {
      if (!n.sort || e.filter) return;
      let $ = a?.order;
      $ && ($ = $ === "asc" ? "desc" : "asc"), u.exec("sort-rows", { key: e.id, add: E.ctrlKey || E.metaKey, order: $ });
    },
    [u, e.id, e.filter, n.sort, a?.order]
  ), y = W(
    (E) => {
      E && E.stopPropagation(), u.exec("collapse-column", { id: e.id, row: r });
    },
    [u, e.id, r]
  ), i = W(
    (E) => {
      E.key === "Enter" && y();
    },
    [y]
  ), R = W(
    (E) => {
      E.key === "Enter" && !e.filter && h(E);
    },
    [h, e.filter]
  ), T = S(
    () => e.collapsed && n.collapsed,
    [e.collapsed, n.collapsed]
  ), N = S(
    () => T && !f && e.collapsible !== "header",
    [T, f, e.collapsible]
  ), l = S(
    () => N ? { top: -v / 2, position: "absolute" } : {},
    [N, v]
  ), H = S(
    () => Ye(
      e.width,
      e.flexgrow,
      n.fixed,
      n.left,
      e.right ?? n.right,
      e.height + (T && N ? v : 0)
    ),
    [
      e.width,
      e.flexgrow,
      n.fixed,
      n.left,
      e.right,
      n.right,
      e.height,
      T,
      N,
      v
    ]
  ), P = S(
    () => Ht(n, e, s),
    [n, e, s]
  ), G = W(() => Object.fromEntries(
    Object.entries(e).filter(([E]) => E !== "cell")
  ), [e]), ae = `wx-cell ${P} ${e.css || ""} wx-collapsed`, oe = [
    "wx-cell",
    P,
    e.css || "",
    e.filter ? "wx-filter" : "",
    n.fixed && n.fixed.right ? "wx-fixed-right" : ""
  ].filter(Boolean).join(" "), V = A(null);
  return z(() => {
    const E = V.current;
    if (!E) return;
    const $ = vt(E, { down: D, move: d, up: m });
    return () => {
      typeof $ == "function" && $();
    };
  }, [D, d, m, vt]), T ? /* @__PURE__ */ c(
    "div",
    {
      className: "wx-RsQD74qC " + ae,
      style: H,
      role: "button",
      "aria-label": `Expand column ${e.text || ""}`,
      "aria-expanded": !e.collapsed,
      tabIndex: 0,
      onKeyDown: i,
      onClick: y,
      "data-header-id": Ee(n.id),
      children: /* @__PURE__ */ c("div", { className: "wx-RsQD74qC wx-text", style: l, children: e.text || "" })
    }
  ) : /* @__PURE__ */ q(
    "div",
    {
      className: "wx-RsQD74qC " + oe,
      style: H,
      onClick: h,
      "data-header-id": Ee(n.id),
      tabIndex: !e._hidden && n.sort && !e.filter ? 0 : void 0,
      role: "columnheader",
      "aria-colindex": e._colindex,
      "aria-colspan": e.colspan > 1 ? e.colspan : void 0,
      "aria-rowspan": e.rowspan > 1 ? e.rowspan : void 0,
      "aria-sort": !a?.order || e.filter ? "none" : a?.order === "asc" ? "ascending" : "descending",
      onKeyDown: R,
      children: [
        e.collapsible ? /* @__PURE__ */ c(
          "div",
          {
            className: "wx-RsQD74qC wx-collapse",
            role: "button",
            "aria-label": e.collapsed ? "Expand column" : "Collapse column",
            "aria-expanded": !e.collapsed,
            tabIndex: 0,
            onKeyDown: i,
            onClick: y,
            children: /* @__PURE__ */ c(
              "i",
              {
                className: `wx-RsQD74qC wxi-angle-${e.collapsed ? "down" : "right"}`
              }
            )
          }
        ) : null,
        e.cell ? (() => {
          const E = e.cell;
          return /* @__PURE__ */ c(
            E,
            {
              api: u,
              cell: G(),
              column: n,
              row: r,
              onAction: ({ action: $, data: fe }) => u.exec($, fe)
            }
          );
        })() : e.filter ? /* @__PURE__ */ c(Vn, { filter: e.filter, column: n }) : /* @__PURE__ */ c("div", { className: "wx-RsQD74qC wx-text", children: e.text || "" }),
        n.resize && p && !e._hidden ? /* @__PURE__ */ c(
          "div",
          {
            className: "wx-RsQD74qC wx-grip",
            role: "presentation",
            "aria-label": "Resize column",
            ref: V,
            onClick: (E) => E.stopPropagation(),
            children: /* @__PURE__ */ c("div", {})
          }
        ) : null,
        x ? /* @__PURE__ */ c("div", { className: "wx-RsQD74qC wx-sort", children: a ? /* @__PURE__ */ q(xe, { children: [
          typeof a.index < "u" ? /* @__PURE__ */ c("div", { className: "wx-RsQD74qC wx-order", children: a.index + 1 }) : null,
          /* @__PURE__ */ c(
            "i",
            {
              className: `wx-RsQD74qC wxi-arrow-${a.order === "asc" ? "up" : "down"}`
            }
          )
        ] }) : null }) : null
      ]
    }
  );
}
function Kn({ cell: t, column: e, row: n, columnStyle: r }) {
  const p = Z(ne), x = S(
    () => Ye(
      t?.width,
      t?.flexgrow,
      e?.fixed,
      e?.left,
      t?.right ?? e?.right,
      t?.height
    ),
    [
      t?.width,
      t?.flexgrow,
      e?.fixed,
      e?.left,
      t?.right,
      e?.right,
      t?.height
    ]
  ), s = S(
    () => Ht(e, t, r),
    [e, t, r]
  ), v = W(() => Object.fromEntries(
    Object.entries(t || {}).filter(([u]) => u !== "cell")
  ), [t]), f = `wx-6Sdi3Dfd wx-cell ${s || ""} ${t?.css || ""}` + (e?.fixed && e?.fixed.right ? " wx-fixed-right" : "");
  return /* @__PURE__ */ c("div", { className: f, style: x, children: !e?.collapsed && !t?.collapsed ? t?.cell ? Ut.createElement(t.cell, {
    api: p,
    cell: v(),
    column: e,
    row: n,
    onAction: ({ action: u, data: g }) => p.exec(u, g)
  }) : /* @__PURE__ */ c("div", { className: "wx-6Sdi3Dfd wx-text", children: t?.text || "" }) : null });
}
function bt({
  deltaLeft: t,
  contentWidth: e,
  columns: n,
  type: r = "header",
  columnStyle: p,
  bodyHeight: x
}) {
  const s = Z(ne), v = Y(s, "_sizes"), f = Y(s, "split"), u = S(() => v?.[`${r}RowHeights`], [v, r]), g = S(() => {
    let d = [];
    if (n && n.length) {
      const m = n[0][r].length;
      for (let h = 0; h < m; h++) {
        let y = 0;
        d.push([]), n.forEach((i, R) => {
          const T = { ...i[r][h] };
          if (y || d[h].push(T), T.colspan > 1) {
            if (y = T.colspan - 1, !rn() && i.right) {
              let N = i.right;
              for (let l = 1; l < T.colspan; l++)
                N -= n[R + l].width;
              T.right = N;
            }
          } else y && y--;
        });
      }
    }
    return d;
  }, [n, r]), a = S(() => f?.left || f?.right, [f]);
  function b(d) {
    return n.find((m) => m.id === d);
  }
  function D(d, m) {
    let h = m;
    return d.rowspan && (h += d.rowspan - 1), h === g.length - 1;
  }
  function C(d, m, h) {
    if (!h.sort) return !1;
    for (let y = g.length - 1; y >= 0; y--) {
      const i = h.header[y];
      if (!i.filter && !i._hidden) return m === y;
    }
    return D(d, m);
  }
  return /* @__PURE__ */ c(
    "div",
    {
      className: `wx-sAsPVaUK wx-${r}`,
      style: { paddingLeft: `${t}px`, width: `${e}px` },
      role: "rowgroup",
      children: g.map((d, m) => /* @__PURE__ */ c(
        "div",
        {
          className: r === "header" ? "wx-sAsPVaUK wx-h-row" : "wx-sAsPVaUK wx-f-row",
          style: { height: `${u?.[m]}px`, display: "flex" },
          role: "row",
          children: d.map((h) => {
            const y = b(h.id);
            return r === "header" ? /* @__PURE__ */ c(
              On,
              {
                cell: h,
                columnStyle: p,
                column: y,
                row: m,
                lastRow: D(h, m),
                bodyHeight: x,
                sortRow: C(h, m, y),
                hasSplit: a
              },
              h.id
            ) : /* @__PURE__ */ c(
              Kn,
              {
                cell: h,
                columnStyle: p,
                column: b(h.id),
                row: m
              },
              h.id
            );
          })
        },
        m
      ))
    }
  );
}
function Pn({ overlay: t }) {
  const e = Z(ne);
  function n(p) {
    return typeof p == "function";
  }
  const r = t;
  return /* @__PURE__ */ c("div", { className: "wx-1ty666CQ wx-overlay", children: n(t) ? /* @__PURE__ */ c(r, { onAction: ({ action: p, data: x }) => e.exec(p, x) }) : t });
}
function Wn(t) {
  const { editor: e, onSave: n, onApply: r } = t, [p, x] = K(e?.value || ""), s = A(null);
  z(() => {
    s.current && s.current.focus();
  }, []), z(() => {
    if (!s.current) return;
    const u = Le(s.current, () => n(!0));
    return () => u.destroy();
  }, [n]);
  function v() {
    s.current && (x(s.current.value), r(s.current.value));
  }
  function f({ key: u }) {
    u === "Enter" && n();
  }
  return /* @__PURE__ */ c(
    "input",
    {
      className: "wx-e7Ao5ejY wx-text",
      onInput: v,
      onKeyDown: f,
      ref: s,
      type: "text",
      value: p
    }
  );
}
function Yn({ editor: t, onAction: e, onSave: n, onApply: r, onCancel: p }) {
  const [x, s] = K(t?.value), [v, f] = K(t?.renderedValue), [u, g] = K(t?.options || []), a = S(() => t?.config?.template, [t]), b = S(() => t?.config?.cell, [t]), D = S(() => t?.config?.dropdown || {}, [t]), C = S(() => ({ trackScroll: !0, ...D }), [D]), d = S(() => (u || []).findIndex((l) => l.id === x), [u, x]), m = A(null), h = A(null), y = W(
    (l) => {
      m.current = l.navigate, h.current = l.keydown, m.current(d);
    },
    [d, m]
  ), i = W(
    (l) => {
      const H = l?.target?.value ?? "";
      f(H);
      const P = H ? (t?.options || []).filter(
        (G) => (G.label || "").toLowerCase().includes(H.toLowerCase())
      ) : t?.options || [];
      g(P), P.length ? m.current(-1 / 0) : m.current(null);
    },
    [t]
  ), R = A(null);
  z(() => {
    R.current && R.current.focus();
  }, []);
  const T = A(n);
  z(() => {
    T.current = n;
  }, [n]), z(() => {
    if (!R.current) return;
    const l = Le(R.current, () => T.current(!0));
    return () => l.destroy();
  }, []), z(() => {
    s(t?.value), f(t?.renderedValue), g(t?.options || []);
  }, [t]);
  const N = W(
    ({ id: l }) => {
      r(l), n();
    },
    [r, n]
  );
  return /* @__PURE__ */ q(xe, { children: [
    /* @__PURE__ */ c(
      "input",
      {
        className: "wx-0UYfSd1x wx-input",
        ref: R,
        value: v ?? "",
        onChange: i,
        onKeyDown: (l) => h.current ? h.current(l, d) : void 0
      }
    ),
    /* @__PURE__ */ c(
      nt,
      {
        items: u,
        onReady: y,
        onSelect: N,
        ...C,
        onCancel: p,
        children: ({ option: l }) => a ? a(l) : b ? /* @__PURE__ */ c(b, { data: l, onAction: e }) : l.label
      }
    )
  ] });
}
function _n({ editor: t, onAction: e, onSave: n, onApply: r, onCancel: p }) {
  const [x] = K(() => t.value || /* @__PURE__ */ new Date()), { template: s, cell: v, dropdown: f = {} } = t?.config || {}, u = S(() => ({
    trackScroll: !0,
    width: "auto",
    ...f
  }), [f]);
  function g({ value: C }) {
    r(C), n();
  }
  const a = A(null), [b, D] = K(null);
  return z(() => {
    a.current && a.current.focus(), typeof window < "u" && window.getSelection && window.getSelection().removeAllRanges();
  }, []), z(() => {
    if (!b) return;
    const C = Le(b, () => n(!0));
    return () => C.destroy();
  }, [b, n]), /* @__PURE__ */ q(xe, { children: [
    /* @__PURE__ */ c(
      "div",
      {
        className: "wx-lNWNYUb6 wx-value",
        ref: a,
        tabIndex: 0,
        onClick: p,
        onKeyDown: (C) => C.preventDefault(),
        children: s ? s(x) : v ? /* @__PURE__ */ c(v, { data: t.value, onAction: e }) : /* @__PURE__ */ c("span", { className: "wx-lNWNYUb6 wx-text", children: t.renderedValue })
      }
    ),
    /* @__PURE__ */ c(Jt, { ...u, onCancel: p, children: /* @__PURE__ */ c("div", { ref: D, className: "wx-lNWNYUb6", children: /* @__PURE__ */ c(
      en,
      {
        value: x,
        onChange: g,
        buttons: t.config?.buttons
      }
    ) }) })
  ] });
}
function Fn(t) {
  const { editor: e } = t, n = t.onAction ?? t.onaction, r = t.onSave ?? t.onsave, p = t.onApply ?? t.onapply, x = t.onCancel ?? t.oncancel, s = e.config || {}, { dropdown: v = {} } = s, f = S(
    () => ({ trackScroll: !0, ...v }),
    [v]
  ), [u] = K(
    e.options.find((i) => i.id === e.value)
  ), [g] = K(e.value), [a] = K(e.options), b = S(
    () => a.findIndex((i) => i.id === g),
    [a, g]
  );
  function D({ id: i }) {
    p(i), r();
  }
  let C;
  const [d, m] = K();
  function h(i) {
    C = i.navigate, m(() => i.keydown), C(b);
  }
  const y = A(null);
  return z(() => {
    y.current && y.current.focus(), typeof window < "u" && window.getSelection && window.getSelection().removeAllRanges();
  }, []), z(() => {
    if (!y.current) return;
    const i = Le(y.current, () => r(!0));
    return () => i.destroy();
  }, [r]), /* @__PURE__ */ q(xe, { children: [
    /* @__PURE__ */ c(
      "div",
      {
        ref: y,
        className: "wx-ywGRk611 wx-value",
        tabIndex: 0,
        onClick: x,
        onKeyDown: (i) => {
          d(i, b), i.preventDefault();
        },
        children: s.template ? s.template(u) : s.cell ? (() => {
          const i = s.cell;
          return /* @__PURE__ */ c(i, { data: u, onAction: n });
        })() : /* @__PURE__ */ c("span", { className: "wx-ywGRk611 wx-text", children: e.renderedValue })
      }
    ),
    /* @__PURE__ */ c(nt, { items: a, onReady: h, onSelect: D, ...f, onCancel: x, children: ({ option: i }) => s.template ? s.template(i) : s.cell ? (() => {
      const R = s.cell;
      return /* @__PURE__ */ c(R, { data: i, onAction: n });
    })() : i.label })
  ] });
}
function Xn(t) {
  const { editor: e, onAction: n, onApply: r, onCancel: p, onSave: x } = t, s = e?.config || {}, v = S(() => ({ trackScroll: !0, ...s?.dropdown || {} }), [s]), f = e?.options ?? [], u = e?.value || [], g = e?.renderedValue, a = f.find((i) => u.includes(i.id)), b = a ? f.indexOf(a) : -1, D = A(null);
  function C({ id: i }) {
    r(i), D.current && D.current.focus();
  }
  let d;
  const [m, h] = K();
  function y(i) {
    d = i.navigate, h(() => i.keydown), d(b);
  }
  return z(() => {
    D.current && D.current.focus(), typeof window < "u" && window.getSelection && window.getSelection().removeAllRanges();
  }, []), z(() => {
    if (!D.current) return;
    const i = Le(D.current, () => x(!0));
    return () => i.destroy();
  }, [x]), /* @__PURE__ */ q(xe, { children: [
    /* @__PURE__ */ c(
      "div",
      {
        ref: D,
        className: "wx-value wx-aacZ4gNU",
        tabIndex: 0,
        onClick: p,
        onKeyDown: (i) => {
          m(i, b), i.preventDefault();
        },
        children: s.template ? s.template(u?.map((i) => f.find((R) => R.id === i))) : s.cell ? (() => {
          const i = s.cell;
          return /* @__PURE__ */ c(
            i,
            {
              data: u.map((R) => f.find((T) => T.id === R))
            }
          );
        })() : /* @__PURE__ */ c("span", { className: "wx-text wx-aacZ4gNU", children: g })
      }
    ),
    /* @__PURE__ */ c(
      nt,
      {
        items: f,
        onReady: y,
        onSelect: C,
        checkboxes: !0,
        multiselect: !0,
        ...v,
        onCancel: p,
        value: u,
        children: ({ option: i }) => /* @__PURE__ */ c("div", { className: "wx-option wx-aacZ4gNU", children: s.template ? s.template(i) : s.cell ? (() => {
          const R = s.cell;
          return /* @__PURE__ */ c(R, { data: i, onAction: n });
        })() : i.label })
      }
    )
  ] });
}
const $t = {
  text: Wn,
  combo: Yn,
  datepicker: _n,
  richselect: Fn,
  multiselect: Xn
};
function co(t, e) {
  $t[t] = e;
}
function Un({ column: t, row: e }) {
  const n = Z(ne), r = Y(n, "editor"), p = W(
    (C, d) => {
      n.exec("close-editor", { ignore: C }), d && setTimeout(() => {
        n.exec("focus-cell", {
          ...d,
          eventSource: "click"
        });
      });
    },
    [n]
  ), x = W(
    (C) => {
      const d = C ? null : { row: r?.id, column: r?.column };
      p(!1, d);
    },
    [r, p]
  ), s = W(() => {
    p(!0, { row: r?.id, column: r?.column });
  }, [r, p]), v = W(
    (C) => {
      n.exec("editor", { value: C });
    },
    [n]
  ), f = W(
    (C) => {
      C.key === "Enter" && r && (t.editor.type === "multiselect" ? v(r.value) : s());
    },
    [r, t.editor, s, v]
  ), u = S(
    () => Ye(
      t.width,
      t.flexgrow,
      t.fixed,
      t.left,
      t.right
    ),
    [t.width, t.flexgrow, t.fixed, t.left, t.right]
  ), g = S(() => {
    let C = t.editor;
    if (typeof C == "function" && (C = C(e, t)), !C) return null;
    let d = typeof C == "string" ? C : C.type;
    return $t[d];
  }, [t, e]), a = A(null);
  z(() => {
    a.current && typeof u == "string" && a.current.setAttribute("style", u);
  }, [u]);
  const b = typeof e.$parent < "u" ? "gridcell" : "cell", D = typeof e.$parent < "u" ? !t.editor : void 0;
  return /* @__PURE__ */ c(
    "div",
    {
      className: "wx-8l724t2g wx-cell wx-editor",
      ref: a,
      style: typeof u == "object" && u !== null ? u : void 0,
      role: b,
      "aria-readonly": D,
      tabIndex: -1,
      onClick: (C) => C.stopPropagation(),
      onDoubleClick: (C) => C.stopPropagation(),
      onKeyDown: f,
      children: g ? /* @__PURE__ */ c(
        g,
        {
          editor: r,
          onSave: x,
          onApply: v,
          onCancel: s,
          onAction: ({ action: C, data: d }) => n.exec(C, d)
        }
      ) : null
    }
  );
}
function Ct(t) {
  const { columns: e, type: n, columnStyle: r } = t, p = Z(ne), { filterValues: x, _columns: s, _sizes: v } = p.getState();
  function f(u) {
    return r ? " " + r(u) : "";
  }
  return /* @__PURE__ */ c(xe, { children: e.map((u, g) => /* @__PURE__ */ c("tr", { children: u.map((a) => {
    const b = s.find((d) => d.id === a.id), D = `wx-print-cell-${n}${f(b)}${a.filter ? " wx-print-cell-filter" : ""}${a.vertical ? " wx-vertical" : ""}`, C = a.cell;
    return /* @__PURE__ */ c(
      "th",
      {
        style: Dt(St(a, v.columnWidth)),
        className: "wx-Gy81xq2u " + D,
        rowSpan: a.rowspan,
        colSpan: a.colspan,
        children: C ? /* @__PURE__ */ c(
          C,
          {
            api: p,
            cell: Object.fromEntries(
              Object.entries(a).filter(([d]) => d !== "cell")
            ),
            column: b,
            row: g
          }
        ) : a.filter ? /* @__PURE__ */ c("div", { className: "wx-Gy81xq2u wx-print-filter", children: ln(x, s, a) }) : /* @__PURE__ */ c("div", { className: "wx-Gy81xq2u wx-text", children: a.text ?? "" })
      },
      a.id
    );
  }) }, g)) });
}
function jn(t) {
  const { columns: e, rowStyle: n, columnStyle: r, cellStyle: p, header: x, footer: s, reorder: v } = t, f = Z(ne), { flatData: u, _sizes: g } = f.getState(), a = x && xt(e, "header", g.headerRowHeights), b = s && xt(e, "footer", g.footerRowHeights);
  function D(d, m) {
    let h = "";
    return h += r ? " " + r(m) : "", h += p ? " " + p(d, m) : "", h;
  }
  function C(d, m) {
    return typeof m.draggable == "function" ? m.draggable(d, m) !== !1 : m.draggable;
  }
  return /* @__PURE__ */ q(
    "table",
    {
      className: `wx-8NTMLH0z wx-print-grid ${e.some((d) => d.flexgrow) ? "wx-flex-columns" : ""}`,
      children: [
        x ? /* @__PURE__ */ c("thead", { children: /* @__PURE__ */ c(
          Ct,
          {
            columns: a,
            type: "header",
            columnStyle: r
          }
        ) }) : null,
        /* @__PURE__ */ c("tbody", { children: u.map((d, m) => /* @__PURE__ */ c(
          "tr",
          {
            className: "wx-8NTMLH0z wx-row" + (n ? " " + n(d) : ""),
            style: { height: `${d.rowHeight || g.rowHeight}px` },
            children: e.map(
              (h) => h.collapsed ? null : /* @__PURE__ */ q(
                "td",
                {
                  className: `wx-8NTMLH0z wx-print-cell wx-cell ${D(d, h)}`,
                  style: Dt(
                    St(h, g.columnWidth)
                  ),
                  children: [
                    v && h.draggable ? /* @__PURE__ */ c("span", { className: "wx-8NTMLH0z wx-print-draggable", children: C(d, h) ? /* @__PURE__ */ c("i", { className: "wx-8NTMLH0z wxi-drag" }) : null }) : null,
                    h.treetoggle ? /* @__PURE__ */ q(xe, { children: [
                      /* @__PURE__ */ c(
                        "span",
                        {
                          style: { marginLeft: d.$level * 28 + "px" }
                        }
                      ),
                      d.$count ? /* @__PURE__ */ c(
                        "i",
                        {
                          className: `wx-8NTMLH0z wx-print-grid-tree-toggle wxi-menu-${d.open !== !1 ? "down" : "right"}`
                        }
                      ) : null
                    ] }) : null,
                    h.cell ? (() => {
                      const y = h.cell;
                      return /* @__PURE__ */ c(y, { api: f, row: d, column: h });
                    })() : /* @__PURE__ */ c("span", { children: Pe(d, h) })
                  ]
                },
                h.id
              )
            )
          },
          m
        )) }),
        s ? /* @__PURE__ */ c("tfoot", { children: /* @__PURE__ */ c(
          Ct,
          {
            columns: b,
            type: "footer",
            columnStyle: r
          }
        ) }) : null
      ]
    }
  );
}
function Bn(t) {
  const { config: e, ...n } = t, r = Z(ne), { _skin: p, _columns: x } = r.getState(), s = S(() => sn(x, e), []), v = A(null);
  return z(() => {
    const f = document.body;
    f.classList.add("wx-print");
    const u = v.current;
    if (!u) return;
    const g = u.cloneNode(!0);
    f.appendChild(g);
    const a = `@media print { @page { size: ${e.paper} ${e.mode}; }`, b = document.createElement("style");
    b.setAttribute("type", "text/css"), b.setAttribute("media", "print"), document.getElementsByTagName("head")[0].appendChild(b), b.appendChild(document.createTextNode(a)), window.print(), b.remove(), f.classList.remove("wx-print"), g.remove();
  }, []), /* @__PURE__ */ c(
    "div",
    {
      className: `wx-4zwCKA7C wx-${p}-theme wx-print-container`,
      ref: v,
      children: s.map((f, u) => /* @__PURE__ */ c("div", { className: "wx-4zwCKA7C wx-print-grid-wrapper", children: /* @__PURE__ */ c(jn, { columns: f, ...n }) }, u))
    }
  );
}
function Gn(t) {
  const {
    header: e,
    footer: n,
    overlay: r,
    multiselect: p,
    onreorder: x,
    rowStyle: s,
    columnStyle: v,
    cellStyle: f,
    autoRowHeight: u,
    resize: g,
    clientWidth: a,
    clientHeight: b,
    responsiveLevel: D,
    hotkeys: C
  } = t, d = Z(ne), m = Y(d, "dynamic"), h = Y(d, "_columns"), y = Y(d, "flatData"), i = Y(d, "split"), R = Y(d, "_sizes"), [T, N] = gn(d, "selectedRows"), l = Y(d, "select"), H = Y(d, "editor"), P = Y(d, "tree"), G = Y(d, "focusCell"), ae = Y(d, "_print"), oe = Y(d, "undo"), V = Y(d, "reorder"), E = Y(d, "_rowHeightFromData"), [$, fe] = K(0);
  z(() => {
    fe(Wt());
  }, []);
  const [Se, _e] = K(0), [me, Fe] = K(0), se = S(() => (h || []).some((o) => !o.hidden && o.flexgrow), [h]), j = S(() => R?.rowHeight || 0, [R]), X = A(null), [Xe, Re] = K(null), [de, Me] = K(null), te = S(() => {
    let o = [], w = 0;
    return i && i.left && (o = (h || []).slice(0, i.left).filter((k) => !k.hidden).map((k) => ({ ...k })), o.forEach((k) => {
      k.fixed = { left: 1, leftSize: i.left }, k.left = w, w += k.width;
    }), o.length && (o[o.length - 1].fixed.left = -1)), { columns: o, width: w };
  }, [i, h]), ie = S(() => {
    let o = [], w = 0;
    if (i && i.right) {
      o = (h || []).slice(i.right * -1).filter((k) => !k.hidden).map((k) => ({ ...k }));
      for (let k = o.length - 1; k >= 0; k--) {
        const M = o[k];
        M.fixed = { right: 1 }, M.right = w, w += M.width;
      }
      o.length && (o[0].fixed = { right: -1 });
    }
    return { columns: o, width: w };
  }, [i, h]), I = S(() => {
    const o = (h || []).slice(i?.left || 0, (h || []).length - (i?.right ?? 0)).filter((w) => !w.hidden);
    return o.forEach((w) => {
      w.fixed = 0;
    }), o;
  }, [h, i]), U = S(() => (h || []).reduce((o, w) => (w.hidden || (o += w.width), o), 0), [h]), J = 1;
  function ue(o, w, k) {
    let M = w, _ = o;
    if (I.length) {
      let O = I.length;
      for (let L = o; L >= 0; L--)
        I[L][k].forEach((Q) => {
          Q.colspan > 1 && L > o - Q.colspan && L < O && (O = L);
        });
      if (O !== I.length && O < o) {
        for (let L = O; L < o; L++)
          M -= I[L].width;
        _ = O;
      }
    }
    return { index: _, delta: M };
  }
  const ee = S(() => {
    let o, w, k;
    const M = Se, _ = Se + (a || 0);
    let O = 0, L = 0, F = 0, Q = 0;
    I.forEach((ge, ke) => {
      M > F && (O = ke, Q = F), F = F + ge.width, _ > F && (L = ke + J);
    });
    const ce = { header: 0, footer: 0 };
    for (let ge = L; ge >= O; ge--)
      ["header", "footer"].forEach((ke) => {
        I[ge] && I[ge][ke].forEach((Xt) => {
          const Je = Xt.colspan;
          if (Je && Je > 1) {
            const ht = Je - (L - ge + 1);
            ht > 0 && (ce[ke] = Math.max(ce[ke], ht));
          }
        });
      });
    const B = ue(O, Q, "header"), le = ue(O, Q, "footer"), Ze = B.delta, $e = B.index, Ve = le.delta, Oe = le.index;
    return se && U > (a || 0) ? o = w = k = [...te.columns, ...I, ...ie.columns] : (o = [
      ...te.columns,
      ...I.slice(O, L + 1),
      ...ie.columns
    ], w = [
      ...te.columns,
      ...I.slice($e, L + ce.header + 1),
      ...ie.columns
    ], k = [
      ...te.columns,
      ...I.slice(Oe, L + ce.footer + 1),
      ...ie.columns
    ]), {
      data: o || [],
      header: w || [],
      footer: k || [],
      d: Q,
      df: Ve,
      dh: Ze
    };
  }, [
    I,
    te,
    ie,
    Se,
    a,
    se,
    U
  ]), ve = S(
    () => e && R?.headerHeight || 0,
    [e, R]
  ), be = S(
    () => n && R?.footerHeight || 0,
    [n, R]
  ), ye = S(() => a && b ? U >= a : !1, [a, b, U]), Ne = S(() => (b || 0) - ve - be - (ye ? $ : 0), [b, ve, be, ye, $]), lt = S(() => Math.ceil((Ne || 0) / (j || 1)) + 1, [Ne, j]), Ue = A([]), [st, Mt] = K(0), [it, At] = K(void 0), re = S(() => {
    let o = 0, w = 0;
    const k = 2;
    if (u) {
      let O = me;
      for (; O > 0; )
        O -= Ue.current[o] || j, o++;
      w = me - O;
      for (let L = Math.max(0, o - k - 1); L < o; L++)
        w -= Ue.current[o - L] || j;
      o = Math.max(0, o - k);
    } else {
      if (E) {
        let O = 0, L = 0;
        for (let B = 0; B < (y || []).length; B++) {
          const le = y[B].rowHeight || j;
          if (L + le > me) {
            O = B;
            break;
          }
          L += le;
        }
        o = Math.max(0, O - k);
        for (let B = 0; B < o; B++)
          w += y[B].rowHeight || j;
        let F = 0, Q = 0;
        for (let B = O + 1; B < (y || []).length; B++) {
          const le = y[B].rowHeight || j;
          if (F++, Q + le > Ne)
            break;
          Q += le;
        }
        const ce = Math.min(
          m ? m.rowCount : (y || []).length,
          O + F + k
        );
        return { d: w, start: o, end: ce };
      }
      o = Math.floor(me / (j || 1)), o = Math.max(0, o - k), w = o * (j || 0);
    }
    const M = m ? m.rowCount : (y || []).length, _ = Math.min(M, o + (lt || 0) + k);
    return { d: w, start: o, end: _ };
  }, [u, E, me, j, m, y, lt, Ne]), je = S(() => {
    const o = m ? m.rowCount : (y || []).length;
    if (u)
      return st + re.d + (o - (it || 0)) * (j || 0);
    if (!E)
      return o * (j || 0);
    let w = 0;
    for (let k = 0; k < o; k++)
      w += y[k]?.rowHeight || j;
    return w;
  }, [
    m,
    y,
    j,
    u,
    E,
    st,
    re.d,
    it
  ]), he = S(() => a && b ? je + ve + be >= b - (U >= (a || 0) ? $ : 0) : !1, [
    a,
    b,
    je,
    ve,
    be,
    U,
    $
  ]), Ce = S(() => se && U <= (a || 0) ? (a || 0) - 0 - (he ? $ : 0) : U, [se, U, a, he, $, ye]), Be = S(() => se && U <= (a || 0) ? a || 0 : Ce < (a || 0) ? U + (he ? $ : 0) : -1, [se, U, a, Ce, he, $]), Ge = A({});
  z(() => {
    if (m && (Ge.current.start !== re.start || Ge.current.end !== re.end)) {
      const { start: o, end: w } = re;
      Ge.current = { start: o, end: w }, d && d.exec && d.exec("request-data", { row: { start: o, end: w } });
    }
  }, [m, re, d]);
  const pe = S(() => m ? y || [] : (y || []).slice(re.start, re.end), [m, y, re]), He = S(() => (T || []).filter(
    (o) => (pe || []).some((w) => w.id === o)
  ), [N, pe]), qe = S(() => re.start, [re.start]), It = W((o) => {
    Fe(o.target.scrollTop), _e(o.target.scrollLeft);
  }, []), zt = W((o) => {
    o.shiftKey && o.preventDefault(), X.current && X.current.focus && X.current.focus();
  }, []), ct = W(() => !!(h || []).find((o) => !!o.draggable), [h]), Qe = A(null), De = A(null), Vt = A({
    dblclick: (o, w) => {
      const k = { id: o, column: mt(w, "data-col-id") };
      d.exec("open-editor", k);
    },
    click: (o, w) => {
      if (w.target.closest("input") || Qe.current) return;
      const k = mt(w, "data-col-id");
      if (G?.id !== o && d.exec("focus-cell", {
        row: o,
        column: k,
        eventSource: "click"
      }), l === !1) return;
      const M = p && (w.ctrlKey || w.metaKey), _ = p && w.shiftKey;
      (M || T.length > 1 || !T.includes(o)) && d.exec("select-row", { id: o, toggle: M, range: _ });
    },
    "toggle-row": (o) => {
      const w = d.getRow(o);
      d.exec(w.open !== !1 ? "close-row" : "open-row", { id: o });
    },
    "ignore-click": () => !1
  }), Ae = S(() => ({
    top: ve,
    bottom: be,
    left: te.width,
    xScroll: ye,
    yScroll: he,
    sense: u && de ? de.offsetHeight : Math.max(R?.rowHeight || 0, 40),
    node: X.current && X.current.firstElementChild
  }), [
    ve,
    be,
    te.width,
    ye,
    he,
    u,
    de,
    R
  ]);
  function Ot(o, w) {
    const { container: k, sourceNode: M, from: _ } = w;
    if (ct() && !M.getAttribute("draggable-data"))
      return !1;
    Re(_), P && d.getRow(_).open && d.exec("close-row", { id: _, nested: !0 });
    const L = et(M), F = L.cloneNode(!0);
    F.classList.remove("wx-selected"), F.querySelectorAll("[tabindex]").forEach((le) => le.setAttribute("tabindex", "-1")), k.appendChild(F), Me(F);
    const Q = Se - ee.d, ce = he ? $ : 0;
    k.style.width = Math.min(
      (a || 0) - ce,
      se && U <= (a || 0) ? Ce : Ce - ce
    ) + Q + "px";
    const B = Ke(L);
    w.offset = {
      x: Q,
      y: -Math.round(B.height / 2)
    }, De.current || (De.current = o.clientY);
  }
  function Kt(o, w) {
    const { from: k } = w, M = w.pos, _ = Ke(X.current);
    M.x = _.x;
    const O = Ae.top;
    if (M.y < O) M.y = O;
    else {
      const L = _.height - (ye && $ > 0 ? $ : Math.round(Ae.sense / 2)) - Ae.bottom;
      M.y > L && (M.y = L);
    }
    if (X.current.contains(w.targetNode)) {
      const L = et(w.targetNode), F = L && We(L);
      if (F && F !== k) {
        w.to = F;
        const Q = u ? de?.offsetHeight : R?.rowHeight;
        if (de && (me === 0 || M.y > O + Q - 1)) {
          const ce = L.getBoundingClientRect(), le = Ke(de).y, Ze = ce.y, $e = le > Ze ? -1 : 1, Ve = $e === 1 ? "after" : "before", Oe = Math.abs(d.getRowIndex(k) - d.getRowIndex(F)), ge = Oe !== 1 ? Ve === "before" ? "after" : "before" : Ve;
          if (Oe === 1 && ($e === -1 && o.clientY > De.current || $e === 1 && o.clientY < De.current))
            return;
          De.current = o.clientY, d.exec("move-item", {
            id: k,
            target: F,
            mode: ge,
            inProgress: !0
          });
        }
      }
      x && x({ event: o, context: w });
    }
    Tn(o, _, w, Ae);
  }
  function Pt(o, w) {
    const { from: k, to: M } = w;
    d.exec("move-item", {
      id: k,
      target: M,
      inProgress: !1
    }), Qe.current = setTimeout(() => {
      Qe.current = 0;
    }, 1), Re(null), Me(null), De.current = null, Et(w);
  }
  function Wt() {
    const o = document.createElement("div");
    o.style.cssText = "position:absolute;left:-1000px;width:100px;padding:0px;margin:0px;min-height:100px;overflow-y:scroll;", document.body.appendChild(o);
    const w = o.offsetWidth - o.clientWidth;
    return document.body.removeChild(o), w;
  }
  const Yt = S(() => Be > 0 ? { width: `${Be}px` } : void 0, [Be]), at = A(null);
  function _t() {
    Promise.resolve().then(() => {
      let o = 0, w = qe;
      const k = at.current;
      k && (Array.from(k.children).forEach((M, _) => {
        Ue.current[qe + _] = M.offsetHeight, o += M.offsetHeight, w++;
      }), Mt(o), At(w));
    });
  }
  z(() => {
    pe && u && _t();
  }, [pe, u, qe]);
  let [Te, Ie] = K();
  z(() => {
    if (G && (!l || !He.length || He.includes(G.row)))
      Ie({ ...G });
    else if (pe.length && ee.data.length) {
      if (!Te || He.length && !He.includes(Te.row) || pe.findIndex((o) => o.id === Te.row) === -1 || ee.data.findIndex(
        (o) => o.id === Te.column && !o.collapsed
      ) === -1) {
        const o = He[0] || pe[0].id, w = ee.data.findIndex((k) => !k.collapsed);
        Ie(w !== -1 ? { row: o, column: ee.data[w].id } : null);
      }
    } else Ie(null);
  }, [G]);
  const dt = A(null);
  z(() => {
    const o = X.current;
    if (!o) return;
    const w = Cn(o, g);
    return () => {
      typeof w == "function" && w();
    };
  }, [g]);
  const ut = A({});
  Object.assign(ut.current, {
    start: Ot,
    move: Kt,
    end: Pt,
    getReorder: () => V,
    getDraggableInfo: () => ({ hasDraggable: ct() })
  }), z(() => {
    const o = X.current;
    return o ? Rn(o, ut).destroy : void 0;
  }, [V, X.current]), z(() => {
    const o = X.current;
    return o ? cn(o, {
      keys: C !== !1 && {
        ...an,
        "ctrl+z": oe,
        "ctrl+y": oe,
        ...C
      },
      exec: (k) => d.exec("hotkey", k)
    }).destroy : void 0;
  }, [d, oe, C]);
  const ze = A({
    scroll: d.getReactiveState().scroll
  });
  ze.current.getWidth = () => (a || 0) - (he ? $ : 0), ze.current.getHeight = () => Ne, ze.current.getScrollMargin = () => te.width + ie.width, z(() => {
    dn(dt.current, ze.current);
  }, []);
  const ft = A(null);
  z(() => {
    const o = ft.current;
    if (!o) return;
    const w = [];
    return w.push(
      Le(o, () => G && d.exec("focus-cell", { eventSource: "click" })).destroy
    ), w.push(wn(o, Vt.current)), () => w.forEach((k) => k());
  }, []);
  const Ft = `wx-grid ${D ? `wx-responsive-${D}` : ""}`;
  return /* @__PURE__ */ q(xe, { children: [
    /* @__PURE__ */ c(
      "div",
      {
        className: "wx-4VuBwK2D " + Ft,
        style: {
          "--header-height": `${ve}px`,
          "--footer-height": `${be}px`,
          "--split-left-width": `${te.width}px`,
          "--split-right-width": `${ie.width}px`
        },
        children: /* @__PURE__ */ c(
          "div",
          {
            ref: X,
            className: "wx-4VuBwK2D wx-table-box",
            style: Yt,
            role: P ? "treegrid" : "grid",
            "aria-colcount": ee.data.length,
            "aria-rowcount": pe.length,
            "aria-multiselectable": P && p ? !0 : void 0,
            tabIndex: -1,
            children: /* @__PURE__ */ q(
              "div",
              {
                ref: dt,
                className: "wx-4VuBwK2D wx-scroll",
                style: {
                  overflowX: ye ? "scroll" : "hidden",
                  overflowY: he ? "scroll" : "hidden"
                },
                onScroll: It,
                children: [
                  e ? /* @__PURE__ */ c("div", { className: "wx-4VuBwK2D wx-header-wrapper", children: /* @__PURE__ */ c(
                    bt,
                    {
                      contentWidth: Ce,
                      deltaLeft: ee.dh,
                      columns: ee.header,
                      columnStyle: v,
                      bodyHeight: Ne - +n
                    }
                  ) }) : null,
                  /* @__PURE__ */ q(
                    "div",
                    {
                      ref: ft,
                      className: "wx-4VuBwK2D wx-body",
                      style: { width: `${Ce}px`, height: `${je}px` },
                      onMouseDown: (o) => zt(o),
                      children: [
                        r ? /* @__PURE__ */ c(Pn, { overlay: r }) : null,
                        /* @__PURE__ */ c(
                          "div",
                          {
                            ref: at,
                            className: "wx-4VuBwK2D wx-data",
                            style: {
                              paddingTop: `${re.d}px`,
                              paddingLeft: `${ee.d}px`
                            },
                            children: pe.map((o, w) => {
                              const k = T.indexOf(o.id) !== -1, M = Xe === o.id, _ = "wx-row" + (u ? " wx-autoheight" : "") + (s ? " " + s(o) : "") + (k ? " wx-selected" : "") + (M ? " wx-inactive" : ""), O = u ? { minHeight: `${o.rowHeight || j}px` } : { height: `${o.rowHeight || j}px` };
                              return /* @__PURE__ */ c(
                                "div",
                                {
                                  className: "wx-4VuBwK2D " + _,
                                  "data-id": Ee(o.id),
                                  "data-context-id": Ee(o.id),
                                  style: O,
                                  role: "row",
                                  "aria-rowindex": w,
                                  "aria-expanded": o.open,
                                  "aria-level": P ? o.$level + 1 : void 0,
                                  "aria-selected": P ? k : void 0,
                                  tabIndex: -1,
                                  children: ee.data.map((L) => L.collapsed ? /* @__PURE__ */ c(
                                    "div",
                                    {
                                      className: "wx-4VuBwK2D wx-cell wx-collapsed"
                                    },
                                    L.id
                                  ) : H?.id === o.id && H.column === L.id ? /* @__PURE__ */ c(Un, { row: o, column: L }, L.id) : /* @__PURE__ */ c(
                                    $n,
                                    {
                                      row: o,
                                      column: L,
                                      columnStyle: v,
                                      cellStyle: f,
                                      reorder: V,
                                      focusable: Te?.row === o.id && Te?.column === L.id
                                    },
                                    L.id
                                  ))
                                },
                                o.id
                              );
                            })
                          }
                        )
                      ]
                    }
                  ),
                  n && (y || []).length ? /* @__PURE__ */ c(
                    bt,
                    {
                      type: "footer",
                      contentWidth: Ce,
                      deltaLeft: ee.df,
                      columns: ee.footer,
                      columnStyle: v
                    }
                  ) : null
                ]
              }
            )
          }
        )
      }
    ),
    ae ? /* @__PURE__ */ c(
      Bn,
      {
        config: ae,
        rowStyle: s,
        columnStyle: v,
        cellStyle: f,
        header: e,
        footer: n,
        reorder: V
      }
    ) : null
  ] });
}
const qn = (t) => t.split("-").map((e) => e ? e.charAt(0).toUpperCase() + e.slice(1) : "").join(""), ao = Bt(function({
  data: e = [],
  columns: n = [],
  rowStyle: r = null,
  columnStyle: p = null,
  cellStyle: x = null,
  selectedRows: s,
  select: v = !0,
  multiselect: f = !1,
  header: u = !0,
  footer: g = !1,
  dynamic: a = null,
  overlay: b = null,
  reorder: D = !1,
  onReorder: C = null,
  autoRowHeight: d = !1,
  sizes: m,
  split: h,
  tree: y = !1,
  autoConfig: i = !1,
  init: R = null,
  responsive: T = null,
  sortMarks: N,
  undo: l = !1,
  hotkeys: H = null,
  filterValues: P,
  ...G
}, ae) {
  const oe = A();
  oe.current = G;
  const V = S(() => new un(xn), []), E = S(() => V.in, [V]), $ = A(null);
  $.current === null && ($.current = new nn((I, U) => {
    const J = "on" + qn(I);
    oe.current && oe.current[J] && oe.current[J](U);
  }), E.setNext($.current));
  const fe = S(
    () => ({
      getState: V.getState.bind(V),
      getReactiveState: V.getReactive.bind(V),
      getStores: () => ({ data: V }),
      exec: E.exec,
      setNext: (I) => ($.current = $.current.setNext(I), $.current),
      intercept: E.intercept.bind(E),
      on: E.on.bind(E),
      detach: E.detach.bind(E),
      getRow: V.getRow.bind(V),
      getRowIndex: V.getRowIndex.bind(V),
      getColumn: V.getColumn.bind(V)
    }),
    [V, E]
  ), [Se, _e] = K(0), [me, Fe] = K(0), [se, j] = K(null), [X, Xe] = K(null), Re = S(() => {
    if (i && !n.length && e.length) {
      const I = e[0], U = [];
      for (let J in I)
        if (J !== "id" && J[0] !== "$") {
          let ue = {
            id: J,
            header: J[0].toUpperCase() + J.slice(1)
          };
          typeof i == "object" && (ue = { ...ue, ...i }), U.push(ue);
        }
      return U;
    }
    return (X && X.columns) ?? n;
  }, [i, n, e, X]), de = S(
    () => (X && X.sizes) ?? m,
    [X, m]
  ), Me = W(
    (I) => {
      if (_e(I.width), Fe(I.height), T) {
        const J = Object.keys(T).map(Number).sort((ue, ee) => ue - ee).find((ue) => I.width <= ue) ?? null;
        J !== se && (Xe(T[J]), j(J));
      }
    },
    [T, se]
  ), te = Z(ot.theme), ie = A(0);
  return z(() => {
    if (!ie.current)
      R && R(fe);
    else {
      const I = V.getState();
      V.init({
        data: e,
        columns: Re,
        split: h || I.split,
        sizes: de || I.sizes,
        selectedRows: s || I.selectedRows,
        dynamic: a,
        tree: y,
        sortMarks: N || I.sortMarks,
        filterValues: P || I.filterValues,
        undo: l,
        reorder: D,
        _skin: te,
        _select: v
      });
    }
    ie.current++;
  }, [
    V,
    e,
    Re,
    h,
    de,
    s,
    a,
    y,
    N,
    P,
    l,
    D,
    te,
    v,
    R,
    fe
  ]), ie.current === 0 && V.init({
    data: e,
    columns: Re,
    split: h || { left: 0 },
    sizes: de || {},
    selectedRows: s || [],
    dynamic: a,
    tree: y,
    sortMarks: N || {},
    filterValues: P || {},
    undo: l,
    reorder: D,
    _skin: te,
    select: v
  }), Gt(
    ae,
    () => ({
      ...fe
    }),
    [fe]
  ), /* @__PURE__ */ c(ne.Provider, { value: fe, children: /* @__PURE__ */ c(tn, { words: rt, optional: !0, children: /* @__PURE__ */ c(
    Gn,
    {
      header: u,
      footer: g,
      overlay: b,
      rowStyle: r,
      columnStyle: p,
      cellStyle: x,
      onReorder: C,
      multiselect: f,
      autoRowHeight: d,
      clientWidth: Se,
      clientHeight: me,
      responsiveLevel: se,
      resize: Me,
      hotkeys: H
    }
  ) }) });
});
function Qn({ item: t }) {
  return /* @__PURE__ */ q(
    "div",
    {
      tabIndex: -1,
      role: "menuitem",
      "aria-label": t.hidden ? `Show ${t.text} column` : `Hide ${t.text} column`,
      children: [
        /* @__PURE__ */ c(
          "div",
          {
            className: "wx-v13lZxja wx-icon" + (t.hidden ? " wx-hidden" : ""),
            children: /* @__PURE__ */ c("i", { className: "wx-v13lZxja wxi-eye" })
          }
        ),
        /* @__PURE__ */ c("span", { children: t.text })
      ]
    }
  );
}
function uo({ columns: t = null, api: e, children: n }) {
  z(() => {
    vn("table-header", Qn);
  }, []);
  function r(f) {
    for (let u = f.header.length - 1; u >= 0; u--) {
      const g = f.header[u].text;
      if (g) return g;
    }
    return f.id;
  }
  function p(f) {
    const u = f.action;
    u && e.exec("hide-column", { id: u.id, mode: !u.hidden });
  }
  function x(f) {
    return f;
  }
  const s = we(e, "_columns"), v = S(() => {
    if (e) {
      const f = Array.isArray(s) ? s : [];
      return (t ? f.filter((g) => t[g.id]) : f).map((g) => {
        const a = r(g);
        return {
          id: g.id,
          text: a,
          type: "table-header",
          hidden: g.hidden
        };
      });
    } else
      return [];
  }, [e, t, s]);
  return /* @__PURE__ */ c(
    kt,
    {
      dataKey: "headerId",
      options: v,
      onClick: p,
      at: "point",
      resolver: x,
      children: typeof n == "function" ? n() : n
    }
  );
}
function fo(t) {
  const { content: e = null, api: n, children: r } = t, p = A(null), x = A(null), [s, v] = K(), [f, u] = K(), [g, a] = K();
  function b(i) {
    for (; i; ) {
      if (i.getAttribute) {
        const R = We(i, "data-row-id"), T = We(i, "data-col-id");
        if (R && n && T) {
          const N = n.getColumn(T);
          return { id: R, col: N, target: i };
        }
      }
      i = i.parentNode;
    }
    return { id: null, col: null, target: null };
  }
  z(() => {
    const i = x.current;
    if (i && s && g) {
      let R = i.getBoundingClientRect(), T = g.left, N = g.top;
      R.right >= s.right && (T = s.width - R.width - 5), R.bottom >= s.bottom && (N = g.top - (R.bottom - s.bottom + 2)), (T !== g.left || N !== g.top) && a({ ...g, left: T, top: N });
    }
  }, [s, g]);
  const D = A(), C = 300, d = (i) => {
    clearTimeout(D.current), D.current = setTimeout(() => {
      i();
    }, C);
  };
  function m(i) {
    const { id: R, target: T, col: N } = b(i.target);
    if (a(null), !R) {
      clearTimeout(D.current);
      return;
    }
    const l = i.clientX;
    d(() => {
      let H = "";
      if (R) {
        const E = h(R);
        u(E), H = y(N, E);
      }
      const P = T.getBoundingClientRect(), G = p.current;
      if (!G) return;
      const ae = G.getBoundingClientRect();
      v(ae);
      const oe = P.top + P.height - ae.top, V = l - ae.left;
      a({ top: oe, left: V, col: N, text: H });
    });
  }
  function h(i) {
    return n.getRow(i);
  }
  function y(i, R) {
    return R = R || f, typeof i.tooltip == "function" ? i.tooltip(R) : Pe(R, i) || "";
  }
  return /* @__PURE__ */ q("div", { className: "wx-RYccULtD wx-area", ref: p, onMouseMove: m, children: [
    g && g.col.tooltip !== !1 && (e || g.text) ? /* @__PURE__ */ c(
      "div",
      {
        className: "wx-RYccULtD tooltip",
        role: "alert",
        tabIndex: 0,
        ref: x,
        style: { top: g.top + "px", left: g.left + "px" },
        children: e ? /* @__PURE__ */ c(e, { data: f }) : g.text
      }
    ) : null,
    r
  ] });
}
function ho({
  api: t,
  options: e = [...fn],
  at: n = "point",
  resolver: r,
  dataKey: p,
  filter: x,
  css: s,
  children: v,
  handler: f,
  onClick: u
}) {
  const g = Z(ot.i18n), a = S(() => g ? g.getGroup("grid") : Tt(rt).getGroup("grid"), [g]), b = we(t, "selectedRows"), D = we(t, "data"), C = we(t, "reorder"), d = S(() => {
    const N = R(e), l = Rt(N);
    return m(l), l;
  }, [e, C, D]);
  function m(N) {
    N.forEach((l) => {
      l.text && (l.text = a(l.text)), l.subtext && (l.subtext = a(l.subtext)), l.data && (l.data = m(l.data));
    });
  }
  const h = S(() => {
    const N = [];
    return d.forEach((l) => {
      switch (l.id) {
        case "move-item:up":
        case "move-item:down":
        case "paste-row": {
          if (!l.isDisabled || !t) {
            N.push(l);
            return;
          }
          const H = l.isDisabled(
            l.id === "paste-row" ? t : b,
            D
          );
          N.push({
            ...l,
            disabled: H
          });
          break;
        }
        default: {
          N.push(l);
          break;
        }
      }
    }), N;
  }, [d, b, D, t]);
  function y(N) {
    return b.includes(N) || t.exec("select-row", { id: N }), N;
  }
  const i = (N) => {
    const l = N.action;
    l && Nt(t, l.id), u && u(N);
  };
  function R(N) {
    return C ? N : N.filter(({ id: l }) => !(l === "move-item:up" || l === "move-item:down"));
  }
  const T = r !== void 0 ? r : y;
  return /* @__PURE__ */ c(
    kt,
    {
      css: `wx-KKGLK8TH wx-table-menu ${s}`,
      at: n,
      dataKey: p,
      handler: f,
      options: h,
      resolver: T,
      filter: x,
      onClick: i,
      children: v
    }
  );
}
function po({
  api: t,
  items: e = [...hn],
  onClick: n,
  ...r
}) {
  const x = Z(ot.i18n)?.getGroup("grid") || Tt(rt).getGroup("grid"), s = we(t, "selectedRows"), v = we(t, "data"), f = we(t, "history"), u = we(t, "reorder"), g = we(t, "undo"), a = [
    "open-editor",
    "delete-row",
    "copy-row",
    "cut-row",
    "paste-row",
    "move-item:up",
    "move-item:down"
  ], b = ["undo", "redo"], D = S(() => {
    const m = y(e), h = Rt(m);
    return i(h), h;
    function y(R) {
      return g && u ? R : R.filter(({ id: T }) => !(!g && (T === "undo" || T === "redo") || !u && (T === "move-item:up" || T === "move-item:down")));
    }
    function i(R) {
      R.forEach((T) => {
        T.text && (T.text = x(T.text)), T.menuText && (T.menuText = x(T.menuText)), T.items && (T.items = i(T.items));
      });
    }
  }, [e, g, u, x]), C = S(() => {
    const m = [], h = s?.length;
    return D.forEach((y) => {
      const i = y.id;
      if (i === "add-row")
        m.push(y);
      else if (a.includes(i)) {
        if (!h) return;
        m.push({
          ...y,
          disabled: y.isDisabled && y.isDisabled(
            i === "paste-row" ? t : s,
            v
          )
        });
      } else b.includes(i) ? m.push({
        ...y,
        disabled: y.isDisabled(f)
      }) : m.push(y);
    }), m;
  }, [D, s, v, f, t, a, b]);
  return /* @__PURE__ */ c(bn, { items: C, onClick: (m) => {
    const h = m.item;
    h && Nt(t, h.id), n && n(m);
  }, ...r });
}
function go({ fonts: t = !0, children: e }) {
  return e ? /* @__PURE__ */ c(pt, { fonts: t, children: e }) : /* @__PURE__ */ c(pt, { fonts: t });
}
function wo({ fonts: t = !0, children: e }) {
  return e ? /* @__PURE__ */ c(gt, { fonts: t, children: e }) : /* @__PURE__ */ c(gt, { fonts: t });
}
function xo({ fonts: t = !0, children: e }) {
  return e ? /* @__PURE__ */ c(wt, { fonts: t, children: e }) : /* @__PURE__ */ c(wt, { fonts: t });
}
mn(yn);
export {
  ho as ContextMenu,
  ao as Grid,
  uo as HeaderMenu,
  go as Material,
  po as Toolbar,
  fo as Tooltip,
  wo as Willow,
  xo as WillowDark,
  vo as defaultMenuOptions,
  bo as defaultToolbarButtons,
  Co as getEditorConfig,
  co as registerInlineEditor
};
