// @svar-ui/react-menu/dist/index.es.js
import { jsxs as D, jsx as o, Fragment as R } from "react/jsx-runtime";
import { useRef as v, useCallback as b, useMemo as O, useState as p, useEffect as z, Fragment as j, forwardRef as E, useImperativeHandle as H } from "react";
import { uid as B, setID as T, calculatePosition as W, clickOutside as J, locateID as _ } from "@svar-ui/lib-dom";
import { Portal as K } from "@svar-ui/react-core";
function S(l, t) {
  return l.map((e) => {
    const n = t(e);
    return e.data && e.data.length && (n.data = S(e.data, t)), n;
  });
}
function q(l, t) {
  const e = [];
  return l.forEach((n) => {
    if (n.data) {
      const a = q(n.data, t);
      a.length && e.push({ ...n, data: a });
    } else
      t(n) && e.push(n);
  }), e;
}
function G(l) {
  return S(l, (t) => {
    const e = { ...t, id: t.id || B() };
    return e.type && (e.comp = e.type), e;
  });
}
const U = {};
function ee(l) {
  return U[l] || l;
}
function ce(l, t) {
  U[l] = t;
}
function te({ onClick: l, onShow: t, option: e }) {
  const n = v(null), a = b(() => {
    t(e.data ? e.id : !1, n.current);
  }, [t, e]), x = b((c) => {
    if (e.data) {
      c.stopPropagation(), t(e.id, n.current);
      return;
    }
    l(c);
  }, [l, t, e]), d = O(() => e && e.comp ? ee(e.comp) : null, [e]);
  return /* @__PURE__ */ D(
    "div",
    {
      ref: n,
      className: `wx-cDCz9rZQ wx-option ${e.css || ""} ${e.disabled ? "wx-disabled" : ""}`,
      "data-id": T(e.id),
      onMouseEnter: a,
      onClick: x,
      children: [
        e.icon ? /* @__PURE__ */ o("i", { className: `wx-cDCz9rZQ wx-icon ${e.icon}` }) : null,
        e.comp ? d ? /* @__PURE__ */ o(d, { item: e, option: e }) : null : /* @__PURE__ */ D("span", { className: "wx-cDCz9rZQ wx-value", children: [
          " ",
          e.text,
          " "
        ] }),
        e.subtext ? /* @__PURE__ */ o("span", { className: "wx-cDCz9rZQ wx-subtext", children: e.subtext }) : null,
        e.data ? /* @__PURE__ */ o("i", { className: "wx-cDCz9rZQ wx-sub-icon wxi-angle-right" }) : null
      ]
    }
  );
}
function X({
  options: l,
  left: t = 0,
  top: e = 0,
  at: n = "bottom",
  parent: a = null,
  mount: x = null,
  context: d = null,
  css: c = "",
  onClick: u
}) {
  const [f, w] = p(-1e4), [h, C] = p(-1e4), [g, s] = p(20), [r, M] = p(), I = v(null), P = v(u), [A, L] = p(!1), [Q, $] = p(null);
  z(() => {
    P.current = u;
  }, [u]);
  const m = b(() => {
    const i = W(I.current, a, n, t, e);
    i && (w(i.x), C(i.y), s(i.z), M(i.width));
  }, [a, n, t, e]);
  z(() => {
    x && x(m);
  }, []);
  const Z = b(() => {
    L(!1);
  }, []), k = b(() => {
    P.current && P.current({ action: null, option: null });
  }, []), N = b((i, y) => {
    L(i), $(y);
  }, []), Y = O(() => G(l), [l]);
  return z(() => {
    m();
  }, [a, m]), z(() => {
    if (I.current)
      return J(I.current, { callback: k, modal: !0 }).destroy;
  }, [k]), /* @__PURE__ */ o(
    "div",
    {
      ref: I,
      "data-wx-menu": "true",
      className: `wx-XMmAGqVx wx-menu ${c}`,
      style: {
        position: "absolute",
        top: h + "px",
        left: f + "px",
        width: r,
        zIndex: g
      },
      onMouseLeave: Z,
      children: Y.map((i) => /* @__PURE__ */ D(j, { children: [
        i.comp === "separator" ? /* @__PURE__ */ o("div", { className: "wx-XMmAGqVx wx-separator" }) : /* @__PURE__ */ o(
          te,
          {
            option: i,
            onShow: N,
            onClick: (y) => {
              if (!i.data && !y.defaultPrevented) {
                const F = { context: d, action: i, option: i, event: y };
                i.handler && i.handler(F), u && u(F), y.stopPropagation();
              }
            }
          }
        ),
        i.data && A === i.id ? /* @__PURE__ */ o(
          X,
          {
            css: c,
            options: i.data,
            at: "right-overlap",
            parent: Q,
            context: d,
            onClick: u
          }
        ) : null
      ] }, i.id))
    }
  );
}
const V = E(function(t, e) {
  const {
    options: n,
    at: a = "bottom",
    resolver: x = null,
    dataKey: d = "contextId",
    filter: c = null,
    css: u = "",
    children: f,
    onClick: w
  } = t, [h, C] = p(null), [g, s] = p(null), [r, M] = p(0), [I, P] = p(0), A = O(
    () => `data-${d.replace(/([a-z0-9])([A-Z])/g, "$1-$2").toLowerCase()}`,
    [d]
  ), L = O(() => h !== null && c ? q(n, (m) => c(m, h)) : n, [h, c, n]), Q = b(
    (m) => {
      s(null), w && w(m);
    },
    [w]
  ), $ = b(
    (m, Z) => {
      if (!m) {
        s(null);
        return;
      }
      if (m.defaultPrevented) return;
      const k = m.target;
      if (k && k.dataset && k.dataset.menuIgnore) return;
      M(m.clientX + 1), P(m.clientY + 1);
      let N = typeof Z < "u" ? Z : _(k, A);
      x && (N = x(N, m), !N) || (C(N), s(k), m.preventDefault());
    },
    [A, x]
  );
  return H(e, () => ({ show: $ }), [$]), /* @__PURE__ */ D(R, { children: [
    f ? /* @__PURE__ */ o("span", { onClick: $, "data-menu-ignore": "true", children: typeof f == "function" ? f() : f }) : null,
    g ? /* @__PURE__ */ o(K, { children: /* @__PURE__ */ o(
      X,
      {
        css: u,
        at: a,
        top: I,
        left: r,
        parent: g,
        context: h,
        onClick: Q,
        options: L
      },
      g
    ) }) : null
  ] });
});
function ae(l) {
  const { css: t = "", menuCss: e = "", options: n, onClick: a } = l, x = O(() => G(n), [n]), [d, c] = p(!1), [u, f] = p([]), w = v(null);
  function h(s) {
    c(null), a && a(s);
  }
  function C(s, r, M) {
    r.data && r.data.length ? d && M ? c(null) : (f(r.data), c(r.id), w.current.show(s, r)) : (w.current.show(null), M ? (a && a({ action: r, option: r }), c(null)) : c(-1));
  }
  function g(s, r) {
    d && C(s, r);
  }
  return /* @__PURE__ */ D(R, { children: [
    /* @__PURE__ */ o("div", { className: `wx-UfhPCLL4 wx-menubar ${t}`, children: x.map((s) => /* @__PURE__ */ o(
      "button",
      {
        className: `wx-UfhPCLL4 wx-option ${d === s.id ? "wx-active" : ""} ${s.disabled ? "wx-disabled" : ""}`,
        onMouseEnter: (r) => g(r, s),
        onClick: (r) => C(r, s, !0),
        children: s.text
      },
      s.id
    )) }),
    /* @__PURE__ */ o(
      V,
      {
        css: e,
        onClick: h,
        options: u,
        ref: w
      }
    )
  ] });
}
const ue = E(function(t, e) {
  const { options: n, at: a = "bottom", css: x = "", children: d, onClick: c } = t, [u, f] = p(null);
  function w(r) {
    f(null), c && c(r);
  }
  const h = b((r) => {
    f(r.target), r.preventDefault();
  }, []);
  H(e, () => ({ show: h }), [h]);
  function C(r) {
    let M = r.target;
    for (; !M.dataset.menuIgnore; )
      f(M), M = M.parentNode;
  }
  const g = v(0), s = v(u);
  return z(() => {
    s.current !== u && (g.current += 1, s.current = u);
  }, [u]), /* @__PURE__ */ D(R, { children: [
    /* @__PURE__ */ o("span", { onClick: C, "data-menu-ignore": "true", children: d }),
    u ? /* @__PURE__ */ o(K, { children: /* @__PURE__ */ o(
      X,
      {
        css: x,
        at: a,
        parent: u,
        options: n,
        onClick: w
      },
      g.current
    ) }) : null
  ] });
}), ie = E(function(t, e) {
  const {
    options: n,
    at: a = "bottom",
    resolver: x = null,
    dataKey: d = "contextId",
    filter: c = null,
    css: u = "",
    children: f,
    onClick: w
  } = t, h = v(null), C = b((g, s) => {
    h.current.show(g, s);
  }, []);
  return H(
    e,
    () => ({
      show: C
    }),
    [C]
  ), /* @__PURE__ */ D(R, { children: [
    f ? /* @__PURE__ */ o("span", { onContextMenu: C, "data-menu-ignore": "true", children: f }) : null,
    /* @__PURE__ */ o(
      V,
      {
        css: u,
        at: a,
        options: n,
        resolver: x,
        dataKey: d,
        filter: c,
        ref: h,
        onClick: w
      }
    )
  ] });
});
export {
  V as ActionMenu,
  ie as ContextMenu,
  ue as DropDownMenu,
  X as Menu,
  ae as MenuBar,
  ce as registerMenuItem
};
