// @svar-ui/react-gantt/dist/index.es.js
import { jsxs as me, jsx as i, Fragment as Oe } from "react/jsx-runtime";
import { createContext as bt, useMemo as v, useState as se, useContext as be, useCallback as b, useRef as X, useEffect as ee, Fragment as vt, useLayoutEffect as Tt, forwardRef as ht, useImperativeHandle as mt } from "react";
import { context as Le, Button as nt, Field as $t, Text as Ct, Combo as Rt, DatePicker as Nt, TimePicker as Mt, Locale as Lt, RichSelect as Et, TwoState as St, Slider as Ht, Counter as At, Checkbox as Dt, Material as st, Willow as rt, WillowDark as ot } from "@svar-ui/react-core";
import { locate as Se, getID as De, locateID as Ae, setID as Fe, dateToString as Ye, locale as je } from "@svar-ui/lib-dom";
import { en as qe } from "@svar-ui/gantt-locales";
import { en as et } from "@svar-ui/core-locales";
import { EventBusRouter as It } from "@svar-ui/lib-state";
import { prepareEditTask as xt, grid as zt, extendDragOptions as Gt, isSegmentMoveAllowed as Wt, DataStore as _t, normalizeLinks as Pt, normalizeZoom as Vt, defaultColumns as Ft, parseTaskDates as lt, defaultTaskTypes as Ot, getToolbarButtons as it, isHandledAction as gt, handleAction as pt, getMenuOptions as ct, getEditorItems as Kt } from "@svar-ui/gantt-store";
import { defaultColumns as Zn, defaultEditorItems as Jn, defaultMenuOptions as es, defaultTaskTypes as ts, defaultToolbarButtons as ns, getEditorItems as ss, getMenuOptions as rs, getToolbarButtons as os, registerScaleUnit as ls } from "@svar-ui/gantt-store";
import { useWritableProp as Ze, useStore as R, useStoreWithCounter as Je, writable as Bt, useStoreLater as He } from "@svar-ui/lib-react";
import { hotkeys as wt } from "@svar-ui/grid-store";
import { Grid as Xt, HeaderMenu as Yt } from "@svar-ui/react-grid";
import { flushSync as jt } from "react-dom";
import { Toolbar as at } from "@svar-ui/react-toolbar";
import { ContextMenu as qt } from "@svar-ui/react-menu";
import { Editor as Qt, registerEditorItem as ze } from "@svar-ui/react-editor";
import { registerEditorItem as cs } from "@svar-ui/react-editor";
const Ie = bt(null);
function Ut(t, s, o) {
  const n = t.getBoundingClientRect(), l = s.querySelector(".wx-body").getBoundingClientRect();
  return {
    top: n.top - l.top,
    left: n.left - l.left,
    dt: n.bottom - o.clientY,
    db: o.clientY - n.top
  };
}
function ut(t) {
  return t && De(t, "data-context-id");
}
const dt = 5;
function Zt(t, s) {
  let o, n, l, N, f, d, u, E, m;
  function V(h) {
    N = h.clientX, f = h.clientY, d = {
      ...Ut(o, t, h),
      y: s.getTask(l).$y
    }, document.body.style.userSelect = "none";
  }
  function T(h) {
    o = Se(h), ut(o) && (l = De(o), m = setTimeout(() => {
      E = !0, s && s.touchStart && s.touchStart(), V(h.touches[0]);
    }, 500), t.addEventListener("touchmove", y), t.addEventListener("contextmenu", c), window.addEventListener("touchend", _));
  }
  function c(h) {
    if (E || m)
      return h.preventDefault(), !1;
  }
  function D(h) {
    h.which === 1 && (o = Se(h), ut(o) && (l = De(o), t.addEventListener("mousemove", B), window.addEventListener("mouseup", M), V(h)));
  }
  function H(h) {
    t.removeEventListener("mousemove", B), t.removeEventListener("touchmove", y), document.body.removeEventListener("mouseup", M), document.body.removeEventListener("touchend", _), document.body.style.userSelect = "", h && (t.removeEventListener("mousedown", D), t.removeEventListener("touchstart", T));
  }
  function S(h) {
    const F = h.clientX - N, q = h.clientY - f;
    if (!n) {
      if (Math.abs(F) < dt && Math.abs(q) < dt || s && s.start && s.start({ id: l, e: h }) === !1)
        return;
      n = o.cloneNode(!0), n.style.pointerEvents = "none", n.classList.add("wx-reorder-task"), n.style.position = "absolute", n.style.left = d.left + "px", n.style.top = d.top + "px", o.style.visibility = "hidden", o.parentNode.insertBefore(n, o);
    }
    if (n) {
      const k = Math.round(Math.max(0, d.top + q));
      if (s && s.move && s.move({ id: l, top: k, detail: u }) === !1)
        return;
      const ne = s.getTask(l), Y = ne.$y;
      if (!d.start && d.y === Y) return G();
      d.start = !0, d.y = ne.$y - 4, n.style.top = k + "px";
      const Q = document.elementFromPoint(
        h.clientX,
        h.clientY
      ), W = Se(Q);
      if (W && W !== o) {
        const O = De(W), $ = W.getBoundingClientRect(), L = $.top + $.height / 2, A = h.clientY + d.db > L && W.nextElementSibling !== o, P = h.clientY - d.dt < L && W.previousElementSibling !== o;
        u?.after === O || u?.before === O ? u = null : A ? u = { id: l, after: O } : P && (u = { id: l, before: O });
      }
    }
  }
  function B(h) {
    S(h);
  }
  function y(h) {
    E ? (h.preventDefault(), S(h.touches[0])) : m && (clearTimeout(m), m = null);
  }
  function _() {
    E = null, m && (clearTimeout(m), m = null), G();
  }
  function M() {
    G();
  }
  function G() {
    o && (o.style.visibility = ""), n && (n.parentNode.removeChild(n), s && s.end && s.end({ id: l, top: d.top })), l = o = n = d = u = null, H();
  }
  return t.style.position !== "absolute" && (t.style.position = "relative"), t.addEventListener("mousedown", D), t.addEventListener("touchstart", T), {
    destroy() {
      H(!0);
    }
  };
}
function Jt({ row: t, column: s }) {
  function o(l, N) {
    return {
      justifyContent: N.align,
      paddingLeft: `${(l.$level - 1) * 20}px`
    };
  }
  const n = s && s._cell;
  return /* @__PURE__ */ me("div", { className: "wx-pqc08MHU wx-content", style: o(t, s), children: [
    !t.$empty && (t.data || t.lazy) ? /* @__PURE__ */ i(
      "i",
      {
        className: `wx-pqc08MHU wx-toggle-icon wxi-menu-${t.open ? "down" : "right"}`,
        "data-action": "open-task"
      }
    ) : /* @__PURE__ */ i("i", { className: "wx-pqc08MHU wx-toggle-placeholder" }),
    /* @__PURE__ */ i("div", { className: "wx-pqc08MHU wx-text", children: n ? /* @__PURE__ */ i(n, { row: t, column: s }) : t.text })
  ] });
}
function ft({ column: t, cell: s }) {
  const o = v(() => t.id, [t?.id]);
  return s || t.id === "add-task" ? /* @__PURE__ */ i("div", { style: { textAlign: t.align }, children: /* @__PURE__ */ i(
    "i",
    {
      className: "wx-9DAESAHW wx-action-icon wxi-plus",
      "data-action": o
    }
  ) }) : null;
}
function en(t) {
  const {
    readonly: s,
    compactMode: o,
    width: n = 0,
    display: l = "all",
    columnWidth: N = 0,
    onTableAPIChange: f
  } = t, [d, u] = Ze(N), [E, m] = se(), V = be(Le.i18n), T = v(() => V.getGroup("gantt"), [V]), c = be(Ie), D = R(c, "scrollTop"), H = R(c, "cellHeight"), S = R(c, "focusTask"), B = R(c, "_selected"), y = R(c, "area"), _ = R(c, "_tasks"), M = R(c, "_scales"), G = R(c, "_headerLength"), h = R(c, "columns"), F = R(c, "_sort"), q = R(c, "calendar"), k = R(c, "durationUnit"), ne = R(c, "splitTasks"), Y = R(c, "filterValues"), [Q, W] = se(null), O = v(() => !_ || !y ? [] : _.slice(y.start, y.end), [_, y]), $ = b(
    (e, r) => {
      if (r === "add-task")
        c.exec(r, {
          target: e,
          task: { text: T("New Task") },
          mode: "child",
          show: !0,
          focus: e ? "grid" : null
        });
      else if (r === "open-task") {
        const a = O.find((x) => x.id === e);
        (a?.data || a?.lazy) && c.exec(r, { id: e, mode: !a.open });
      }
    },
    [O]
  ), L = b(
    (e) => {
      const r = Ae(e), a = e.target.dataset.action;
      a && e.preventDefault(), r ? a === "add-task" || a === "open-task" ? $(r, a) : c.exec("select-task", {
        id: r,
        toggle: e.ctrlKey || e.metaKey,
        range: e.shiftKey,
        show: "xy",
        focus: "grid"
      }) : a === "add-task" && $(null, a);
    },
    [c, $]
  ), A = X(null), P = X(null), [K, ue] = se(0), [p, C] = se(0), [w, Z] = se(!1);
  ee(() => {
    const e = P.current;
    if (!e || typeof ResizeObserver > "u") return;
    const r = () => {
      ue(e.clientWidth), C(e.clientHeight);
    };
    r();
    const a = new ResizeObserver(r);
    return a.observe(e), () => a.disconnect();
  }, []);
  const te = v(() => (Q && !O.find((r) => r.id === Q.id) ? [...O, Q] : O).map((r) => ({ ...r })), [O, Q]), ie = X(te);
  ee(() => {
    ie.current = te;
  }, [te]);
  const re = X(null), ce = b(
    (e) => {
      const r = e.id, { before: a, after: x } = e, I = e.onMove;
      let z = a || x, U = a ? "before" : "after";
      if (I) {
        if (U === "after") {
          const oe = ie.current.findIndex((Me) => Me.id === r), ye = ie.current.findIndex(
            (Me) => Me.id === z
          ), Ne = ie.current[ye];
          oe - ye === 1 ? U = "before" : Ne && Ne.data && Ne.open && (U = "before", z = Ne.data[0].id);
        }
        re.current = { id: r, [U]: z };
      } else re.current = null;
      c.exec("move-task", {
        id: r,
        mode: U,
        target: z,
        inProgress: I
      });
    },
    [c]
  ), de = v(() => {
    let e = (h || []).map((x) => {
      x = { ...x };
      const I = [...x.header];
      return I.forEach((z) => {
        z.text && (z.text = T(z.text));
      }), x.header = I, x;
    });
    const r = e.findIndex((x) => x.id === "text"), a = e.findIndex((x) => x.id === "add-task");
    if (r !== -1 && (e[r].cell && (e[r]._cell = e[r].cell), e[r].cell = Jt), a !== -1) {
      e[a].cell = e[a].cell || ft;
      const x = e[a].header[0];
      if (e[a].header[0].cell = x.cell || ft, s)
        e.splice(a, 1);
      else if (o) {
        const [I] = e.splice(a, 1);
        e.unshift(I);
      }
    }
    return e.length > 0 && (e[e.length - 1].resize = !1), e;
  }, [h, T, s, o]), ae = v(() => y?.from ?? 0, [y]), pe = v(() => M?.height ?? 0, [M]), xe = v(() => !o && l !== "grid" ? (d ?? 0) > (n ?? 0) : (d ?? 0) > (K ?? 0), [o, l, d, n, K]), Te = v(() => {
    const e = {
      minHeight: `${p + (H ?? 0) * 4}px`
    };
    return xe && l === "all" || l === "grid" && xe ? e.width = d : l === "grid" && (e.width = "100%"), e;
  }, [xe, l, d, p, H]), $e = v(() => l === "all" ? `${n}px` : l === "grid" ? "calc(100% - 4px)" : de.find((e) => e.id === "add-task") ? "50px" : "0", [l, n, de]), ge = v(() => {
    if (te && F?.length) {
      const e = {};
      return F.forEach(({ key: r, order: a }, x) => {
        e[r] = {
          order: a,
          ...F.length > 1 && { index: x }
        };
      }), e;
    }
    return {};
  }, [te, F]), Ce = v(() => ge ? { ...Y } : Y, [ge, Y]), Ee = b(() => de.some((e) => e.flexgrow && !e.hidden), [de]), fe = v(() => Ee(), [Ee, w]), we = v(() => {
    let e = l === "chart" ? de.filter((a) => a.id === "add-task") : de;
    const r = l === "all" ? n : K;
    if (!fe) {
      let a = d, x = !1;
      if (de.some((I) => I.$width)) {
        let I = 0;
        a = de.reduce((z, U) => (U.hidden || (I += U.width, z += U.$width || U.width), z), 0), I > a && a >= r && (x = !0);
      }
      if (x || a < r) {
        let I = 1;
        return x || (I = (r - 50) / (a - 50 || 1)), e.map((z) => (z.id !== "add-task" && !z.hidden && (z.$width || (z.$width = z.width), z.width = z.$width * I), z));
      }
    }
    return e;
  }, [l, de, fe, d, n, K]), g = b(
    (e) => {
      if (!Ee()) {
        const r = we.reduce((a, x) => (e && x.$width && (x.$width = x.width), a + (x.hidden ? 0 : x.width)), 0);
        r !== d && u(r);
      }
      Z(!0), Z(!1);
    },
    [Ee, we, d, u]
  ), J = b(() => {
    de.filter((r) => r.flexgrow && !r.hidden).length === 1 && de.forEach((r) => {
      r.$width && !r.flexgrow && !r.hidden && (r.width = r.$width);
    });
  }, [de]), le = b(
    (e) => {
      if (!s) {
        const r = Ae(e), a = Ae(e, "data-col-id");
        !(a && de.find((I) => I.id === a))?.editor && r && c.exec("show-editor", { id: r });
      }
    },
    [c, s, de]
  ), j = v(
    () => Array.isArray(B) ? B.map((e) => e.id) : [],
    [B]
  ), he = b(() => {
    if (A.current && te !== null) {
      const e = A.current.querySelector(".wx-body");
      e && (e.style.top = -((D ?? 0) - (ae ?? 0)) + "px");
    }
  }, [te, D, ae]);
  ee(() => {
    A.current && he();
  }, [D, ae, he]), ee(() => {
    const e = A.current;
    if (!e) return;
    const r = e.querySelector(".wx-table-box .wx-body");
    if (!r || typeof ResizeObserver > "u") return;
    const a = new ResizeObserver(() => {
      he();
    });
    return a.observe(r), () => {
      a.disconnect();
    };
  }, [we, Te, l, $e, te, he]);
  const Re = X(!1);
  ee(() => {
    if (!S || !E) return;
    const { id: e, column: r } = S;
    r && (Re.current || (Re.current = !0, requestAnimationFrame(() => {
      const a = E.getState().focusCell;
      E.exec("focus-cell", {
        row: e,
        column: a?.column || de[0]?.id
      }), Re.current = !1;
    })));
  }, [S, E]);
  const _e = b(
    ({ id: e }) => {
      if (s) return !1;
      c.getTask(e).open && c.exec("open-task", { id: e, mode: !1 });
      const r = c.getState()._tasks.find((a) => a.id === e);
      if (W(r || null), !r) return !1;
    },
    [c, s]
  ), Ge = b(
    ({ id: e, top: r }) => {
      re.current ? ce({ ...re.current, onMove: !1 }) : c.exec("drag-task", {
        id: e,
        top: r + (ae ?? 0),
        inProgress: !1
      }), W(null);
    },
    [c, ce, ae]
  ), Pe = b(
    ({ id: e, top: r, detail: a }) => {
      a && ce({ ...a, onMove: !0 }), c.exec("drag-task", {
        id: e,
        top: r + (ae ?? 0),
        inProgress: !0
      });
    },
    [c, ce, ae]
  );
  ee(() => {
    const e = A.current;
    return e ? Zt(e, {
      start: _e,
      end: Ge,
      move: Pe,
      getTask: c.getTask
    }).destroy : void 0;
  }, [c, _e, Ge, Pe]);
  const Ve = b(
    (e) => {
      const { key: r, isInput: a } = e;
      if (!a && (r === "arrowup" || r === "arrowdown"))
        return e.eventSource = "grid", c.exec("hotkey", e), !1;
      if (r === "enter") {
        const x = E?.getState().focusCell;
        if (x) {
          const { row: I, column: z } = x;
          z === "add-task" ? $(I, "add-task") : z === "text" && $(I, "open-task");
        }
      }
    },
    [c, $, E]
  ), ke = X(null), We = () => {
    ke.current = {
      setTableAPI: m,
      handleHotkey: Ve,
      sortVal: F,
      api: c,
      adjustColumns: J,
      setColumnWidth: g,
      tasks: O,
      calendarVal: q,
      durationUnitVal: k,
      splitTasksVal: ne,
      onTableAPIChange: f
    };
  };
  We(), ee(() => {
    We();
  }, [
    m,
    Ve,
    F,
    c,
    J,
    g,
    O,
    q,
    k,
    ne,
    f
  ]);
  const Ke = b((e) => {
    m(e), e.intercept("hotkey", (r) => ke.current.handleHotkey(r)), e.intercept("scroll", () => !1), e.intercept("select-row", () => !1), e.intercept("sort-rows", (r) => {
      const a = ke.current.sortVal, { key: x, add: I } = r, z = a ? a.find((oe) => oe.key === x) : null;
      let U = "asc";
      return z && (U = !z || z.order === "asc" ? "desc" : "asc"), c.exec("sort-tasks", {
        key: x,
        order: U,
        add: I
      }), !1;
    }), e.intercept("filter-rows", (r) => {
      const { key: a, value: x } = r;
      return c.exec("filter-tasks", {
        key: a,
        value: x,
        open: !0
      }), !1;
    }), e.on("resize-column", () => {
      ke.current.setColumnWidth(!0);
    }), e.on("hide-column", (r) => {
      r.mode || ke.current.adjustColumns(), ke.current.setColumnWidth();
    }), e.intercept("update-cell", (r) => {
      const { id: a, column: x, value: I } = r, z = ke.current.tasks.find((U) => U.id === a);
      if (z) {
        const U = { ...z };
        let oe = I;
        oe && !isNaN(oe) && !(oe instanceof Date) && (oe *= 1), U[x] = oe, xt(
          U,
          {
            calendar: ke.current.calendarVal,
            durationUnit: ke.current.durationUnitVal,
            splitTasks: ke.current.splitTasksVal
          },
          x
        ), c.exec("update-task", {
          id: a,
          task: U
        });
      }
      return !1;
    }), f && f(e);
  }, []);
  return /* @__PURE__ */ i(
    "div",
    {
      className: "wx-rHj6070p wx-table-container",
      style: { flex: `0 0 ${$e}` },
      ref: P,
      children: /* @__PURE__ */ i(
        "div",
        {
          ref: A,
          style: Te,
          className: "wx-rHj6070p wx-table",
          onClick: L,
          onDoubleClick: le,
          children: /* @__PURE__ */ i(
            Xt,
            {
              init: Ke,
              sizes: {
                rowHeight: H,
                headerHeight: (pe ?? 0) / (G ?? 1)
              },
              rowStyle: (e) => e.$reorder ? "wx-rHj6070p wx-reorder-task" : "wx-rHj6070p",
              columnStyle: (e) => `wx-rHj6070p wx-text-${e.align}${e.id === "add-task" ? " wx-action" : ""}`,
              data: te,
              columns: we,
              selectedRows: [...j],
              sortMarks: ge,
              filterValues: Ce
            }
          )
        }
      )
    }
  );
}
function tn({ borders: t = "" }) {
  const s = be(Ie), o = R(s, "cellWidth"), n = R(s, "cellHeight"), l = X(null), [N, f] = se("#e4e4e4");
  ee(() => {
    if (typeof getComputedStyle < "u" && l.current) {
      const u = getComputedStyle(l.current).getPropertyValue(
        "--wx-gantt-border"
      );
      f(u ? u.substring(u.indexOf("#")) : "#1d1e261a");
    }
  }, []);
  const d = {
    width: "100%",
    height: "100%",
    background: o != null && n != null ? `url(${zt(o, n, N, t)})` : void 0,
    position: "absolute"
  };
  return /* @__PURE__ */ i("div", { ref: l, style: d });
}
function nn({ onSelectLink: t, selectedLink: s, readonly: o }) {
  const n = be(Ie), l = R(n, "_links"), N = R(n, "criticalPath"), f = X(null), d = b(
    (u) => {
      const E = u?.target?.classList;
      !E?.contains("wx-line") && !E?.contains("wx-delete-button") && t(null);
    },
    [t]
  );
  return ee(() => {
    if (!o && s && f.current) {
      const u = (E) => {
        f.current && !f.current.contains(E.target) && d(E);
      };
      return document.addEventListener("click", u), () => {
        document.removeEventListener("click", u);
      };
    }
  }, [o, s, d]), /* @__PURE__ */ me("svg", { className: "wx-dkx3NwEn wx-links", children: [
    (l || []).map((u) => {
      const E = "wx-dkx3NwEn wx-line" + (N && u.critical ? " wx-critical" : "") + (o ? "" : " wx-line-selectable");
      return /* @__PURE__ */ i(
        "polyline",
        {
          className: E,
          points: u.$p,
          onClick: () => !o && t(u.id),
          "data-link-id": Fe(u.id)
        },
        u.id
      );
    }),
    !o && s && /* @__PURE__ */ i(
      "polyline",
      {
        ref: f,
        className: "wx-dkx3NwEn wx-line wx-line-selected wx-line-selectable wx-delete-link",
        points: s.$p
      }
    )
  ] });
}
function sn(t) {
  const { task: s, type: o } = t;
  function n(N) {
    const f = s.segments[N];
    return {
      left: `${f.$x}px`,
      top: "0px",
      width: `${f.$w}px`,
      height: "100%"
    };
  }
  function l(N) {
    if (!s.progress) return 0;
    const f = s.duration * s.progress / 100, d = s.segments;
    let u = 0, E = 0, m = null;
    do {
      const V = d[E];
      E === N && (u > f ? m = 0 : m = Math.min((f - u) / V.duration, 1) * 100), u += V.duration, E++;
    } while (m === null && E < d.length);
    return m || 0;
  }
  return /* @__PURE__ */ i("div", { className: "wx-segments wx-GKbcLEGA", children: s.segments.map((N, f) => /* @__PURE__ */ me(
    "div",
    {
      className: `wx-segment wx-bar wx-${o} wx-GKbcLEGA`,
      "data-segment": f,
      style: n(f),
      children: [
        s.progress ? /* @__PURE__ */ i("div", { className: "wx-progress-wrapper", children: /* @__PURE__ */ i(
          "div",
          {
            className: "wx-progress-percent wx-GKbcLEGA",
            style: { width: `${l(f)}%` }
          }
        ) }) : null,
        /* @__PURE__ */ i("div", { className: "wx-content", children: N.text || "" })
      ]
    },
    f
  )) });
}
function rn(t) {
  const { rollup: s, parent: o } = t;
  return /* @__PURE__ */ i(
    "div",
    {
      "data-tooltip-id": s.id,
      className: `wx-GKbcLEGA wx-rollup wx-${s.type}-rollup`,
      style: {
        left: `${s.$x_rollup}px`,
        top: `${o.$y + o.$h + s.$y_rollup_relative}px`,
        width: `${s.$w_rollup}px`,
        height: `${s.$h_rollup}px`
      }
    }
  );
}
function on(t) {
  const { readonly: s, taskTemplate: o } = t, n = be(Ie), [l, N] = Je(n, "_tasks"), [f, d] = Je(n, "_links"), u = R(n, "area"), E = R(n, "_scales"), m = R(n, "taskTypes"), V = R(n, "baselines"), T = R(n, "_selected"), c = R(n, "rollups"), D = R(n, "_rollups"), H = R(n, "focusTask"), S = R(n, "criticalPath"), B = R(n, "tasks"), y = R(n, "schedule"), _ = R(n, "splitTasks"), M = R(n, "summary"), G = R(n, "slack"), h = v(() => {
    if (!u || !Array.isArray(l)) return [];
    const e = u.start ?? 0, r = u.end ?? 0;
    return l.slice(e, r).map((a) => ({ ...a }));
  }, [N, u]), F = v(
    () => E.lengthUnitWidth,
    [E]
  ), q = X(!1), [k, ne] = se(void 0), [Y, Q] = se(null), W = X(null), [O, $] = se(null), L = v(() => O && {
    ...f.find((e) => e.id === O)
  }, [O, d]), [A, P] = se(void 0), K = X(null), [ue, p] = se(0), C = X(null), w = v(() => {
    const e = C.current;
    return !!(T.length && e && e.contains(document.activeElement));
  }, [T, C.current]), Z = v(() => w && T[T.length - 1]?.id, [w, T]);
  ee(() => {
    if (H && H.column === !1) {
      const { id: e } = H, r = C.current?.querySelector(
        `.wx-bar[data-id='${Fe(e)}']`
      );
      r && r.focus({ preventScroll: !0 });
    }
  }, [H]), ee(() => {
    const e = C.current;
    if (e && (p(e.offsetWidth || 0), typeof ResizeObserver < "u")) {
      const r = new ResizeObserver((a) => {
        a[0] && p(a[0].contentRect.width);
      });
      return r.observe(e), () => r.disconnect();
    }
  }, [C.current]);
  const te = b(() => {
    document.body.style.userSelect = "none";
  }, []), ie = b(() => {
    document.body.style.userSelect = "";
  }, []), re = b(
    (e, r, a) => {
      if (r.target.classList.contains("wx-line") || (a || (a = n.getTask(De(e))), a.type === "milestone" || a.type === "summary")) return "";
      const x = Se(r, "data-segment");
      x && (e = x);
      const { left: I, width: z } = e.getBoundingClientRect(), U = (r.clientX - I) / z;
      let oe = 0.2 / (z > 200 ? z / 200 : 1);
      return U < oe ? "start" : U > 1 - oe ? "end" : "";
    },
    [n]
  ), ce = b(
    (e, r) => {
      const { clientX: a } = r, x = De(e), I = n.getTask(x), z = r.target.classList;
      if (!r.target.closest(".wx-delete-button") && !s) {
        if (z.contains("wx-progress-marker")) {
          const { progress: U } = n.getTask(x);
          W.current = {
            id: x,
            x: a,
            progress: U,
            dx: 0,
            node: e,
            marker: r.target
          }, r.target.classList.add("wx-progress-in-drag");
        } else {
          const U = re(e, r, I) || "move", oe = {
            id: x,
            mode: U,
            x: a,
            dx: 0,
            l: I.$x,
            w: I.$w
          };
          if (_ && I.segments?.length) {
            const ye = Se(r, "data-segment");
            ye && (oe.segmentIndex = ye.dataset.segment * 1, Gt(I, oe));
          }
          Q(oe);
        }
        te();
      }
    },
    [n, s, re, te, _]
  ), de = b(
    (e) => {
      if (e.button !== 0) return;
      const r = Se(e);
      r && ce(r, e);
    },
    [ce]
  ), ae = b(
    (e) => {
      const r = Se(e);
      r && (K.current = setTimeout(() => {
        P(!0), ce(r, e.touches[0]);
      }, 300));
    },
    [ce]
  ), pe = b((e) => {
    $(e);
  }, []), xe = b(() => {
    if (W.current) {
      const { dx: e, id: r, marker: a, value: x } = W.current;
      W.current = null, typeof x < "u" && e && n.exec("update-task", {
        id: r,
        task: { progress: x },
        inProgress: !1
      }), a.classList.remove("wx-progress-in-drag"), q.current = !0, ie();
    } else if (Y) {
      const { id: e, mode: r, dx: a, l: x, w: I, start: z, segment: U, index: oe } = Y;
      if (Q(null), z) {
        const ye = Math.round(a / F);
        if (!ye)
          n.exec("drag-task", {
            id: e,
            width: I,
            left: x,
            inProgress: !1,
            ...U && { segmentIndex: oe }
          });
        else {
          let Ne = {}, Me = n.getTask(e);
          U && (Me = Me.segments[oe]), r === "move" ? (Ne.start = Me.start, Ne.end = Me.end) : Ne[r] = Me[r], n.exec("update-task", {
            id: e,
            diff: ye,
            task: Ne,
            ...U && { segmentIndex: oe }
          });
        }
        q.current = !0;
      }
      ie();
    }
  }, [n, ie, Y, F]), Te = b(
    (e, r) => {
      const { clientX: a } = r;
      if (!s)
        if (W.current) {
          const { node: x, x: I, id: z } = W.current, U = W.current.dx = a - I, oe = Math.round(U / x.offsetWidth * 100);
          let ye = W.current.progress + oe;
          W.current.value = ye = Math.min(
            Math.max(0, ye),
            100
          ), n.exec("update-task", {
            id: z,
            task: { progress: ye },
            inProgress: !0
          });
        } else if (Y) {
          pe(null);
          const { mode: x, l: I, w: z, x: U, id: oe, start: ye, segment: Ne, index: Me } = Y, Qe = n.getTask(oe), ve = a - U, tt = Math.round(F) || 1;
          if (!ye && Math.abs(ve) < 20 || x === "start" && z - ve < tt || x === "end" && z + ve < tt || x === "move" && (ve < 0 && I + ve < 0 || ve > 0 && I + z + ve > ue) || Y.segment && !Wt(Qe, Y))
            return;
          const Ue = { ...Y, dx: ve };
          let Be, Xe;
          if (x === "start" ? (Be = I + ve, Xe = z - ve) : x === "end" ? (Be = I, Xe = z + ve) : x === "move" && (Be = I + ve, Xe = z), n.exec("drag-task", {
            id: oe,
            width: Xe,
            left: Be,
            inProgress: !0,
            start: ye,
            ...Ne && { segmentIndex: Me }
          }), !Ue.start && (x === "move" && Qe.$x === I || x !== "move" && Qe.$w === z)) {
            q.current = !0, xe();
            return;
          }
          Ue.start = !0, Q(Ue);
        } else {
          const x = Se(e);
          if (x) {
            const I = n.getTask(De(x)), U = Se(e, "data-segment") || x, oe = re(U, r, I);
            U.style.cursor = oe && !s ? "col-resize" : "pointer";
          }
        }
    },
    [
      n,
      s,
      Y,
      F,
      ue,
      re,
      pe,
      xe
    ]
  ), $e = b(
    (e) => {
      Te(e, e);
    },
    [Te]
  ), ge = b(
    (e) => {
      A ? (e.preventDefault(), Te(e, e.touches[0])) : K.current && (clearTimeout(K.current), K.current = null);
    },
    [A, Te]
  ), Ce = b(() => {
    xe();
  }, [xe]), Ee = b(() => {
    P(null), K.current && (clearTimeout(K.current), K.current = null), xe();
  }, [xe]);
  ee(() => (window.addEventListener("mouseup", Ce), () => {
    window.removeEventListener("mouseup", Ce);
  }), [Ce]);
  const fe = b(
    (e) => {
      if (!s) {
        const r = Ae(e.target);
        if (r && !e.target.classList.contains("wx-link")) {
          const a = Ae(e.target, "data-segment");
          n.exec("show-editor", {
            id: r,
            ...a !== null && { segmentIndex: a }
          });
        }
      }
    },
    [n, s]
  ), we = ["e2s", "s2s", "e2e", "s2e"], g = b((e, r) => we[(e ? 1 : 0) + (r ? 0 : 2)], []), J = b(
    (e, r) => {
      const a = k.id, x = k.start;
      return e === a ? !0 : !!f.find((I) => I.target === e && I.source === a && I.type === g(x, r));
    },
    [k, d, g]
  ), le = b(() => {
    k && ne(null);
  }, [k]), j = b(
    (e) => {
      if (q.current) {
        q.current = !1;
        return;
      }
      const r = Ae(e.target);
      if (r) {
        const a = e.target.classList;
        if (a.contains("wx-link")) {
          const x = a.contains("wx-left");
          if (!k) {
            ne({ id: r, start: x });
            return;
          }
          k.id !== r && !J(r, x) && n.exec("add-link", {
            link: {
              source: k.id,
              target: r,
              type: g(k.start, x)
            }
          });
        } else if (a.contains("wx-delete-button-icon"))
          n.exec("delete-link", { id: O }), $(null);
        else {
          const x = Ae(e.target, "data-segment");
          n.exec("select-task", {
            id: r,
            toggle: e.ctrlKey || e.metaKey,
            range: e.shiftKey,
            ...x !== null && { segmentIndex: x }
          });
        }
      }
      le();
    },
    [
      n,
      k,
      d,
      L,
      J,
      g,
      le
    ]
  ), he = b((e) => ({
    left: `${e.$x}px`,
    top: `${e.$y}px`,
    width: `${e.$w}px`,
    height: `${e.$h}px`,
    lineHeight: `${e.$h}px`
  }), []), Re = b((e) => ({
    left: `${e.$x_base}px`,
    top: `${e.$y_base}px`,
    width: `${e.$w_base}px`,
    height: `${e.$h_base}px`
  }), []), _e = b((e) => ({
    left: `${e.$x_slack}px`,
    top: `${e.$y}px`,
    width: `${e.$w_slack}px`,
    height: `${e.$h}px`
  }), []), Ge = b(
    (e) => {
      if (A || K.current)
        return e.preventDefault(), !1;
    },
    [A]
  ), Pe = v(
    () => m.map((e) => e.id),
    [m]
  ), Ve = b(
    (e) => {
      let r = Pe.includes(e) ? e : "task";
      return ["task", "milestone", "summary"].includes(e) || (r = `task ${r}`), r;
    },
    [Pe]
  ), ke = b(
    (e) => {
      n.exec(e.action, e.data);
    },
    [n]
  ), We = b(
    (e) => S && e.critical,
    [S]
  ), Ke = b(
    (e) => {
      if (y?.auto) {
        const r = B.getSummaryId(e, !0), a = B.getSummaryId(k.id, !0);
        return k?.id && !(Array.isArray(r) ? r : [r]).includes(
          k.id
        ) && !(Array.isArray(a) ? a : [a]).includes(e);
      }
      return k;
    },
    [y, B, k]
  );
  return /* @__PURE__ */ me(
    "div",
    {
      className: "wx-GKbcLEGA wx-bars",
      ref: C,
      onContextMenu: Ge,
      onMouseDown: de,
      onMouseMove: $e,
      onTouchStart: ae,
      onTouchMove: ge,
      onTouchEnd: Ee,
      onClick: j,
      onDoubleClick: fe,
      onDragStart: (e) => (e.preventDefault(), !1),
      children: [
        G ? h.map(
          (e) => e.$visibleSlack ? /* @__PURE__ */ i(
            "div",
            {
              className: `wx-GKbcLEGA wx-slack wx-slack-${e.type}`,
              style: _e(e)
            },
            e.id
          ) : null
        ) : null,
        /* @__PURE__ */ i(
          nn,
          {
            onSelectLink: pe,
            selectedLink: L,
            readonly: s
          }
        ),
        h.map((e) => {
          if (e.$skip && e.$skip_baseline && !(c && D?.[e.id])) return null;
          const r = `wx-bar wx-${Ve(e.type)}` + (A && Y && e.id === Y.id ? " wx-touch" : "") + (k && k.id === e.id ? " wx-selected" : "") + (We(e) ? " wx-critical" : "") + (e.$reorder ? " wx-reorder-task" : "") + (_ && e.segments ? " wx-split" : ""), a = "wx-link wx-left" + (k ? " wx-visible" : "") + (!k || !J(e.id, !0) && Ke(e.id) ? " wx-target" : "") + (k && k.id === e.id && k.start ? " wx-selected" : "") + (We(e) ? " wx-critical" : ""), x = "wx-link wx-right" + (k ? " wx-visible" : "") + (!k || !J(e.id, !1) && Ke(e.id) ? " wx-target" : "") + (k && k.id === e.id && !k.start ? " wx-selected" : "") + (We(e) ? " wx-critical" : "");
          return /* @__PURE__ */ me(vt, { children: [
            !e.$skip && /* @__PURE__ */ me(
              "div",
              {
                className: "wx-GKbcLEGA " + r,
                style: he(e),
                "data-id": Fe(e.id),
                "data-tooltip-id": Fe(e.id),
                tabIndex: Z === e.id ? 0 : -1,
                children: [
                  s ? null : e.id === L?.target && L?.type[2] === "s" ? /* @__PURE__ */ i(
                    nt,
                    {
                      type: "danger",
                      css: "wx-left wx-delete-button wx-delete-link",
                      children: /* @__PURE__ */ i("i", { className: "wxi-close wx-delete-button-icon" })
                    }
                  ) : /* @__PURE__ */ i("div", { className: "wx-GKbcLEGA " + a, children: /* @__PURE__ */ i("div", { className: "wx-GKbcLEGA wx-inner" }) }),
                  e.type !== "milestone" ? /* @__PURE__ */ me(Oe, { children: [
                    e.progress && !(_ && e.segments) ? /* @__PURE__ */ i("div", { className: "wx-GKbcLEGA wx-progress-wrapper", children: /* @__PURE__ */ i(
                      "div",
                      {
                        className: "wx-GKbcLEGA wx-progress-percent",
                        style: { width: `${e.progress}%` }
                      }
                    ) }) : null,
                    !s && !(_ && e.segments) && !(e.type === "summary" && M?.autoProgress) ? /* @__PURE__ */ i(
                      "div",
                      {
                        className: "wx-GKbcLEGA wx-progress-marker",
                        style: { left: `calc(${e.progress}% - 10px)` },
                        children: e.progress
                      }
                    ) : null,
                    o ? /* @__PURE__ */ i(o, { data: e, api: n, onAction: ke }) : _ && e.segments ? /* @__PURE__ */ i(sn, { task: e, type: Ve(e.type) }) : /* @__PURE__ */ i("div", { className: "wx-GKbcLEGA wx-content", children: e.text || "" })
                  ] }) : /* @__PURE__ */ me(Oe, { children: [
                    /* @__PURE__ */ i("div", { className: "wx-GKbcLEGA wx-content" }),
                    o ? /* @__PURE__ */ i(o, { data: e, api: n, onAction: ke }) : /* @__PURE__ */ i("div", { className: "wx-GKbcLEGA wx-text-out", children: e.text })
                  ] }),
                  s ? null : e.id === L?.target && L?.type[2] === "e" ? /* @__PURE__ */ i(
                    nt,
                    {
                      type: "danger",
                      css: "wx-right wx-delete-button wx-delete-link",
                      children: /* @__PURE__ */ i("i", { className: "wxi-close wx-delete-button-icon" })
                    }
                  ) : /* @__PURE__ */ i("div", { className: "wx-GKbcLEGA " + x, children: /* @__PURE__ */ i("div", { className: "wx-GKbcLEGA wx-inner" }) })
                ]
              }
            ),
            c && D?.[e.id] ? D[e.id].map((I, z) => /* @__PURE__ */ i(rn, { rollup: I, parent: e }, z)) : null,
            V && !e.$skip_baseline ? /* @__PURE__ */ i(
              "div",
              {
                className: "wx-GKbcLEGA wx-baseline" + (e.type === "milestone" ? " wx-milestone" : ""),
                style: Re(e)
              }
            ) : null
          ] }, e.id);
        })
      ]
    }
  );
}
function ln(t) {
  const { highlightTime: s } = t, o = be(Ie), n = R(o, "_scales");
  return /* @__PURE__ */ i("div", { className: "wx-ZkvhDKir wx-scale", style: { width: n.width }, children: (n?.rows || []).map((l, N) => /* @__PURE__ */ i(
    "div",
    {
      className: "wx-ZkvhDKir wx-row",
      style: { height: `${l.height}px` },
      children: (l.cells || []).map((f, d) => {
        const u = s ? s(f.date, f.unit) : "", E = "wx-cell " + (f.css || "") + " " + (u || "");
        return /* @__PURE__ */ i(
          "div",
          {
            className: "wx-ZkvhDKir " + E,
            style: { width: `${f.width}px` },
            children: f.value
          },
          d
        );
      })
    },
    N
  )) });
}
const cn = /* @__PURE__ */ new Map();
function an(t) {
  const s = X(null), o = X(0), n = X(null), l = typeof window < "u" && window.__RENDER_METRICS_ENABLED__;
  s.current === null && (s.current = performance.now()), o.current++, ee(() => {
    if (l)
      return cancelAnimationFrame(n.current), n.current = requestAnimationFrame(() => {
        const N = {
          label: t,
          time: performance.now() - s.current,
          renders: o.current,
          timestamp: Date.now()
        };
        cn.set(t, N), window.dispatchEvent(
          new CustomEvent("render-metric", { detail: N })
        );
      }), () => cancelAnimationFrame(n.current);
  });
}
function un(t) {
  const {
    readonly: s,
    fullWidth: o,
    fullHeight: n,
    taskTemplate: l,
    cellBorders: N,
    highlightTime: f
  } = t, d = be(Ie), [u, E] = Je(d, "_selected"), m = R(d, "scrollTop"), V = R(d, "scrollLeft"), T = R(d, "cellHeight"), c = R(d, "_scales"), D = R(d, "_markers"), H = R(d, "zoom"), [S, B] = se(), y = X(null), _ = X({ top: null, left: null }), M = X(!1), G = 1, h = v(() => {
    const p = [];
    return u && u.length && T && u.forEach((C) => {
      p.push({ height: `${T}px`, top: `${C.$y - 3}px` });
    }), p;
  }, [E, T]), F = v(
    () => Math.max(S || 0, n),
    [S, n]
  );
  Tt(() => {
    const p = y.current;
    if (p) {
      if (M.current) {
        M.current = !1;
        return;
      }
      typeof m == "number" && (_.current.top = m, p.scrollTop = m), typeof V == "number" && (_.current.left = V, p.scrollLeft = V);
    }
  }, [m, V]);
  const q = () => {
    k(), ne();
  };
  function k() {
    const p = y.current;
    if (!p) return;
    const C = {};
    p.scrollTop !== _.current.top && (C.top = p.scrollTop), p.scrollLeft !== _.current.left && (C.left = p.scrollLeft), Object.keys(C).length && (C.top !== void 0 && (_.current.top = C.top), C.left !== void 0 && (_.current.left = C.left), M.current = !0, d.exec("scroll-chart", C));
  }
  function ne() {
    const p = y.current, w = Math.ceil((S || 0) / (T || 1)) + 1, Z = Math.floor((p && p.scrollTop || 0) / (T || 1)), te = Math.max(0, Z - G), ie = Z + w + G, re = te * (T || 0);
    d.exec("render-data", {
      start: te,
      end: ie,
      from: re
    });
  }
  ee(() => {
    ne();
  }, [S]);
  const Y = X(performance.now()), Q = X(!1), W = 3e-3;
  function O(p, C, w) {
    return Math.max(Math.min(p, w), C);
  }
  function $(p) {
    const w = Math.abs(p) < 50 ? 4e-3 : 0.01, Z = performance.now(), te = Math.min(Z - Y.current, 50);
    Y.current = Z;
    const ie = O(
      -p * w,
      -W * te,
      W * te
    );
    return Math.exp(ie);
  }
  const L = b((p) => {
    if (H && (p.ctrlKey || p.metaKey)) {
      p.preventDefault();
      const C = y.current, w = $(p.deltaY), Z = p.clientX - (C ? C.getBoundingClientRect().left : 0);
      Q.current || (Q.current = !0, requestAnimationFrame(() => {
        d.exec("zoom-scale", {
          dir: w > 1 ? 1 : -1,
          ratio: Math.abs(1 - w),
          offset: Z
        }), Q.current = !1;
      }));
    }
  }, [H, d]);
  function A(p) {
    const C = f(p.date, p.unit);
    return C ? {
      css: C,
      width: p.width
    } : null;
  }
  const P = v(() => c && (c.minUnit === "hour" || c.minUnit === "day") && f ? c.rows[c.rows.length - 1].cells.map(A) : null, [c, f]), K = b(
    (p) => {
      p.eventSource = "chart", d.exec("hotkey", p);
    },
    [d]
  );
  ee(() => {
    const p = y.current;
    if (!p) return;
    const C = () => B(p.clientHeight);
    C();
    const w = new ResizeObserver(() => C());
    return w.observe(p), () => {
      w.disconnect();
    };
  }, [y.current]);
  const ue = X(null);
  return ee(() => {
    const p = y.current;
    if (p && !ue.current)
      return ue.current = wt(p, {
        keys: {
          arrowup: !0,
          arrowdown: !0
        },
        exec: (C) => K(C)
      }), () => {
        ue.current?.destroy(), ue.current = null;
      };
  }, []), ee(() => {
    const p = y.current;
    if (!p) return;
    const C = L;
    return p.addEventListener("wheel", C), () => {
      p.removeEventListener("wheel", C);
    };
  }, [L]), an("chart"), /* @__PURE__ */ me(
    "div",
    {
      className: "wx-mR7v2Xag wx-chart",
      tabIndex: -1,
      ref: y,
      onScroll: q,
      children: [
        /* @__PURE__ */ i(ln, { highlightTime: f, scales: c }),
        D && D.length ? /* @__PURE__ */ i(
          "div",
          {
            className: "wx-mR7v2Xag wx-markers",
            style: { height: `${F}px` },
            children: D.map((p, C) => /* @__PURE__ */ i(
              "div",
              {
                className: `wx-mR7v2Xag wx-marker ${p.css || ""}`,
                style: { left: `${p.left}px` },
                children: /* @__PURE__ */ i("div", { className: "wx-mR7v2Xag wx-content", children: p.text })
              },
              C
            ))
          }
        ) : null,
        /* @__PURE__ */ me(
          "div",
          {
            className: "wx-mR7v2Xag wx-area",
            style: { width: `${o}px`, height: `${F}px` },
            children: [
              P ? /* @__PURE__ */ i(
                "div",
                {
                  className: "wx-mR7v2Xag wx-gantt-holidays",
                  style: { height: "100%" },
                  children: P.map(
                    (p, C) => p ? /* @__PURE__ */ i(
                      "div",
                      {
                        className: "wx-mR7v2Xag " + p.css,
                        style: {
                          width: `${p.width}px`,
                          left: `${C * p.width}px`
                        }
                      },
                      C
                    ) : null
                  )
                }
              ) : null,
              /* @__PURE__ */ i(tn, { borders: N }),
              u && u.length ? u.map(
                (p, C) => p.$y ? /* @__PURE__ */ i(
                  "div",
                  {
                    className: "wx-mR7v2Xag wx-selected",
                    "data-id": Fe(p.id),
                    style: h[C]
                  },
                  p.id
                ) : null
              ) : null,
              /* @__PURE__ */ i(on, { readonly: s, taskTemplate: l })
            ]
          }
        )
      ]
    }
  );
}
function dn(t) {
  const {
    position: s = "after",
    size: o = 4,
    dir: n = "x",
    onMove: l,
    onDisplayChange: N,
    compactMode: f,
    containerWidth: d = 0,
    leftThreshold: u = 50,
    rightThreshold: E = 50
  } = t, [m, V] = Ze(t.value ?? 0), [T, c] = Ze(t.display ?? "all");
  function D(A) {
    let P = 0;
    s == "center" ? P = o / 2 : s == "before" && (P = o);
    const K = {
      size: [o + "px", "auto"],
      p: [A - P + "px", "0px"],
      p2: ["auto", "0px"]
    };
    if (n != "x")
      for (let ue in K) K[ue] = K[ue].reverse();
    return K;
  }
  const [H, S] = se(!1), [B, y] = se(null), _ = X(0), M = X(), G = X(T);
  ee(() => {
    G.current = T;
  }, [T]), ee(() => {
    B === null && m > 0 && y(m);
  }, [B, m]);
  function h(A) {
    return n == "x" ? A.clientX : A.clientY;
  }
  const F = b(
    (A) => {
      const P = M.current + h(A) - _.current;
      V(P);
      let K;
      P <= u ? K = "chart" : d - P <= E ? K = "grid" : K = "all", G.current !== K && (c(K), G.current = K), l && l(P);
    },
    [d, u, E, l]
  ), q = b(() => {
    document.body.style.cursor = "", document.body.style.userSelect = "", S(!1), window.removeEventListener("mousemove", F), window.removeEventListener("mouseup", q);
  }, [F]), k = v(
    () => T !== "all" ? "auto" : n == "x" ? "ew-resize" : "ns-resize",
    [T, n]
  ), ne = b(
    (A) => {
      !f && (T === "grid" || T === "chart") || (_.current = h(A), M.current = m, S(!0), document.body.style.cursor = k, document.body.style.userSelect = "none", window.addEventListener("mousemove", F), window.addEventListener("mouseup", q));
    },
    [k, F, q, m, f, T]
  );
  function Y() {
    c("all"), B !== null && (V(B), l && l(B));
  }
  function Q(A) {
    if (f) {
      const P = T === "chart" ? "grid" : "chart";
      c(P), N(P);
    } else if (T === "grid" || T === "chart")
      Y(), N("all");
    else {
      const P = A === "left" ? "chart" : "grid";
      c(P), N(P);
    }
  }
  function W() {
    Q("left");
  }
  function O() {
    Q("right");
  }
  const $ = v(() => D(m), [m, s, o, n]), L = [
    "wx-resizer",
    `wx-resizer-${n}`,
    `wx-resizer-display-${T}`,
    H ? "wx-resizer-active" : ""
  ].filter(Boolean).join(" ");
  return /* @__PURE__ */ me(
    "div",
    {
      className: "wx-pFykzMlT " + L,
      onMouseDown: ne,
      style: { width: $.size[0], height: $.size[1], cursor: k },
      children: [
        /* @__PURE__ */ me("div", { className: "wx-pFykzMlT wx-button-expand-box", children: [
          /* @__PURE__ */ i("div", { className: "wx-pFykzMlT wx-button-expand-content wx-button-expand-left", children: /* @__PURE__ */ i(
            "i",
            {
              className: "wx-pFykzMlT wxi-menu-left",
              onClick: W
            }
          ) }),
          /* @__PURE__ */ i("div", { className: "wx-pFykzMlT wx-button-expand-content wx-button-expand-right", children: /* @__PURE__ */ i(
            "i",
            {
              className: "wx-pFykzMlT wxi-menu-right",
              onClick: O
            }
          ) })
        ] }),
        /* @__PURE__ */ i("div", { className: "wx-pFykzMlT wx-resizer-line" })
      ]
    }
  );
}
const fn = 650;
function yt(t) {
  let s;
  function o() {
    s = new ResizeObserver((l) => {
      for (let N of l)
        if (N.target === document.body) {
          let f = N.contentRect.width <= fn;
          t(f);
        }
    }), s.observe(document.body);
  }
  function n() {
    s && (s.disconnect(), s = null);
  }
  return {
    observe: o,
    disconnect: n
  };
}
function hn(t) {
  const {
    taskTemplate: s,
    readonly: o,
    cellBorders: n,
    highlightTime: l,
    onTableAPIChange: N
  } = t, f = be(Ie), d = R(f, "_tasks"), u = R(f, "_scales"), E = R(f, "cellHeight"), m = R(f, "columns"), V = R(f, "undo"), [T, c] = se(!1);
  let [D, H] = se(0);
  const [S, B] = se(0), [y, _] = se(0), [M, G] = se(void 0), [h, F] = se("all"), q = X(null), k = b(
    (w) => {
      c((Z) => (w !== Z && (w ? (q.current = h, h === "all" && F("grid")) : (!q.current || q.current === "all") && F("all")), w));
    },
    [h]
  );
  ee(() => {
    const w = yt(k);
    return w.observe(), () => {
      w.disconnect();
    };
  }, [k]);
  const ne = v(() => {
    let w;
    return m.every((Z) => Z.width && !Z.flexgrow) ? w = m.reduce((Z, te) => Z + parseInt(te.width), 0) : h === "chart" ? w = parseInt(m.find((Z) => Z.id === "action")?.width) || 50 : w = 440, D = w, w;
  }, [m, h]);
  ee(() => {
    H(ne);
  }, [ne]);
  const Y = v(
    () => (S ?? 0) - (M ?? 0),
    [S, M]
  ), Q = v(() => u.width, [u]), W = v(
    () => d.length * E,
    [d, E]
  ), O = v(
    () => u.height + W + Y,
    [u, W, Y]
  ), $ = X(null), L = X({
    ganttWidth: 0,
    gridWidth: 0,
    ganttHeight: 0,
    rScalesHeight: 0,
    scrollSize: 0
  });
  ee(() => {
    L.current = {
      ganttWidth: S ?? 0,
      gridWidth: D,
      ganttHeight: y ?? 0,
      rScalesHeight: u.height,
      scrollSize: Y
    };
  }, [S, D, y, u, Y]);
  const A = b(() => {
    const { ganttWidth: w, gridWidth: Z, ganttHeight: te, rScalesHeight: ie, scrollSize: re } = L.current;
    f.exec("resize-chart", {
      width: w - Z,
      height: te - ie,
      scrollSize: re
    });
  }, [f]);
  ee(() => {
    let w;
    return $.current && (w = new ResizeObserver(A), w.observe($.current)), () => {
      w && w.disconnect();
    };
  }, [$.current, A]);
  const P = X(null), K = X(null), ue = b(() => {
    const w = P.current;
    w && f.exec("scroll-chart", {
      top: w.scrollTop
    });
  }, [f]);
  ee(() => {
    const w = P.current, Z = K.current;
    if (!w || !Z) return;
    const te = () => {
      jt(() => {
        _(w.offsetHeight), B(w.offsetWidth), G(Z.offsetWidth);
      });
    }, ie = new ResizeObserver(te);
    return ie.observe(w), () => ie.disconnect();
  }, [P.current]);
  const p = X(null), C = X(null);
  return ee(() => {
    C.current && (C.current.destroy(), C.current = null);
    const w = p.current;
    if (w)
      return C.current = wt(w, {
        keys: {
          "ctrl+c": !0,
          "ctrl+v": !0,
          "ctrl+x": !0,
          "ctrl+d": !0,
          backspace: !0,
          "ctrl+z": V,
          "ctrl+y": V
        },
        exec: (Z) => {
          Z.isInput || f.exec("hotkey", Z);
        }
      }), () => {
        C.current?.destroy(), C.current = null;
      };
  }, [V]), /* @__PURE__ */ i("div", { className: "wx-jlbQoHOz wx-gantt", ref: P, onScroll: ue, children: /* @__PURE__ */ i(
    "div",
    {
      className: "wx-jlbQoHOz wx-pseudo-rows",
      style: { height: O, width: "100%" },
      ref: K,
      children: /* @__PURE__ */ i(
        "div",
        {
          className: "wx-jlbQoHOz wx-stuck",
          style: {
            height: y,
            width: M
          },
          children: /* @__PURE__ */ me("div", { tabIndex: 0, className: "wx-jlbQoHOz wx-layout", ref: p, children: [
            m.length ? /* @__PURE__ */ me(Oe, { children: [
              /* @__PURE__ */ i(
                en,
                {
                  display: h,
                  compactMode: T,
                  columnWidth: ne,
                  width: D,
                  readonly: o,
                  fullHeight: W,
                  onTableAPIChange: N
                }
              ),
              /* @__PURE__ */ i(
                dn,
                {
                  value: D,
                  display: h,
                  compactMode: T,
                  containerWidth: S,
                  onMove: (w) => H(w),
                  onDisplayChange: (w) => F(w)
                }
              )
            ] }) : null,
            /* @__PURE__ */ i("div", { className: "wx-jlbQoHOz wx-content", ref: $, children: /* @__PURE__ */ i(
              un,
              {
                readonly: o,
                fullWidth: Q,
                fullHeight: W,
                taskTemplate: s,
                cellBorders: n,
                highlightTime: l
              }
            ) })
          ] })
        }
      )
    }
  ) });
}
function mn(t) {
  return {
    year: "%Y",
    quarter: `${t("Q")} %Q`,
    month: "%M",
    week: `${t("Week")} %w`,
    day: "%M %j",
    hour: "%H:%i"
  };
}
function xn(t, s) {
  return typeof t == "function" ? t : Ye(t, s);
}
function kt(t, s) {
  return t.map(({ format: o, ...n }) => ({
    ...n,
    format: xn(o, s)
  }));
}
function gn(t, s) {
  const o = mn(s);
  for (let n in o)
    o[n] = Ye(o[n], t);
  return o;
}
function pn(t, s) {
  if (!t || !t.length) return t;
  const o = Ye("%d-%m-%Y", s);
  return t.map((n) => n.template ? n : n.id === "start" || n.id === "end" ? {
    ...n,
    //store locale template for unscheduled tasks
    _template: (l) => o(l),
    template: (l) => o(l)
  } : n.id === "duration" ? {
    ...n,
    _template: (l) => l,
    template: (l) => l
  } : n);
}
function wn(t, s) {
  return t.levels ? {
    ...t,
    levels: t.levels.map((o) => ({
      ...o,
      scales: kt(o.scales, s)
    }))
  } : t;
}
const yn = (t) => t.split("-").map((s) => s ? s.charAt(0).toUpperCase() + s.slice(1) : "").join(""), kn = [
  { unit: "month", step: 1, format: "%F %Y" },
  { unit: "day", step: 1, format: "%j" }
], Pn = ht(function({
  taskTemplate: s = null,
  markers: o = [],
  taskTypes: n = Ot,
  tasks: l = [],
  selected: N = [],
  activeTask: f = null,
  links: d = [],
  scales: u = kn,
  columns: E = Ft,
  start: m = null,
  end: V = null,
  lengthUnit: T = "day",
  durationUnit: c = "day",
  cellWidth: D = 100,
  cellHeight: H = 38,
  scaleHeight: S = 36,
  readonly: B = !1,
  cellBorders: y = "full",
  zoom: _ = !1,
  baselines: M = !1,
  rollups: G = !1,
  highlightTime: h = null,
  init: F = null,
  autoScale: q = !0,
  unscheduledTasks: k = !1,
  criticalPath: ne = null,
  schedule: Y = { type: "forward" },
  projectStart: Q = null,
  projectEnd: W = null,
  calendar: O = null,
  undo: $ = !1,
  splitTasks: L = !1,
  summary: A = null,
  slack: P = !1,
  _export: K = !1,
  ...ue
}, p) {
  const C = X();
  C.current = ue;
  const w = v(() => new _t(Bt), []), Z = v(() => ({ ...et, ...qe }), []), te = be(Le.i18n), ie = v(() => te ? te.extend(Z, !0) : je(Z), [te, Z]), re = v(() => ie.getRaw().calendar, [ie]), ce = v(() => {
    let fe = {
      zoom: wn(_, re),
      scales: kt(u, re),
      columns: pn(E, re),
      links: Pt(d),
      cellWidth: D
    };
    return fe.zoom && (fe = {
      ...fe,
      ...Vt(
        fe.zoom,
        gn(re, ie.getGroup("gantt")),
        fe.scales,
        D
      )
    }), fe;
  }, [_, u, E, d, D, re, ie]), de = X(null);
  de.current !== l && (K || lt(l, { durationUnit: c, splitTasks: L, calendar: O }), de.current = l), ee(() => {
    K || lt(l, { durationUnit: c, splitTasks: L, calendar: O });
  }, [l, c, O, L, K]);
  const ae = v(() => w.in, [w]), pe = X(null);
  pe.current === null && (pe.current = new It((fe, we) => {
    const g = "on" + yn(fe);
    C.current && C.current[g] && C.current[g](we);
  }), ae.setNext(pe.current));
  const [xe, Te] = se(null), $e = X(null);
  $e.current = xe;
  const ge = v(
    () => ({
      getState: w.getState.bind(w),
      getReactiveState: w.getReactive.bind(w),
      getStores: () => ({ data: w }),
      exec: ae.exec,
      setNext: (fe) => (pe.current = pe.current.setNext(fe), pe.current),
      intercept: ae.intercept.bind(ae),
      on: ae.on.bind(ae),
      detach: ae.detach.bind(ae),
      getTask: w.getTask.bind(w),
      serialize: () => w.serialize(),
      getTable: (fe) => fe ? new Promise((we) => setTimeout(() => we($e.current), 1)) : $e.current,
      getHistory: () => w.getHistory()
    }),
    [w, ae]
  );
  mt(
    p,
    () => ({
      ...ge
    }),
    [ge]
  );
  const Ce = X(0);
  ee(() => {
    Ce.current ? w.init({
      tasks: l,
      links: ce.links,
      start: m,
      columns: ce.columns,
      end: V,
      lengthUnit: T,
      cellWidth: ce.cellWidth,
      cellHeight: H,
      scaleHeight: S,
      scales: ce.scales,
      taskTypes: n,
      zoom: ce.zoom,
      selected: N,
      activeTask: f,
      baselines: M,
      rollups: G === !0 ? { type: "closest" } : G,
      autoScale: q,
      unscheduledTasks: k,
      markers: o,
      durationUnit: c,
      criticalPath: ne,
      schedule: Y,
      projectStart: Q,
      projectEnd: W,
      calendar: O,
      slack: P,
      undo: $,
      _weekStart: re.weekStart,
      splitTasks: L,
      summary: A
    }) : F && F(ge), Ce.current++;
  }, [
    ge,
    F,
    l,
    ce,
    m,
    V,
    T,
    H,
    S,
    n,
    N,
    f,
    M,
    G,
    q,
    k,
    o,
    c,
    ne,
    Y,
    Q,
    W,
    O,
    P,
    $,
    re,
    L,
    A,
    w
  ]), Ce.current === 0 && w.init({
    tasks: l,
    links: ce.links,
    start: m,
    columns: ce.columns,
    end: V,
    lengthUnit: T,
    cellWidth: ce.cellWidth,
    cellHeight: H,
    scaleHeight: S,
    scales: ce.scales,
    taskTypes: n,
    zoom: ce.zoom,
    selected: N,
    activeTask: f,
    baselines: M,
    rollups: G === !0 ? { type: "closest" } : G,
    autoScale: q,
    unscheduledTasks: k,
    markers: o,
    durationUnit: c,
    criticalPath: ne,
    schedule: Y,
    projectStart: Q,
    projectEnd: W,
    calendar: O,
    slack: P,
    undo: $,
    _weekStart: re.weekStart,
    splitTasks: L,
    summary: A
  });
  const Ee = v(() => O ? (fe, we) => we === "day" && !O.getDayHours(fe) || we === "hour" && !O.getDayHours(fe) ? "wx-weekend" : "" : h, [O, h]);
  return /* @__PURE__ */ i(Le.i18n.Provider, { value: ie, children: /* @__PURE__ */ i(Ie.Provider, { value: ge, children: /* @__PURE__ */ i(
    hn,
    {
      taskTemplate: s,
      readonly: B,
      cellBorders: y,
      highlightTime: Ee,
      onTableAPIChange: Te
    }
  ) }) });
});
function Vn({ api: t = null, items: s = [] }) {
  const o = be(Le.i18n), n = v(() => o || je(qe), [o]), l = v(() => n.getGroup("gantt"), [n]), N = He(t, "_selected"), f = He(t, "undo"), d = He(t, "history"), u = He(t, "splitTasks"), E = ["undo", "redo"], m = v(() => {
    const T = it({ undo: !0, splitTasks: !0 });
    return (s.length ? s : it({
      undo: f,
      splitTasks: u
    })).map((D) => {
      let H = { ...D, disabled: !1 };
      return H.handler = gt(T, H.id) ? (S) => pt(t, S.id, null, l) : H.handler, H.text && (H.text = l(H.text)), H.menuText && (H.menuText = l(H.menuText)), H;
    });
  }, [s, t, l, f, u]), V = v(() => {
    const T = [];
    return m.forEach((c) => {
      const D = c.id;
      if (D === "add-task")
        T.push(c);
      else if (E.includes(D))
        E.includes(D) && T.push({
          ...c,
          disabled: c.isDisabled(d)
        });
      else {
        if (!N?.length || !t) return;
        T.push({
          ...c,
          disabled: c.isDisabled && N.some((H) => c.isDisabled(H, t.getState()))
        });
      }
    }), T.filter((c, D) => {
      if (t && c.isHidden)
        return !N.some((H) => c.isHidden(H, t.getState()));
      if (c.comp === "separator") {
        const H = T[D + 1];
        if (!H || H.comp === "separator") return !1;
      }
      return !0;
    });
  }, [t, N, d, m]);
  return o ? /* @__PURE__ */ i(at, { items: V }) : /* @__PURE__ */ i(Le.i18n.Provider, { value: n, children: /* @__PURE__ */ i(at, { items: V }) });
}
const Fn = ht(function({
  options: s = [],
  api: o = null,
  resolver: n = null,
  filter: l = null,
  at: N = "point",
  children: f,
  onClick: d,
  css: u
}, E) {
  const m = X(null), V = X(null), T = be(Le.i18n), c = v(() => T || je({ ...qe, ...et }), [T]), D = v(() => c.getGroup("gantt"), [c]), H = He(o, "taskTypes"), S = He(o, "selected"), B = He(o, "_selected"), y = He(o, "splitTasks"), _ = He(o, "summary"), M = v(
    () => ({
      splitTasks: y,
      taskTypes: H,
      summary: _
    }),
    [y, H, _]
  ), G = v(() => ct(M), [M]);
  ee(() => {
    o && (o.on("scroll-chart", () => {
      m.current && m.current.show && m.current.show();
    }), o.on("drag-task", () => {
      m.current && m.current.show && m.current.show();
    }));
  }, [o]);
  function h($) {
    return $.map((L) => (L = { ...L }, L.text && (L.text = D(L.text)), L.subtext && (L.subtext = D(L.subtext)), L.data && (L.data = h(L.data)), L));
  }
  function F() {
    const $ = s.length ? s : ct(M);
    return h($);
  }
  const q = v(() => F(), [o, s, M, D]), k = v(
    () => B && B.length ? B : [],
    [B]
  ), ne = b(
    ($, L) => {
      let A = $ ? o?.getTask($) : null;
      if (n) {
        const P = n($, L);
        A = P === !0 ? A : P;
      }
      if (A) {
        const P = Ae(L.target, "data-segment");
        P !== null ? V.current = { id: A.id, segmentIndex: P } : V.current = A.id, (!Array.isArray(S) || !S.includes(A.id)) && o && o.exec && o.exec("select-task", { id: A.id });
      }
      return A;
    },
    [o, n, S]
  ), Y = b(
    ($) => {
      const L = $.action;
      L && (gt(G, L.id) && pt(o, L.id, V.current, D), d && d($));
    },
    [o, D, d, G]
  ), Q = b(
    ($, L) => {
      const A = k.length ? k : L ? [L] : [];
      let P = l ? A.every((K) => l($, K)) : !0;
      if (P && ($.isHidden && (P = !A.some(
        (K) => $.isHidden(K, o.getState(), V.current)
      )), $.isDisabled)) {
        const K = A.some(
          (ue) => $.isDisabled(ue, o.getState(), V.current)
        );
        $.disabled = K;
      }
      return P;
    },
    [l, k, o]
  );
  mt(E, () => ({
    show: ($, L) => {
      m.current && m.current.show && m.current.show($, L);
    }
  }));
  const W = b(($) => {
    m.current && m.current.show && m.current.show($);
  }, []), O = /* @__PURE__ */ me(Oe, { children: [
    /* @__PURE__ */ i(
      qt,
      {
        filter: Q,
        options: q,
        dataKey: "id",
        resolver: ne,
        onClick: Y,
        at: N,
        ref: m,
        css: u
      }
    ),
    /* @__PURE__ */ i("span", { onContextMenu: W, "data-menu-ignore": "true", children: typeof f == "function" ? f() : f })
  ] });
  if (!T && Le.i18n?.Provider) {
    const $ = Le.i18n.Provider;
    return /* @__PURE__ */ i($, { value: c, children: O });
  }
  return O;
});
function bn({ api: t, autoSave: s, onLinksChange: o }) {
  const l = be(Le.i18n).getGroup("gantt"), N = R(t, "activeTask"), f = R(t, "_activeTask"), d = R(t, "_links"), u = R(t, "schedule"), E = R(t, "unscheduledTasks"), [m, V] = se();
  function T() {
    if (N) {
      const S = d.filter((y) => y.target == N).map((y) => ({ link: y, task: t.getTask(y.source) })), B = d.filter((y) => y.source == N).map((y) => ({ link: y, task: t.getTask(y.target) }));
      return [
        { title: l("Predecessors"), data: S },
        { title: l("Successors"), data: B }
      ];
    }
  }
  ee(() => {
    V(T());
  }, [N, d]);
  const c = v(
    () => [
      { id: "e2s", label: l("End-to-start") },
      { id: "s2s", label: l("Start-to-start") },
      { id: "e2e", label: l("End-to-end") },
      { id: "s2e", label: l("Start-to-end") }
    ],
    [l]
  );
  function D(S) {
    s ? t.exec("delete-link", { id: S }) : (V(
      (B) => (B || []).map((y) => ({
        ...y,
        data: y.data.filter((_) => _.link.id !== S)
      }))
    ), o && o({
      id: S,
      action: "delete-link",
      data: { id: S }
    }));
  }
  function H(S, B) {
    s ? t.exec("update-link", {
      id: S,
      link: B
    }) : (V(
      (y) => (y || []).map((_) => ({
        ..._,
        data: _.data.map(
          (M) => M.link.id === S ? { ...M, link: { ...M.link, ...B } } : M
        )
      }))
    ), o && o({
      id: S,
      action: "update-link",
      data: {
        id: S,
        link: B
      }
    }));
  }
  return /* @__PURE__ */ i(Oe, { children: (m || []).map(
    (S, B) => S.data.length ? /* @__PURE__ */ i("div", { className: "wx-j93aYGQf wx-links", children: /* @__PURE__ */ i(Le.fieldId.Provider, { value: null, children: /* @__PURE__ */ i($t, { label: S.title, position: "top", children: /* @__PURE__ */ i("table", { children: /* @__PURE__ */ i("tbody", { children: S.data.map((y) => /* @__PURE__ */ me("tr", { children: [
      /* @__PURE__ */ i("td", { className: "wx-j93aYGQf wx-cell", children: /* @__PURE__ */ i("div", { className: "wx-j93aYGQf wx-task-name", children: y.task.text || "" }) }),
      u?.auto && y.link.type === "e2s" ? /* @__PURE__ */ i("td", { className: "wx-j93aYGQf wx-cell wx-link-lag", children: /* @__PURE__ */ i(
        Ct,
        {
          type: "number",
          placeholder: l("Lag"),
          value: y.link.lag,
          disabled: E && f?.unscheduled,
          onChange: (_) => {
            _.input || H(y.link.id, { lag: _.value });
          }
        }
      ) }) : null,
      /* @__PURE__ */ i("td", { className: "wx-j93aYGQf wx-cell", children: /* @__PURE__ */ i("div", { className: "wx-j93aYGQf wx-wrapper", children: /* @__PURE__ */ i(
        Rt,
        {
          value: y.link.type,
          placeholder: l("Select link type"),
          options: c,
          onChange: (_) => H(y.link.id, { type: _.value }),
          children: ({ option: _ }) => _.label
        }
      ) }) }),
      /* @__PURE__ */ i("td", { className: "wx-j93aYGQf wx-cell", children: /* @__PURE__ */ i(
        "i",
        {
          className: "wx-j93aYGQf wxi-delete wx-delete-icon",
          onClick: () => D(y.link.id),
          role: "button"
        }
      ) })
    ] }, y.link.id)) }) }) }) }) }, B) : null
  ) });
}
function vn(t) {
  const { value: s, time: o, format: n, onchange: l, onChange: N, ...f } = t, d = N ?? l;
  function u(E) {
    const m = new Date(E.value);
    m.setHours(s.getHours()), m.setMinutes(s.getMinutes()), d && d({ value: m });
  }
  return /* @__PURE__ */ me("div", { className: "wx-hFsbgDln date-time-controll", children: [
    /* @__PURE__ */ i(
      Nt,
      {
        ...f,
        value: s,
        onChange: u,
        format: n,
        buttons: ["today"],
        clear: !1
      }
    ),
    o ? /* @__PURE__ */ i(Mt, { value: s, onChange: d, format: n }) : null
  ] });
}
ze("select", Et);
ze("date", vn);
ze("twostate", St);
ze("slider", Ht);
ze("counter", At);
ze("links", bn);
ze("checkbox", Dt);
function On({
  api: t,
  items: s = [],
  css: o = "",
  layout: n = "default",
  readonly: l = !1,
  placement: N = "sidebar",
  bottomBar: f = !0,
  topBar: d = !0,
  autoSave: u = !0,
  focus: E = !1,
  hotkeys: m = {}
}) {
  const V = be(Le.i18n), T = v(() => V || je({ ...qe, ...et }), [V]), c = v(() => T.getGroup("gantt"), [T]), D = T.getRaw(), H = v(() => {
    const g = D.gantt?.dateFormat || D.formats?.dateFormat;
    return Ye(g, D.calendar);
  }, [D]), S = v(() => {
    if (d === !0 && !l) {
      const g = [
        { comp: "icon", icon: "wxi-close", id: "close" },
        { comp: "spacer" },
        {
          comp: "button",
          type: "danger",
          text: c("Delete"),
          id: "delete"
        }
      ];
      return u ? { items: g } : {
        items: [
          ...g,
          {
            comp: "button",
            type: "primary",
            text: c("Save"),
            id: "save"
          }
        ]
      };
    }
    return d;
  }, [d, l, u, c]), [B, y] = se(!1), _ = v(
    () => B ? "wx-full-screen" : "",
    [B]
  ), M = b((g) => {
    y(g);
  }, []);
  ee(() => {
    const g = yt(M);
    return g.observe(), () => {
      g.disconnect();
    };
  }, [M]);
  const G = R(t, "_activeTask"), h = R(t, "activeTask"), F = R(t, "unscheduledTasks"), q = R(t, "rollups"), k = R(t, "summary"), ne = R(t, "links"), Y = R(t, "splitTasks"), Q = v(
    () => Y && h?.segmentIndex,
    [Y, h]
  ), W = v(
    () => Q || Q === 0,
    [Q]
  ), O = R(t, "taskTypes"), $ = v(
    () => Kt({ unscheduledTasks: F, rollups: q, summary: k, taskTypes: O }),
    [F, q, k, O]
  ), L = R(t, "undo"), [A, P] = se({}), [K, ue] = se(null), [p, C] = se(), [w, Z] = se(null), te = v(() => {
    if (!G) return null;
    let g;
    if (W && G.segments ? g = { ...G.segments[Q] } : g = { ...G }, l) {
      let J = { parent: g.parent };
      return $.forEach(({ key: le, comp: j }) => {
        if (j !== "links") {
          const he = g[le];
          j === "date" && he instanceof Date ? J[le] = H(he) : j === "slider" && le === "progress" ? J[le] = `${he}%` : J[le] = he;
        }
      }), J;
    }
    return g || null;
  }, [G, W, Q, l, $, H]);
  ee(() => {
    C(te);
  }, [te]), ee(() => {
    P({}), Z(null), ue(null);
  }, [h]);
  function ie(g, J) {
    return g.map((le) => {
      const j = { ...le };
      if (le.config && (j.config = { ...j.config }), j.comp === "links" && t && (j.api = t, j.autoSave = u, j.onLinksChange = de), j.comp === "select" && j.key === "type") {
        const he = j.options ?? [];
        j.options = he.map((Re) => ({
          ...Re,
          label: c(Re.label)
        }));
      }
      return j.comp === "slider" && j.key === "progress" && (j.labelTemplate = (he) => `${c(j.label)} ${he}%`), j.label && (j.label = c(j.label)), j.config?.placeholder && (j.config.placeholder = c(j.config.placeholder)), J && (j.isDisabled && j.isDisabled(J, t.getState()) ? j.disabled = !0 : delete j.disabled), j;
    });
  }
  const re = v(() => {
    let g = s.length ? s : $;
    return g = ie(g, p), p ? g.filter(
      (J) => !J.isHidden || !J.isHidden(p, t.getState())
    ) : g;
  }, [s, $, p, c, t, u]), ce = v(
    () => re.map((g) => g.key),
    [re]
  );
  function de({ id: g, action: J, data: le }) {
    P((j) => ({
      ...j,
      [g]: { action: J, data: le }
    }));
  }
  const ae = b(() => {
    for (let g in A)
      if (ne.byId(g)) {
        const { action: J, data: le } = A[g];
        t.exec(J, le);
      }
  }, [t, A, ne]), pe = b(() => {
    const g = h?.id || h;
    if (W) {
      if (G?.segments) {
        const J = G.segments.filter(
          (le, j) => j !== Q
        );
        t.exec("update-task", {
          id: g,
          task: { segments: J }
        });
      }
    } else
      t.exec("delete-task", { id: g });
  }, [t, h, W, G, Q]), xe = b(() => {
    t.exec("show-editor", { id: null });
  }, [t]), Te = b(
    (g) => {
      const { item: J, changes: le } = g;
      J.id === "delete" && pe(), J.id === "save" && (le.length ? xe() : ae()), J.comp && xe();
    },
    [t, h, u, ae, pe, xe]
  ), $e = b(
    (g, J, le) => (F && g.type === "summary" && (g.unscheduled = !1), xt(g, t.getState(), J), le || ue(!1), g),
    [F, t]
  ), ge = b(
    (g) => {
      g = {
        ...g,
        unscheduled: F && g.unscheduled && g.type !== "summary"
      }, delete g.links, delete g.data, (ce.indexOf("duration") === -1 || g.segments && !g.duration) && delete g.duration;
      const J = {
        id: h?.id || h,
        task: g,
        ...W && { segmentIndex: Q }
      };
      u && K && (J.inProgress = K), t.exec("update-task", J), u || ae();
    },
    [
      t,
      h,
      F,
      u,
      ae,
      ce,
      W,
      Q,
      K
    ]
  ), Ce = b(
    (g) => {
      let { update: J, key: le, input: j } = g;
      if (j && ue(!0), g.update = $e({ ...J }, le, j), !u) C(g.update);
      else if (!w && !j) {
        const he = re.find((Ge) => Ge.key === le), Re = J[le];
        (!he.validation || he.validation(Re)) && (!he.required || Re) && ge(g.update);
      }
    },
    [u, $e, w, re, ge]
  ), Ee = b(
    (g) => {
      u || ge(g.values);
    },
    [u, ge]
  ), fe = b((g) => {
    Z(g.errors);
  }, []), we = v(
    () => L ? {
      "ctrl+z": (g) => {
        g.preventDefault(), t.exec("undo");
      },
      "ctrl+y": (g) => {
        g.preventDefault(), t.exec("redo");
      }
    } : {},
    [L, t]
  );
  return te ? /* @__PURE__ */ i(Lt, { children: /* @__PURE__ */ i(
    Qt,
    {
      css: `wx-XkvqDXuw wx-gantt-editor ${_} ${o}`,
      items: re,
      values: te,
      topBar: S,
      bottomBar: f,
      placement: N,
      layout: n,
      readonly: l,
      autoSave: u,
      focus: E,
      onAction: Te,
      onSave: Ee,
      onValidation: fe,
      onChange: Ce,
      hotkeys: m && { ...we, ...m }
    }
  ) }) : null;
}
const Kn = ({ children: t, columns: s = null, api: o }) => {
  const [n, l] = se(null);
  return ee(() => {
    o && o.getTable(!0).then(l);
  }, [o]), /* @__PURE__ */ i(Yt, { api: n, columns: s, children: t });
};
function Bn(t) {
  const { api: s, content: o, children: n } = t, l = X(null), N = X(null), [f, d] = se({}), [u, E] = se({}), [m, V] = se({}), T = X(m);
  T.current = m;
  function c(M) {
    const G = M;
    for (; M; ) {
      if (M.getAttribute) {
        const h = De(M, "data-tooltip-id"), F = De(M, "data-tooltip-at"), q = M.getAttribute("data-tooltip");
        if (h || q) {
          const k = Ae(G, "data-segment");
          return { id: h, tooltip: q, target: M, at: F, segment: k };
        }
      }
      M = M.parentNode;
    }
    return { id: null, tooltip: null, target: null, at: null, segment: null };
  }
  ee(() => {
    const M = N.current, G = T.current;
    if (M && G && (G.text || o)) {
      const h = M.getBoundingClientRect();
      let F = !1, q = G.left, k = G.top;
      h.right >= f.right && (q = f.width - h.width - 5, F = !0), h.bottom >= f.bottom && (k = G.top - (h.bottom - f.bottom + 2), F = !0), F && V((ne) => ne && { ...ne, left: q, top: k });
    }
  }, [f, o]);
  const D = X(null), H = 300, S = (M) => {
    clearTimeout(D.current), D.current = setTimeout(() => {
      M();
    }, H);
  };
  function B(M) {
    let { id: G, tooltip: h, target: F, at: q, segment: k } = c(M.target);
    if (V(null), E({}), !h)
      if (G)
        h = _(G, k);
      else {
        clearTimeout(D.current);
        return;
      }
    const ne = M.clientX;
    S(() => {
      if (G) {
        const L = { data: y(G) };
        k != null && (L.segmentIndex = k), E(L);
      }
      const Y = F.getBoundingClientRect(), Q = l.current, W = Q ? Q.getBoundingClientRect() : { top: 0, left: 0, right: 0, bottom: 0, width: 0, height: 0 };
      let O, $;
      q === "left" ? (O = Y.top + 5 - W.top, $ = Y.right + 5 - W.left) : (O = Y.top + Y.height - W.top, $ = ne - W.left), d(W), V({ top: O, left: $, text: h });
    });
  }
  function y(M) {
    return s?.getTask(M) || null;
  }
  function _(M, G) {
    const h = y(M);
    return G !== null && h?.segments ? h.segments[G]?.text || "" : h?.text || "";
  }
  return /* @__PURE__ */ me(
    "div",
    {
      className: "wx-KG0Lwsqo wx-tooltip-area",
      ref: l,
      onMouseMove: B,
      children: [
        m && (m.text || o) ? /* @__PURE__ */ i(
          "div",
          {
            className: "wx-KG0Lwsqo wx-gantt-tooltip",
            ref: N,
            style: { top: `${m.top}px`, left: `${m.left}px` },
            children: o ? /* @__PURE__ */ i(o, { ...u }) : m.text ? /* @__PURE__ */ i("div", { className: "wx-KG0Lwsqo wx-gantt-tooltip-text", children: m.text }) : null
          }
        ) : null,
        n
      ]
    }
  );
}
function Xn({ fonts: t = !0, children: s }) {
  return s ? /* @__PURE__ */ i(st, { fonts: t, children: s() }) : /* @__PURE__ */ i(st, { fonts: t });
}
function Yn({ fonts: t = !0, children: s }) {
  return s ? /* @__PURE__ */ i(rt, { fonts: t, children: s }) : /* @__PURE__ */ i(rt, { fonts: t });
}
function jn({ fonts: t = !0, children: s }) {
  return s ? /* @__PURE__ */ i(ot, { fonts: t, children: s }) : /* @__PURE__ */ i(ot, { fonts: t });
}
const Tn = "2.6.1", $n = {
  version: Tn
}, qn = $n.version;
export {
  Fn as ContextMenu,
  On as Editor,
  Pn as Gantt,
  Kn as HeaderMenu,
  Xn as Material,
  Vn as Toolbar,
  Bn as Tooltip,
  Yn as Willow,
  jn as WillowDark,
  Zn as defaultColumns,
  Jn as defaultEditorItems,
  es as defaultMenuOptions,
  ts as defaultTaskTypes,
  ns as defaultToolbarButtons,
  ss as getEditorItems,
  rs as getMenuOptions,
  os as getToolbarButtons,
  cs as registerEditorItem,
  ls as registerScaleUnit,
  qn as version
};
