// @svar-ui/react-core/dist/index.es.js
import { jsx as l, jsxs as R, Fragment as _ } from "react/jsx-runtime";
import { createContext as fe, useContext as K, useState as B, useCallback as D, useRef as S, useEffect as W, useMemo as A, useLayoutEffect as me, memo as Ue, useId as _e, forwardRef as Qe, useImperativeHandle as qe } from "react";
import { useWritableProp as J, snippet as ae, useHotkeys as Ke } from "@svar-ui/lib-react";
import { uid as oe, env as pe, calculatePosition as Ze, getAbsParent as ze, clickOutside as ve, locale as Ae, locateID as Xe, getDuodecade as Be, dateToString as ne, delegateClick as Ee, setEnv as et } from "@svar-ui/lib-dom";
import { locale as bn } from "@svar-ui/lib-dom";
import { createPortal as tt } from "react-dom";
import { en as We } from "@svar-ui/core-locales";
import { en as Vn } from "@svar-ui/core-locales";
const ie = fe(""), Oe = fe({}), q = fe(null), ue = fe(null), Ut = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  fieldId: ue,
  helpers: Oe,
  i18n: q,
  theme: ie
}, Symbol.toStringTag, { value: "Module" }));
function z(e) {
  const n = K(ue), [t] = B(() => e || n || oe());
  return t;
}
function _t({
  value: e = "",
  id: n,
  placeholder: t = "",
  title: s = "",
  disabled: r = !1,
  error: u = !1,
  readonly: i = !1,
  onChange: c
}) {
  const a = z(n), [o, f] = J(e), d = D(
    (x) => {
      const w = x.target.value;
      f(w), c && c({ value: w, input: !0 });
    },
    [c]
  ), p = D(
    (x) => {
      const w = x.target.value;
      f(w), c && c({ value: w });
    },
    [c]
  ), v = S(null);
  return W(() => {
    const x = p, w = v.current;
    return w.addEventListener("change", x), () => {
      w && w.removeEventListener("change", x);
    };
  }, [p]), /* @__PURE__ */ l(
    "textarea",
    {
      className: `wx-3yFVAC wx-textarea ${u ? "wx-error" : ""}`,
      id: a,
      disabled: r,
      placeholder: t,
      readOnly: i,
      title: s,
      value: o,
      onInput: d,
      ref: v
    }
  );
}
function re({
  type: e = "",
  css: n = "",
  icon: t = "",
  disabled: s = !1,
  title: r = "",
  text: u = "",
  children: i,
  onClick: c
}) {
  const a = A(() => {
    let f = e ? e.split(" ").filter((d) => d !== "").map((d) => "wx-" + d).join(" ") : "";
    return n + (n ? " " : "") + f;
  }, [e, n]), o = (f) => {
    c && c(f);
  };
  return /* @__PURE__ */ R(
    "button",
    {
      title: r,
      className: `wx-2ZWgb4 wx-button ${a} ${t && !i ? "wx-icon" : ""}`,
      disabled: s,
      onClick: o,
      children: [
        t && /* @__PURE__ */ l("i", { className: "wx-2ZWgb4 " + t }),
        i || u || " "
      ]
    }
  );
}
function Ye({
  id: e,
  label: n = "",
  inputValue: t = "",
  value: s = !1,
  onChange: r,
  disabled: u = !1,
  style: i
}) {
  const c = z(e), [a, o] = J(s);
  return /* @__PURE__ */ R("div", { style: i, className: "wx-2IvefP wx-checkbox", children: [
    /* @__PURE__ */ l(
      "input",
      {
        type: "checkbox",
        id: c,
        disabled: u,
        className: "wx-2IvefP wx-check",
        checked: a,
        value: t,
        onChange: ({ target: d }) => {
          const p = d.checked;
          o(p), r && r({ value: p, inputValue: t });
        }
      }
    ),
    /* @__PURE__ */ R("label", { htmlFor: c, className: "wx-2IvefP wx-label", children: [
      /* @__PURE__ */ l("span", { className: "wx-2IvefP wx-before" }),
      n && /* @__PURE__ */ l("span", { className: "wx-2IvefP wx-after", children: n })
    ] })
  ] });
}
function Qt({
  options: e = [],
  value: n = [],
  type: t = "",
  onChange: s
}) {
  const r = D(
    (u) => {
      let i;
      u.value ? i = [...n, u.inputValue] : i = n.filter((c) => c !== u.inputValue), s && s({ value: i });
    },
    [n, s]
  );
  return /* @__PURE__ */ l(ue.Provider, { value: null, children: /* @__PURE__ */ l("div", { className: `wx-q8xwRD wx-checkboxgroup ${t && `wx-${t}`}`, children: e.map((u) => /* @__PURE__ */ l("div", { className: "wx-q8xwRD wx-item", children: /* @__PURE__ */ l(
    Ye,
    {
      label: u.label,
      inputValue: u.id,
      value: n.includes(u.id),
      onChange: r
    }
  ) }, u.id)) }) });
}
function nt({ theme: e = "", target: n, children: t }) {
  const s = S(null), r = S(null), [u, i] = B(null);
  s.current || (s.current = document.createElement("div"));
  const c = K(ie);
  return W(() => {
    i(
      n || rt(r.current) || pe.getTopNode(r.current)
    );
  }, [r.current]), /* @__PURE__ */ R(_, { children: [
    /* @__PURE__ */ l("span", { ref: r, style: { display: "none" } }),
    r.current && u ? tt(
      /* @__PURE__ */ l(
        "div",
        {
          className: `wx-3ZWsT0 wx-${e || c}-theme`,
          children: t
        }
      ),
      u
    ) : null
  ] });
}
function rt(e) {
  const n = pe.getTopNode(e);
  for (; e && e !== n && !e.getAttribute("data-wx-portal-root"); )
    e = e.parentNode;
  return e;
}
function lt({
  left: e = 0,
  top: n = 0,
  at: t = "bottom",
  parent: s = null,
  width: r = "auto",
  css: u = "",
  onCancel: i,
  children: c,
  trackScroll: a = !1
}) {
  const [o, f] = B(0), [d, p] = B(0), [v, x] = B("auto"), [w, h] = B(!1), m = S(null), g = S(null);
  function T(N) {
    return s && (r + "").indexOf("%") > -1 ? r.replace(/(\d+)%/, (k, M) => (M = M * s.offsetWidth / 100 + "px", r.replace(k, M))) : r && r !== "auto" ? r : N;
  }
  const b = D(() => {
    if (!m.current) return;
    const N = Ze(m.current, s, t, e, n);
    N && (f(N.x), p(N.y), x(T(N.width)));
  }, [m, s, t, e, n, r]);
  return me(() => {
    const N = (k) => {
      i && k.target !== g.current && m.current && !m.current.contains(k.target) && i(k);
    };
    return requestAnimationFrame(() => {
      b(), h(!0), a && m.current && (g.current = ze(m.current), g.current && g.current.addEventListener("scroll", N, !0));
    }), () => {
      a && g.current && g.current.removeEventListener("scroll", N, !0);
    };
  }, []), me(() => {
    b();
  }, [s, t, e, n]), W(() => {
    const N = (k) => {
      i && i(k);
    };
    if (m.current)
      return ve(m.current, N).destroy;
  }, [i]), /* @__PURE__ */ l(
    "div",
    {
      ref: m,
      className: ["wx-37M6Fj", "wx-popup", u].filter(Boolean).join(" "),
      style: {
        position: "absolute",
        visibility: w ? "visible" : "hidden",
        top: d + "px",
        left: o + "px",
        width: v
      },
      children: c
    }
  );
}
function ct({
  position: e = "bottom",
  align: n = "start",
  autoFit: t = !0,
  onCancel: s = null,
  width: r = "100%",
  css: u = "",
  children: i
}) {
  const c = S(null), [a, o] = B(e), [f, d] = B(n);
  return W(() => {
    if (t) {
      const p = c.current;
      if (p) {
        const v = p.getBoundingClientRect(), x = document.body.getBoundingClientRect();
        v.right >= x.right && d("end"), v.bottom >= x.bottom && o("top");
      }
    }
  }, []), W(() => {
    if (c.current) {
      const p = (v) => {
        s && s(v);
      };
      return ve(c.current, p).destroy;
    }
  }, [s]), /* @__PURE__ */ l(
    "div",
    {
      ref: c,
      className: [
        "wx-dropdown",
        `wx-${a}-${f}`,
        u || "",
        "wx-aaaVytZW"
      ].filter(Boolean).join(" "),
      style: { width: r },
      children: i
    }
  );
}
function le({
  position: e = "bottom",
  align: n = "start",
  autoFit: t = !0,
  inline: s = !1,
  onCancel: r,
  width: u = "100%",
  children: i,
  ...c
}) {
  const [a, o] = B(void 0), f = S(null), d = A(() => `${e}-${n}`, [e, n]);
  return me(() => {
    f.current && o(f.current.parentNode);
  }, []), /* @__PURE__ */ R(_, { children: [
    a && (s ? /* @__PURE__ */ l(
      ct,
      {
        onCancel: r,
        position: e,
        align: n,
        autoFit: t,
        width: u,
        ...c,
        children: i
      }
    ) : /* @__PURE__ */ l(nt, { children: /* @__PURE__ */ l(lt, { parent: a, at: d, onCancel: r, width: u, ...c, children: i }) })),
    /* @__PURE__ */ l("span", { ref: f, className: "wx-portal-node wx-32GZ52" })
  ] });
}
const st = [
  "#00a037",
  "#37a9ef",
  "#f5a623",
  "#ff4c3b",
  "#a0a0a0",
  "#000000",
  "#ffffff"
];
function qt({
  colors: e = st,
  value: n = "",
  onChange: t,
  id: s,
  clear: r = !1,
  placeholder: u = "",
  title: i = "",
  disabled: c = !1,
  error: a = !1,
  dropdown: o = {}
}) {
  const f = z(s), [d, p] = B(n), [v, x] = B(!1), w = D(
    (g, T) => {
      g.stopPropagation(), p(T), x(!1), t && t({ value: T });
    },
    [t]
  ), h = D(
    (g) => {
      g.stopPropagation(), p(""), t && t({ value: "" });
    },
    [t]
  ), m = D(() => {
    if (c) return !1;
    x(!0);
  }, [c]);
  return /* @__PURE__ */ R("div", { className: "wx-121TgJ wx-colorselect", onClick: m, children: [
    /* @__PURE__ */ l(
      "input",
      {
        className: "wx-121TgJ wx-input" + (a ? " wx-error" : "") + (v ? " wx-focus" : ""),
        title: i,
        value: d,
        readOnly: !0,
        id: f,
        placeholder: u,
        disabled: c
      }
    ),
    r && d && !c && /* @__PURE__ */ l("i", { className: "wx-121TgJ wx-clear wxi-close", onClick: h }),
    d ? /* @__PURE__ */ l(
      "div",
      {
        className: "wx-121TgJ wx-color wx-selected",
        style: { backgroundColor: d || "#00a037" }
      }
    ) : /* @__PURE__ */ l("div", { className: "wx-121TgJ wx-empty wx-selected" }),
    v && /* @__PURE__ */ l(le, { onCancel: () => x(!1), ...o, children: /* @__PURE__ */ R("div", { className: "wx-121TgJ wx-colors", children: [
      /* @__PURE__ */ l(
        "div",
        {
          className: "wx-121TgJ wx-empty",
          onClick: (g) => w(g, "")
        }
      ),
      e.map((g, T) => /* @__PURE__ */ l(
        "div",
        {
          className: "wx-121TgJ wx-color",
          style: { backgroundColor: g },
          onClick: (b) => w(b, g)
        },
        T
      ))
    ] }) })
  ] });
}
function be(e, n) {
  const t = e.firstChild, { moveBlockSlider: s, moveLineSlider: r } = n || {};
  function u(a) {
    i(a), t.focus(), window.addEventListener("mousemove", i), window.addEventListener("mouseup", c);
  }
  function i(a) {
    let o, f, d, p;
    const v = e.getBoundingClientRect();
    o = v.left, f = v.top, d = a.clientX - o, p = a.clientY - f, s && n.moveBlockSlider(d, p), r && n.moveLineSlider(d);
  }
  function c() {
    t.blur(), window.removeEventListener("mousemove", i), window.removeEventListener("mouseup", c);
  }
  return e.addEventListener("mousedown", u), {
    destroy() {
      e.removeEventListener("mousedown", u);
    }
  };
}
const se = {
  _toHex: [
    "0",
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
    "A",
    "B",
    "C",
    "D",
    "E",
    "F"
  ],
  toHex(e, n) {
    e = parseInt(e, 10);
    let t = "";
    for (; e > 0; )
      t = this._toHex[e % 16] + t, e = Math.floor(e / 16);
    for (; t.length < n; ) t = `0${t}`;
    return t;
  },
  rgbToHex(e) {
    let n = [];
    return typeof e == "string" ? e.replace(/[\d+.]+/g, (t) => n.push(parseFloat(t))) : Array.isArray(e) && (n = e), n[3] === 0 ? "" : n.slice(0, 3).map((t) => this.toHex(Math.floor(t), 2)).join("");
  },
  hexToDec(e) {
    return parseInt(e, 16);
  },
  // only handles the six-digit hex color format
  hexToRgb(e) {
    e = e.charAt(0) === "#" ? e.slice(1) : e;
    const n = e.match(/.{1,2}/g);
    let [t, s, r] = n;
    return t = this.hexToDec(t), s = this.hexToDec(s), r = this.hexToDec(r), [t, s, r];
  },
  hsvToRgb(e, n, t) {
    let s, r, u, i, c, a, o, f;
    switch (s = Math.floor(e / 60) % 6, r = e / 60 - s, u = t * (1 - n), i = t * (1 - r * n), c = t * (1 - (1 - r) * n), a = 0, o = 0, f = 0, s) {
      case 0:
        a = t, o = c, f = u;
        break;
      case 1:
        a = i, o = t, f = u;
        break;
      case 2:
        a = u, o = t, f = c;
        break;
      case 3:
        a = u, o = i, f = t;
        break;
      case 4:
        a = c, o = u, f = t;
        break;
      case 5:
        a = t, o = u, f = i;
        break;
    }
    return a = Math.floor(a * 255), o = Math.floor(o * 255), f = Math.floor(f * 255), [a, o, f];
  },
  rgbToHsv(e, n, t) {
    let s, r, u;
    e /= 255, n /= 255, t /= 255;
    let i = Math.min(e, n, t), c = Math.max(e, n, t);
    return r = 0, s = c === 0 ? 0 : 1 - i / c, u = c, c === i ? r = 0 : c === e && n >= t ? r = 60 * (n - t) / (c - i) + 0 : c === e && n < t ? r = 60 * (n - t) / (c - i) + 360 : c === n ? r = 60 * (t - e) / (c - i) + 120 : c === t && (r = 60 * (e - n) / (c - i) + 240), [r, s, u];
  },
  hexToHvs(e) {
    const n = this.hexToRgb(e);
    return this.rgbToHsv(...n);
  },
  hvsToHex(e, n, t) {
    const s = this.hsvToRgb(e, n, t);
    return "#" + this.rgbToHex(s).toUpperCase();
  }
}, at = (e) => /^([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$/.test(e);
function De(e) {
  return e ? (e = e.toString(16), e.charAt(0) === "#" && (e = e.slice(1)), at(e) ? (e.length === 3 && (e = e.slice(0, 3).split("").reduce((n, t) => n + t + t, "")), `#${e.toUpperCase()}`) : null) : null;
}
function X() {
  return Ae(We);
}
const Ve = "Block", Te = "Line";
function ot({
  value: e = "#65D3B3",
  button: n = !1,
  onChange: t
}) {
  const s = (K(q) || X()).getGroup("core"), r = S(null), u = S(null), [i, c] = B(), [a, o] = B(), [f, d] = B(), [p, v] = B(), [x, w] = J(e), [h, m] = B(e), g = S(null), T = S(null), b = A(() => De(x) || "#65D3B3", [x]), N = A(
    () => se.hvsToHex(f || 0, 1, 1),
    [f]
  ), k = D(
    (L, P) => {
      const { width: C, height: O } = r.current.getBoundingClientRect();
      P < 0 ? P = 0 : P > O && (P = O), L < 0 ? L = 0 : L > C && (L = C), c(P), o(L);
      let U, Y;
      const I = C / 100, G = O / 100;
      U = Math.ceil(L / I) / 100, Y = Math.ceil(Math.abs(P / G - 100)) / 100;
      const Q = se.hvsToHex(
        f,
        U,
        Y
      );
      w(Q), m(Q), t && t({ value: Q, input: !0 });
    },
    [b, f, t]
  ), M = D(
    (L) => {
      const P = u.current.getBoundingClientRect().width;
      L < 0 ? L = 0 : L > P && (L = P);
      const C = Math.round(L * 359 / P), O = Math.max(Math.min(C, 359), 0);
      d(O), v(L);
      let U, Y;
      [, U, Y] = se.hexToHvs(b);
      const I = se.hvsToHex(O, U, Y);
      w(I), m(I), t && t({ value: I, input: !0 });
    },
    [b, t]
  );
  g.current = k, T.current = M;
  const $ = D((L) => {
    const [P, C, O] = se.hexToHvs(L || b);
    if (d(P), r.current) {
      const { width: U, height: Y } = r.current.getBoundingClientRect();
      u.current && v(
        P * u.current.getBoundingClientRect().width / 359
      ), o(C * U), c(Math.abs(Y * (O - 1)));
    }
  }, [b]), V = D(
    ({ target: L }) => {
      const P = De(L.value);
      m(L.value), P && (w(P), t && t({ value: P, input: !0 }), $(P));
    },
    [t, $]
  ), F = D(
    (L) => {
      L.stopPropagation(), t && t({ value: b });
    },
    [b, t]
  ), j = D(
    (L) => {
      const P = L.target, C = P.getAttribute("data-slider") === Ve, O = P.getAttribute("data-slider") === Te;
      let U = window.getComputedStyle(P), Y = parseFloat(U.left), I = parseFloat(U.top);
      const G = L.code;
      if (C) {
        switch (G) {
          case "ArrowLeft": {
            Y--;
            break;
          }
          case "ArrowRight": {
            Y++;
            break;
          }
          case "ArrowDown": {
            I++;
            break;
          }
          case "ArrowUp": {
            I--;
            break;
          }
          default:
            return;
        }
        k(Y, I);
      }
      if (O) {
        if (G === "ArrowLeft" || G === "ArrowDown") Y--;
        else if (G === "ArrowRight" || G === "ArrowUp") Y++;
        else return;
        M(Y);
      }
      L.preventDefault();
    },
    [k, M]
  );
  W(() => {
    const L = [];
    return r.current && L.push(
      be(r.current, {
        moveBlockSlider: (P, C) => {
          g.current?.(P, C);
        }
      }).destroy
    ), u.current && L.push(
      be(u.current, {
        moveLineSlider: (P) => {
          T.current?.(P);
        }
      }).destroy
    ), requestAnimationFrame(() => $()), () => {
      L.forEach((P) => P());
    };
  }, []);
  const H = "wx-1yoKzq";
  return /* @__PURE__ */ R("div", { className: H + " wx-colorboard", children: [
    /* @__PURE__ */ l(
      "div",
      {
        className: H + " wx-color-block",
        style: { background: N },
        ref: r,
        children: /* @__PURE__ */ l(
          "div",
          {
            className: H + " wx-color-block-slider wx-slider",
            style: {
              background: b,
              top: i + "px",
              left: a + "px"
            },
            tabIndex: 0,
            "data-slider": Ve,
            onKeyDown: j
          }
        )
      }
    ),
    /* @__PURE__ */ l("div", { className: H + " wx-color-line", ref: u, children: /* @__PURE__ */ l(
      "div",
      {
        className: H + " wx-color-line-slider wx-slider",
        style: { background: N, left: p + "px" },
        tabIndex: 0,
        "data-slider": Te,
        onKeyDown: j
      }
    ) }),
    /* @__PURE__ */ R("div", { className: H + " wx-color-controls", children: [
      /* @__PURE__ */ l(
        "div",
        {
          className: H + " wx-color",
          style: { background: b }
        }
      ),
      /* @__PURE__ */ l(
        "input",
        {
          type: "text",
          className: H + " wx-text",
          value: h,
          onChange: V
        }
      )
    ] }),
    n && /* @__PURE__ */ l(re, { onClick: F, type: "secondary", children: s("select") })
  ] });
}
function Kt({
  value: e = "",
  id: n,
  placeholder: t = "",
  title: s = "",
  disabled: r = !1,
  error: u = !1,
  clear: i = !1,
  onChange: c,
  dropdown: a = {}
}) {
  const o = z(n);
  let [f, d] = J(e);
  const [p, v] = B(!1);
  function x() {
    if (r) return !1;
    v(!0);
  }
  function w(m) {
    if (m.input) return;
    v(!1);
    const g = m.value;
    d(g), c && c({ value: g });
  }
  function h(m) {
    m.stopPropagation(), d(""), c && c({ value: "" });
  }
  return /* @__PURE__ */ R("div", { className: "wx-mmT8WH wx-colorpicker", onClick: x, children: [
    /* @__PURE__ */ l(
      "input",
      {
        className: "wx-mmT8WH wx-input" + (u ? " wx-error" : "") + (p ? " wx-focus" : ""),
        title: s,
        value: f,
        readOnly: !0,
        id: o,
        placeholder: t,
        disabled: r
      }
    ),
    /* @__PURE__ */ l("div", { className: "wx-mmT8WH wx-color", style: { background: f } }),
    i && !r && f && /* @__PURE__ */ l("i", { className: "wx-mmT8WH wxi-close", onClick: h }),
    p && /* @__PURE__ */ l(le, { onCancel: () => v(!1), ...a, children: /* @__PURE__ */ l(ot, { value: f, button: "true", onChange: w }) })
  ] });
}
function it() {
  let e = null, n = !1, t = !1, s, r, u, i;
  const c = (v, x, w, h, m) => {
    s = v, r = x, u = w, i = h, t = m;
  }, a = (v) => {
    e = v, n = e !== null, u(e);
  }, o = (v, x) => {
    if (!t && v !== null && s) {
      const w = s.querySelectorAll(".wx-list > .wx-item")[v];
      w && (w.scrollIntoView({ block: "nearest" }), x && x.preventDefault());
    }
  }, f = (v, x) => {
    const w = v === null ? null : Math.max(0, Math.min(e + v, r.length - 1));
    w !== e && (a(w), s ? o(w, x) : requestAnimationFrame(() => o(w, x)));
  };
  return { move: (v) => {
    const x = Xe(v), w = r.findIndex((h) => h.id === x);
    w !== e && a(w);
  }, keydown: (v, x) => {
    switch (v.code) {
      case "Enter":
        n ? i() : a(0);
        break;
      case "Space":
        n || a(0);
        break;
      case "Escape":
        u(e = null);
        break;
      case "Tab":
        u(e = null);
        break;
      case "ArrowDown":
        f(n ? 1 : x || 0, v);
        break;
      case "ArrowUp":
        f(n ? -1 : x || 0, v);
        break;
    }
  }, init: c, navigate: f };
}
const Re = 10, ut = Ue(function({ data: n, checkboxes: t, children: s, value: r }) {
  return /* @__PURE__ */ R(_, { children: [
    t && /* @__PURE__ */ l(
      Ye,
      {
        id: n.id,
        style: { marginRight: "8px", pointerEvents: "none" },
        value: r && r.includes(n.id)
      }
    ),
    s ? ae(s, { option: n }) : n.label
  ] });
});
function $e({
  visibleItems: e,
  visibleRange: n,
  navIndex: t,
  checkboxes: s,
  children: r,
  value: u,
  measureRefCallback: i
}) {
  return e.map((c, a) => {
    const o = n.start + a;
    return /* @__PURE__ */ l(
      "div",
      {
        ref: a === 0 ? i : void 0,
        className: `wx-233fr7 wx-item ${o === t ? "wx-focus" : ""}`,
        "data-id": c.id,
        children: /* @__PURE__ */ l(
          ut,
          {
            data: c,
            checkboxes: s,
            children: r,
            value: u
          }
        )
      },
      c.id
    );
  });
}
function ge({
  items: e = [],
  children: n,
  onSelect: t,
  onReady: s,
  virtualized: r = !1,
  checkboxes: u,
  multiselect: i,
  value: c,
  ...a
}) {
  const o = S(), f = S(!1), d = S(it()), [p, v] = B(null), x = S(p), [w, h] = B(0), [m, g] = B(24), T = S(!1), b = (K(q) || X()).getGroup("core"), N = S(e);
  W(() => {
    N.current = e;
  }, [e]);
  const k = S(c);
  W(() => {
    k.current = c;
  }, [c]);
  const M = S(0);
  W(() => {
    M.current = m;
  }, [m]);
  const $ = D((y) => {
    y && y.stopPropagation();
    const E = N.current, Z = k.current, ce = x.current;
    let ee;
    const te = E[ce]?.id;
    i ? Z && Z.includes(te) ? ee = Z.filter((we) => we !== te) : ee = [...Z || [], te] : ee = te, t && t({ id: ee });
  }, [i, t]), V = S($);
  W(() => {
    V.current = $;
  }, [$]);
  const F = D((y) => {
    if (o.current) {
      let E = y * M.current - o.current.clientHeight + M.current;
      h(E), o.current.scrollTop = E;
    }
  }, [M.current]), j = D(
    (y) => {
      o.current = y, r && (y && !f.current && x.current !== null && requestAnimationFrame(() => {
        o.current && (F(x.current), f.current = !0);
      }), y || (f.current = !1));
    },
    [F]
  ), H = D((y) => {
    if (!y || T.current) return;
    const E = y.getBoundingClientRect().height;
    E && (g(E), M.current = E, T.current = !0);
  }, []);
  W(() => {
    d.current.init(
      o.current,
      e,
      (y) => {
        v(y), x.current = y;
      },
      (y) => V.current(y),
      r
    );
  }, [e, o.current, r]), W(() => {
    const y = (Z, ce) => {
      d.current.navigate(Z, ce), F(x.current);
    }, E = (Z, ce) => {
      const ee = x.current, te = N.current;
      if (ee !== 0 && ee !== te.length - 1 && (Z.key === "ArrowDown" || Z.key === "ArrowUp") && Z.preventDefault(), d.current.keydown(Z, ce), o.current) {
        const we = o.current.scrollTop <= x.current * M.current, Ge = o.current.scrollTop + o.current.clientHeight >= x.current * M.current + M.current;
        we ? Ge || F(x.current) : (o.current.scrollTop = x.current * M.current, h(o.current.scrollTop));
      }
    };
    s && s({
      navigate: r ? y : d.current.navigate,
      keydown: r ? E : d.current.keydown,
      move: d.current.move
    });
  }, []);
  const L = S(0), P = S(e);
  P.current !== e && (P.current = e, L.current += 1);
  const C = D(() => {
    d.current.navigate(null);
  }, [d]), O = D((y) => {
    r && h(y.target.scrollTop);
  }, [r]), U = A(
    () => Math.ceil((o.current?.clientHeight || 0) / m),
    [m, w]
  ), Y = A(() => {
    if (!r) return { start: 0, end: e.length };
    if (!e.length) return { start: 0, end: 0 };
    const y = Math.floor(w / m), E = y + U;
    return {
      start: Math.max(0, y - Re),
      end: Math.min(e.length, E + Re)
    };
  }, [r, e, w, m, U]), I = A(() => {
    if (!r) return e;
    const { start: y, end: E } = Y;
    return e.slice(y, E).map((Z) => ({
      ...Z
    }));
  }, [r, e, Y]), G = Y.start * m, Q = e.length * m;
  return W(() => {
    p === null && (f.current = !1);
  }, [p]), p === null ? null : /* @__PURE__ */ l(le, { onCancel: C, ...a, children: /* @__PURE__ */ l(
    "div",
    {
      className: "wx-233fr7 wx-list",
      ref: j,
      onClick: $,
      onMouseMove: d.current.move,
      onScroll: O,
      children: e.length ? r ? /* @__PURE__ */ l(
        "div",
        {
          className: "wx-233fr7 wx-list-wrapper",
          style: { height: `${Q}px` },
          children: /* @__PURE__ */ l(
            "div",
            {
              className: "wx-233fr7 wx-list-content",
              style: { transform: `translateY(${G}px)` },
              children: /* @__PURE__ */ l(
                $e,
                {
                  visibleRange: Y,
                  visibleItems: I,
                  navIndex: p,
                  measureRefCallback: H,
                  checkboxes: u,
                  children: n,
                  value: c
                }
              )
            }
          )
        }
      ) : /* @__PURE__ */ l(
        $e,
        {
          visibleRange: Y,
          visibleItems: I,
          navIndex: p,
          measureRefCallback: H,
          checkboxes: u,
          children: n,
          value: c
        }
      ) : /* @__PURE__ */ l("div", { className: "wx-233fr7 wx-no-data", children: b("No data") })
    }
  ) }, L.current);
}
function Zt({
  value: e = "",
  id: n,
  options: t = [],
  textOptions: s = null,
  textField: r = "label",
  placeholder: u = "",
  title: i = "",
  disabled: c = !1,
  error: a = !1,
  clear: o = !1,
  children: f,
  onChange: d,
  dropdown: p = {}
}) {
  const v = z(n), x = S(null), w = S(null), [h, m] = J(e), [g, T] = B(!1), [b, N] = B(""), k = S(null), M = S(!1), $ = A(() => {
    if (g) return b;
    if (h || h === 0) {
      const I = (s || t).find((G) => G.id === h);
      if (I) return I[r];
    }
    return "";
  }, [g, b, h, s, t, r]), V = A(() => !$ || !g ? t : t.filter(
    (I) => I[r].toLowerCase().includes($.toLowerCase())
  ), [$, g, t, r]), F = D(
    () => V.findIndex((I) => I.id === h),
    [V, h]
  ), j = D((I) => {
    x.current = I.navigate, w.current = I.keydown;
  }, []), H = D(
    (I, G) => {
      if (I || I === 0) {
        let Q = t.find((y) => y.id === I);
        if (T(!1), G && x.current(null), Q && h !== Q.id) {
          const y = Q.id;
          m(y), d && d({ value: y });
        }
      }
      !M.current && G && k.current.focus();
    },
    [t, h, d]
  ), L = D(
    ({ id: I }) => {
      H(I, !0);
    },
    [H]
  ), P = D(
    (I) => {
      I && I.stopPropagation(), m(""), T(!1), d && d({ value: "" });
    },
    [d]
  ), C = D(
    (I) => {
      if (!t.length) return;
      if (I === "" && o) {
        P();
        return;
      }
      let G = t.find((y) => y[r] === I);
      G || (G = t.find(
        (y) => y[r].toLowerCase().includes(I.toLowerCase())
      ));
      const Q = G ? G.id : h || t[0].id;
      H(Q, !1);
    },
    [t, r, o, h, H, P]
  ), O = D(() => {
    N(k.current.value), T(!0), V.length && x.current(0);
  }, [V.length, x]), U = D(() => {
    M.current = !0;
  }, []), Y = D(() => {
    M.current = !1, setTimeout(() => {
      M.current || C($);
    }, 200);
  }, [C, $]);
  return /* @__PURE__ */ R(
    "div",
    {
      className: "wx-1j11Jk wx-combo",
      onClick: () => x.current(F()),
      onKeyDown: (I) => w.current(I, F()),
      title: i,
      children: [
        /* @__PURE__ */ l(
          "input",
          {
            className: "wx-1j11Jk wx-input " + (a ? "wx-error" : ""),
            id: v,
            ref: k,
            value: $,
            disabled: c,
            placeholder: u,
            onFocus: U,
            onBlur: Y,
            onInput: O
          }
        ),
        o && !c && h ? /* @__PURE__ */ l("i", { className: "wx-1j11Jk wx-icon wxi-close", onClick: P }) : /* @__PURE__ */ l("i", { className: "wx-1j11Jk wx-icon wxi-angle-down" }),
        !c && /* @__PURE__ */ l(ge, { items: V, onReady: j, onSelect: L, ...p, children: ({ option: I }) => /* @__PURE__ */ l(_, { children: f ? f({ option: I }) : I[r] }) })
      ]
    }
  );
}
function Ne({
  value: e = "",
  id: n,
  readonly: t = !1,
  focus: s = !1,
  select: r = !1,
  type: u = "text",
  placeholder: i = "",
  disabled: c = !1,
  error: a = !1,
  inputStyle: o = {},
  title: f = "",
  css: d = "",
  icon: p = "",
  clear: v = !1,
  onChange: x
}) {
  const w = z(n), [h, m] = J(e), g = S(null), T = A(
    () => p && d.indexOf("wx-icon-left") === -1 ? "wx-icon-right " + d : d,
    [p, d]
  ), b = A(
    () => p && d.indexOf("wx-icon-left") !== -1,
    [p, d]
  );
  W(() => {
    const V = setTimeout(() => {
      s && g.current && g.current.focus(), r && g.current && g.current.select();
    }, 1);
    return () => clearTimeout(V);
  }, [s, r]);
  const N = D(
    (V) => {
      const F = V.target.value;
      m(F), x && x({ value: F, input: !0 });
    },
    [x]
  ), k = D(
    (V) => x && x({ value: V.target.value }),
    [x]
  );
  function M(V) {
    V.stopPropagation(), m(""), x && x({ value: "" });
  }
  let $ = u;
  return u !== "password" && u !== "number" && ($ = "text"), W(() => {
    const V = k, F = g.current;
    return F.addEventListener("change", V), () => {
      F && F.removeEventListener("change", V);
    };
  }, [k]), /* @__PURE__ */ R(
    "div",
    {
      className: `wx-hQ64J4 wx-text ${T} ${a ? "wx-error" : ""} ${c ? "wx-disabled" : ""} ${v ? "wx-clear" : ""}`,
      children: [
        /* @__PURE__ */ l(
          "input",
          {
            className: "wx-hQ64J4 wx-input",
            ref: g,
            id: w,
            readOnly: t,
            disabled: c,
            placeholder: i,
            type: $,
            style: o,
            title: f,
            value: h,
            onInput: N
          }
        ),
        v && !c && h ? /* @__PURE__ */ R(_, { children: [
          /* @__PURE__ */ l("i", { className: "wx-hQ64J4 wx-icon wxi-close", onClick: M }),
          b && /* @__PURE__ */ l("i", { className: `wx-hQ64J4 wx-icon ${p}` })
        ] }) : p ? /* @__PURE__ */ l("i", { className: `wx-hQ64J4 wx-icon ${p}` }) : null
      ]
    }
  );
}
function dt({ date: e, type: n, part: t, onShift: s }) {
  const { calendar: r, formats: u } = K(q).getRaw(), i = e.getFullYear(), c = A(() => {
    switch (n) {
      case "month":
        return ne(u.monthYearFormat, r)(e);
      case "year":
        return ne(u.yearFormat, r)(e);
      case "duodecade": {
        const { start: o, end: f } = Be(i), d = ne(u.yearFormat, r);
        return `${d(new Date(o, 0, 1))} - ${d(new Date(f, 11, 31))}`;
      }
      default:
        return "";
    }
  }, [e, n, i, r, u]);
  function a() {
    s && s({ diff: 0, type: n });
  }
  return /* @__PURE__ */ R("div", { className: "wx-8HQVQV wx-header", children: [
    t !== "right" ? /* @__PURE__ */ l(
      "i",
      {
        className: "wx-8HQVQV wx-pager wxi-angle-left",
        onClick: () => s && s({ diff: -1, type: n })
      }
    ) : /* @__PURE__ */ l("span", { className: "wx-8HQVQV wx-spacer" }),
    /* @__PURE__ */ l("span", { className: "wx-8HQVQV wx-label", onClick: a, children: c }),
    t !== "left" ? /* @__PURE__ */ l(
      "i",
      {
        className: "wx-8HQVQV wx-pager wxi-angle-right",
        onClick: () => s && s({ diff: 1, type: n })
      }
    ) : /* @__PURE__ */ l("span", { className: "wx-8HQVQV wx-spacer" })
  ] });
}
function ke({ onClick: e, children: n }) {
  return /* @__PURE__ */ l("button", { className: "wx-3s8W4d wx-button", onClick: e, children: n });
}
function ft({
  value: e,
  current: n,
  part: t = "",
  markers: s = null,
  onCancel: r,
  onChange: u
}) {
  const i = (K(q) || X()).getRaw().calendar, c = (i.weekStart || 7) % 7, a = i.dayShort.slice(c).concat(i.dayShort.slice(0, c)), o = (b, N, k) => new Date(
    b.getFullYear(),
    b.getMonth() + (N || 0),
    b.getDate() + (k || 0)
  );
  let f = t !== "normal";
  function d(b) {
    const N = b.getDay();
    return N === 0 || N === 6;
  }
  function p() {
    const b = o(n, 0, 1 - n.getDate());
    return b.setDate(b.getDate() - (b.getDay() - (c - 7)) % 7), b;
  }
  function v() {
    const b = o(n, 1, -n.getDate());
    return b.setDate(b.getDate() + (6 - b.getDay() + c) % 7), b;
  }
  const x = S(0);
  function w(b, N) {
    N.timeStamp !== x.current && (x.current = N.timeStamp, N.stopPropagation(), u && u(new Date(b)), r && r());
  }
  const h = A(() => t == "normal" ? [e ? o(e).valueOf() : 0] : e ? [
    e.start ? o(e.start).valueOf() : 0,
    e.end ? o(e.end).valueOf() : 0
  ] : [0, 0], [t, e]), m = A(() => {
    const b = p(), N = v(), k = n.getMonth();
    let M = [];
    for (let $ = b; $ <= N; $.setDate($.getDate() + 1)) {
      const V = {
        day: $.getDate(),
        in: $.getMonth() === k,
        date: $.valueOf()
      };
      let F = "";
      if (F += V.in ? "" : " wx-inactive", F += h.indexOf(V.date) > -1 ? " wx-selected" : "", f) {
        const j = V.date == h[0], H = V.date == h[1];
        j && !H ? F += " wx-left" : H && !j && (F += " wx-right"), V.date > h[0] && V.date < h[1] && (F += " wx-inrange");
      }
      if (F += d($) ? " wx-weekend" : "", s) {
        const j = s($);
        j && (F += " " + j);
      }
      M.push({ ...V, css: F });
    }
    return M;
  }, [n, h, f, s]), g = S(null);
  let T = S({});
  return T.current.click = w, W(() => {
    Ee(g.current, T.current);
  }, []), /* @__PURE__ */ R("div", { children: [
    /* @__PURE__ */ l("div", { className: "wx-398RBS wx-weekdays", children: a.map((b) => /* @__PURE__ */ l("div", { className: "wx-398RBS wx-weekday", children: b }, b)) }),
    /* @__PURE__ */ l("div", { className: "wx-398RBS wx-days", ref: g, children: m.map((b) => /* @__PURE__ */ l(
      "div",
      {
        className: `wx-398RBS wx-day ${b.css} ${b.in ? "" : "wx-out"}`,
        "data-id": b.date,
        children: b.day
      },
      b.date
    )) })
  ] });
}
function wt({
  value: e,
  current: n,
  part: t,
  onCancel: s,
  onChange: r,
  onShift: u
}) {
  const i = S(/* @__PURE__ */ new Date()), [c, a] = J(e || i.current), [o, f] = J(n || i.current), d = K(q).getRaw().calendar, p = d.monthShort || [], v = A(() => o.getMonth(), [o]), x = D(
    (m, g) => {
      if (m != null) {
        g.stopPropagation();
        const T = new Date(o);
        T.setMonth(m), f(T), u && u({ current: T });
      }
      t === "normal" && a(new Date(o)), s && s();
    },
    [o, t, u, s]
  ), w = D(() => {
    const m = new Date(je(c, t) || o);
    m.setMonth(o.getMonth()), m.setFullYear(o.getFullYear()), r && r(m);
  }, [c, o, t, r]), h = D(
    (m) => {
      const g = m.target.closest("[data-id]");
      if (g) {
        const T = parseInt(g.getAttribute("data-id"), 10);
        x(T, m);
      }
    },
    [x]
  );
  return /* @__PURE__ */ R(_, { children: [
    /* @__PURE__ */ l("div", { className: "wx-34U8T8 wx-months", onClick: h, children: p.map((m, g) => /* @__PURE__ */ l(
      "div",
      {
        className: "wx-34U8T8 wx-month" + (v === g ? " wx-current" : ""),
        "data-id": g,
        children: m
      },
      g
    )) }),
    /* @__PURE__ */ l("div", { className: "wx-34U8T8 wx-buttons", children: /* @__PURE__ */ l(ke, { onClick: w, children: d.done }) })
  ] });
}
const xe = "wx-1XEF33", xt = ({ value: e, current: n, onCancel: t, onChange: s, onShift: r, part: u }) => {
  const i = K(q).getRaw().calendar, [c, a] = J(n), [o, f] = J(e), d = A(() => c.getFullYear(), [c]), p = A(() => {
    const { start: m, end: g } = Be(d), T = [];
    for (let b = m; b <= g; ++b)
      T.push(b);
    return T;
  }, [d]), v = {
    click: x
  };
  function x(m, g) {
    if (m) {
      g.stopPropagation();
      const T = new Date(c);
      T.setFullYear(m), a(T), r && r({ current: T });
    }
    u === "normal" && f(new Date(c)), t && t();
  }
  function w() {
    const m = new Date(je(o, u) || c);
    m.setFullYear(c.getFullYear()), s && s(m);
  }
  const h = S(null);
  return W(() => {
    h.current && Ee(h.current, v);
  }, []), /* @__PURE__ */ R(_, { children: [
    /* @__PURE__ */ l("div", { className: xe + " wx-years", ref: h, children: p.map((m, g) => /* @__PURE__ */ l(
      "div",
      {
        className: xe + ` wx-year ${d == m ? "wx-current" : ""} ${g === 0 ? "wx-prev-decade" : ""} ${g === 11 ? "wx-next-decade" : ""}`,
        "data-id": m,
        children: m
      },
      g
    )) }),
    /* @__PURE__ */ l("div", { className: xe + " wx-buttons", children: /* @__PURE__ */ l(ke, { onClick: w, children: i.done }) })
  ] });
}, Me = {
  month: {
    component: ft,
    next: ht,
    prev: mt
  },
  year: {
    component: wt,
    next: vt,
    prev: pt
  },
  duodecade: {
    component: xt,
    next: Nt,
    prev: gt
  }
};
function mt(e) {
  return e = new Date(e), e.setMonth(e.getMonth() - 1), e;
}
function ht(e) {
  return e = new Date(e), e.setMonth(e.getMonth() + 1), e;
}
function pt(e) {
  return e = new Date(e), e.setFullYear(e.getFullYear() - 1), e;
}
function vt(e) {
  return e = new Date(e), e.setFullYear(e.getFullYear() + 1), e;
}
function gt(e) {
  return e = new Date(e), e.setFullYear(e.getFullYear() - 10), e;
}
function Nt(e) {
  return e = new Date(e), e.setFullYear(e.getFullYear() + 10), e;
}
function je(e, n) {
  let t;
  if (n === "normal") t = e;
  else {
    const { start: s, end: r } = e;
    n === "left" ? t = s : n == "right" ? t = r : t = s && r;
  }
  return t;
}
const kt = ["clear", "today"];
function yt(e) {
  if (e === "done") return -1;
  if (e === "clear") return null;
  if (e === "today") return /* @__PURE__ */ new Date();
}
function de({
  value: e,
  current: n,
  onCurrentChange: t,
  part: s = "normal",
  markers: r = null,
  buttons: u,
  onShift: i,
  onChange: c
}) {
  const a = K(q).getGroup("calendar"), [o, f] = B("month"), d = Array.isArray(u) ? u : u ? kt : [], p = (m, g) => {
    m.preventDefault(), c && c({ value: g });
  }, v = () => {
    o === "duodecade" ? f("year") : o === "year" && f("month");
  }, x = (m) => {
    const { diff: g, current: T } = m;
    if (g === 0) {
      o === "month" ? f("year") : o === "year" && f("duodecade");
      return;
    }
    if (g) {
      const b = Me[o];
      t(g > 0 ? b.next(n) : b.prev(n));
    } else T && t(T);
    i && i();
  }, w = (m) => {
    f("month"), c && c({ select: !0, value: m });
  }, h = A(() => Me[o].component, [o]);
  return /* @__PURE__ */ l(
    "div",
    {
      className: `wx-2Gr4AS wx-calendar ${s !== "normal" && s !== "both" ? "wx-part" : ""}`,
      children: /* @__PURE__ */ R("div", { className: "wx-2Gr4AS wx-wrap", children: [
        /* @__PURE__ */ l(dt, { date: n, part: s, type: o, onShift: x }),
        /* @__PURE__ */ R("div", { children: [
          /* @__PURE__ */ l(
            h,
            {
              value: e,
              current: n,
              onCurrentChange: t,
              part: s,
              markers: r,
              onCancel: v,
              onChange: w,
              onShift: x
            }
          ),
          o === "month" && d.length > 0 && /* @__PURE__ */ l("div", { className: "wx-2Gr4AS wx-buttons", children: d.map((m) => /* @__PURE__ */ l("div", { className: "wx-2Gr4AS wx-button-item", children: /* @__PURE__ */ l(
            ke,
            {
              onClick: (g) => p(g, yt(m)),
              children: a(m)
            }
          ) }, m)) })
        ] })
      ] })
    }
  );
}
function he(e) {
  let { words: n = null, optional: t = !1, children: s } = e, r = K(q);
  const u = A(() => {
    let i = r;
    return (!i || !i.extend) && (i = Ae(We)), n !== null && (i = i.extend(n, t)), i;
  }, [n, t, r]);
  return /* @__PURE__ */ l(q.Provider, { value: u, children: s });
}
function Ie(e, n, t, s) {
  if (!e || s) {
    const r = n ? new Date(n) : /* @__PURE__ */ new Date();
    r.setDate(1), t(r);
  } else if (e.getDate() !== 1) {
    const r = new Date(e);
    r.setDate(1), t(r);
  }
}
const bt = ["clear", "today"];
function Dt({
  value: e,
  current: n,
  markers: t = null,
  buttons: s = bt,
  onChange: r
}) {
  const [u, i] = J(e), [c, a] = J(n);
  W(() => {
    Ie(c, u, a, !1);
  }, [u, c]);
  const o = D(
    (d) => {
      const p = d.value;
      p ? (i(new Date(p)), Ie(c, new Date(p), a, !0)) : i(null), r && r({ value: p ? new Date(p) : null });
    },
    [r, c]
  ), f = D(
    (d) => {
      a(d);
    },
    [a]
  );
  return c ? /* @__PURE__ */ l(he, { children: /* @__PURE__ */ l(
    de,
    {
      value: u,
      current: c,
      markers: t,
      buttons: s,
      onChange: o,
      onCurrentChange: f
    }
  ) }) : null;
}
function ye(e) {
  return e || (e = {}), e.width || (e.width = "unset"), e;
}
const Vt = ["clear", "today"];
function zt({
  value: e,
  id: n,
  disabled: t = !1,
  error: s = !1,
  placeholder: r = "",
  format: u = "",
  buttons: i = Vt,
  css: c = "",
  title: a = "",
  editable: o = !1,
  clear: f = !1,
  onChange: d,
  dropdown: p = {}
}) {
  const { calendar: v, formats: x } = (K(q) || X()).getRaw(), w = u || x?.dateFormat;
  let h = typeof w == "function" ? w : ne(w, v);
  const [m, g] = B(e), [T, b] = B(!1);
  W(() => {
    g(e);
  }, [e]);
  function N() {
    b(!1);
  }
  function k(V) {
    const F = V === m || V && m && V.valueOf() === m.valueOf() || !V && !m;
    g(V), F || d && d({ value: V }), setTimeout(N, 1);
  }
  const M = A(
    () => m ? h(m) : "",
    [m, h]
  );
  function $({ value: V, input: F }) {
    if (!o && !f || F) return;
    let j = typeof o == "function" ? o(V) : V ? new Date(V) : null;
    j = isNaN(j) ? m || null : j || null, k(j);
  }
  return W(() => {
    const V = N;
    return window.addEventListener("scroll", V), () => window.removeEventListener("scroll", V);
  }, []), /* @__PURE__ */ R("div", { className: "wx-1lKOFG wx-datepicker", onClick: () => b(!0), children: [
    /* @__PURE__ */ l(
      Ne,
      {
        css: c,
        title: a,
        value: M,
        id: n,
        readonly: !o,
        disabled: t,
        error: s,
        placeholder: r,
        onInput: N,
        onChange: $,
        icon: "wxi-calendar",
        inputStyle: {
          cursor: "pointer",
          width: "100%",
          paddingRight: "calc(var(--wx-input-icon-size) + var(--wx-input-icon-indent) * 2)"
        },
        clear: f
      }
    ),
    T && !t && /* @__PURE__ */ l(
      le,
      {
        onCancel: N,
        ...ye(p),
        children: /* @__PURE__ */ l(
          Dt,
          {
            buttons: i,
            value: m,
            onChange: (V) => k(V.value)
          }
        )
      }
    )
  ] });
}
const Tt = ["clear", "today"], Rt = ({
  start: e,
  end: n,
  current: t,
  months: s = 2,
  markers: r = null,
  buttons: u = Tt,
  onChange: i
}) => {
  const [c, a] = J(e), [o, f] = J(n), [d, p] = B(), [v, x] = B(), w = D((N, k, M) => {
    const $ = new Date(N);
    return $.setMonth($.getMonth() + k), M && $.valueOf() === M.valueOf() ? M : $;
  }, []), h = D(
    (N) => {
      p(() => {
        const k = new Date(N);
        return k.setDate(1), x(w(k, 1)), k;
      });
    },
    [w]
  ), m = D(
    (N) => {
      x(() => {
        const k = new Date(N);
        return k.setDate(1), p(w(k, -1)), k;
      });
    },
    [w]
  );
  W(() => {
    const N = c;
    d || h(N ? new Date(N) : t || /* @__PURE__ */ new Date());
  }, [c, t, d, h]);
  const g = D(
    (N) => {
      const k = N.value;
      let M = c, $ = o;
      const V = k === -1;
      return V || (N.select ? !c || o ? (a(M = k), f($ = null)) : c > k ? (f($ = c), a(M = k)) : f($ = k) : k ? (a(M = new Date(k)), f($ = new Date(k))) : (a(M = null), f($ = null))), (V || !u.includes("done")) && i && i({ start: M, end: $ }), [M, $];
    },
    [c, o, u, i]
  ), T = D(
    (N) => {
      const [k] = g(N);
      k && h(new Date(k));
    },
    [h, c, o]
  ), b = D(
    (N) => {
      const [, k] = g(N);
      k && m(new Date(k));
    },
    [m, c, o]
  );
  return !d || !v ? null : s === 1 ? /* @__PURE__ */ l(he, { children: /* @__PURE__ */ l(
    de,
    {
      value: { start: c, end: o },
      current: d,
      markers: r,
      buttons: u,
      part: "both",
      onChange: T,
      onCurrentChange: h
    }
  ) }) : /* @__PURE__ */ l(he, { children: /* @__PURE__ */ R("div", { className: "wx-p2jCaW wx-rangecalendar", children: [
    /* @__PURE__ */ l("div", { className: "wx-p2jCaW wx-half", children: /* @__PURE__ */ l(
      de,
      {
        value: { start: c, end: o },
        current: d,
        markers: r,
        buttons: !1,
        part: "left",
        onChange: T,
        onCurrentChange: h
      }
    ) }),
    /* @__PURE__ */ l("div", { className: "wx-p2jCaW wx-half", children: /* @__PURE__ */ l(
      de,
      {
        value: { start: c, end: o },
        current: v,
        markers: r,
        buttons: u,
        part: "right",
        onChange: b,
        onCurrentChange: m
      }
    ) })
  ] }) });
}, $t = ["clear", "today"], Xt = ({
  value: e,
  onChange: n,
  id: t,
  disabled: s = !1,
  error: r = !1,
  placeholder: u = "",
  css: i = "",
  title: c = "",
  format: a = "",
  months: o = 2,
  buttons: f = $t,
  editable: d = !1,
  clear: p = !1,
  dropdown: v
}) => {
  const [x, w] = J(e), { calendar: h, formats: m } = (K(q) || X()).getRaw(), g = a || m?.dateFormat, T = typeof g == "function" ? g : ne(g, h), [b, N] = B(!1), k = D(() => {
    N(!1);
  }, []), M = A(() => x ? x.start ? T(x.start) + (x.end ? ` - ${T(x.end)}` : "") : T(x) : "", [x, T]), $ = A(() => x && x.start || null, [x]), V = A(() => x && x.end || null, [x]), F = D(
    (H) => {
      const L = H.start || H.end ? { start: H.start, end: H.end } : null;
      w && w(L), (H.start && H.end || !H.start && !H.end) && (n && n({ value: L }), setTimeout(k, 1));
    },
    [n, k, w]
  ), j = D(
    (H) => {
      if (!d && !p) return;
      const { value: L, input: P } = H;
      if (P) return;
      const [C, O] = L.split(" -").map((U, Y) => {
        const I = U.trim();
        let G = typeof d == "function" ? d(I) : I ? new Date(I) : null, Q = Y === 0 ? $ : V;
        return isNaN(G) ? Q || null : G || null;
      });
      F({ start: C, end: O });
    },
    [d, p, $, V, F]
  );
  return W(() => {
    const H = () => {
      k();
    };
    return window.addEventListener("scroll", H), () => {
      window.removeEventListener("scroll", H);
    };
  }, [k]), /* @__PURE__ */ R(
    "div",
    {
      className: `wx-1b0WDQ wx-daterangepicker ${s ? "wx-disabled" : ""} ${r ? "wx-error" : ""}`,
      onClick: () => N(!0),
      children: [
        /* @__PURE__ */ l(
          Ne,
          {
            css: i,
            title: c,
            value: M,
            id: t,
            readonly: !d,
            disabled: s,
            placeholder: u,
            error: r,
            onChange: j,
            icon: "wxi-calendar",
            inputStyle: {
              cursor: "pointer",
              width: "100%",
              paddingRight: "calc(var(--wx-input-icon-size) + var(--wx-input-icon-indent) * 2)"
            },
            clear: p
          }
        ),
        b && !s && /* @__PURE__ */ l(
          le,
          {
            onCancel: k,
            ...ye(v),
            children: /* @__PURE__ */ l(
              Rt,
              {
                onCancel: k,
                buttons: f,
                start: $,
                end: V,
                months: o,
                onChange: F
              }
            )
          }
        )
      ]
    }
  );
};
function en({
  hotkey: e = null,
  toggleButton: n,
  children: t
}) {
  const s = S(null), [r, u] = B(!1), i = A(
    () => `wxi-${r ? "collapse" : "expand"}`,
    [r]
  ), c = S(null);
  return c.current = () => {
    const a = s.current;
    !r && a ? a.requestFullscreen() : r && document.exitFullscreen(), u((o) => !o);
  }, Ke({ [e]: () => c.current() }, s), W(() => {
    const a = () => {
      u(document.fullscreenElement === s.current);
    };
    return document.addEventListener("fullscreenchange", a), () => {
      document.removeEventListener("fullscreenchange", a);
    };
  }, []), /* @__PURE__ */ R("div", { className: "wx-13aa5W wx-fullscreen", ref: s, tabIndex: -1, children: [
    t,
    n ? n(c.current, r) : /* @__PURE__ */ l(re, { css: "wx-13aa5W wx-fullscreen-button", onClick: c.current, children: /* @__PURE__ */ l("i", { className: `wx-13aa5W ${i} wx-fullscreen-icon`, children: " " }) })
  ] });
}
function tn({ css: e = "", title: n = "", children: t, onClick: s }) {
  return t ? /* @__PURE__ */ l(
    "i",
    {
      title: n,
      role: "img",
      className: `wx-mdnST1 wx-icon ${e}`,
      onClick: s,
      children: t
    }
  ) : /* @__PURE__ */ l(
    "i",
    {
      title: n,
      className: `wx-mdnST1 wx-icon ${e}`,
      onClick: s
    }
  );
}
const Mt = [];
function nn({
  id: e,
  value: n = Mt,
  options: t = [],
  textOptions: s = null,
  textField: r = "label",
  placeholder: u = "",
  title: i = "",
  disabled: c = !1,
  error: a = !1,
  checkboxes: o = !1,
  onChange: f,
  children: d,
  dropdown: p = {}
}) {
  const v = z(e), x = S(null), w = S(null), [h, m] = J(n), [g, T] = B(""), [b, N] = B(!1), k = S(null), M = A(() => h ? (s || t).filter((C) => h.includes(C.id)) : [], [h, s, t]), $ = A(() => {
    const C = t;
    return g ? C.filter((O) => O[r].toLowerCase().includes(g.toLowerCase())) : C;
  }, [g, t, r]), V = D((C) => {
    x.current = C.navigate, w.current = C.keydown;
  }, []), F = D(() => {
    $.length ? x.current(0) : x.current(null);
  }, [$]), j = D(
    (C) => {
      const { id: O } = C;
      O && (m(O), f && f({ value: O }), k.current.focus());
    },
    [f]
  ), H = D(
    (C, O) => {
      O && O.stopPropagation();
      const U = h.filter((Y) => Y !== C);
      m(U), f && f({ value: U });
    },
    [h, f]
  ), L = D(
    () => h && h.length ? $.findIndex((C) => C.id === h[0]) : 0,
    [h, $]
  ), P = D(() => {
    c || (k.current.focus(), x.current(L()));
  }, [c, L]);
  return /* @__PURE__ */ R(
    "div",
    {
      title: i,
      className: `wx-12Wj21 wx-multicombo ${a ? "wx-error" : ""} ${c ? "wx-disabled" : ""} ${M.length ? "wx-not-empty" : ""} ${b && !c ? "wx-focus" : ""}`,
      onClick: P,
      onKeyDown: (C) => w.current(C, L()),
      children: [
        /* @__PURE__ */ R("div", { className: "wx-12Wj21 wx-wrapper", children: [
          /* @__PURE__ */ l("div", { className: "wx-12Wj21 wx-tags", children: M.map((C) => /* @__PURE__ */ R("div", { className: "wx-12Wj21 wx-tag", children: [
            d ? ae(d, { option: C }) : C[r],
            !c && /* @__PURE__ */ l(
              "i",
              {
                className: "wx-12Wj21 wx-icon wxi-close",
                onClick: (O) => H(C.id, O)
              }
            )
          ] }, C.id)) }),
          /* @__PURE__ */ R("div", { className: "wx-12Wj21 wx-select", children: [
            /* @__PURE__ */ l(
              "input",
              {
                className: "wx-12Wj21 wx-input",
                id: v,
                type: "text",
                value: g,
                onChange: (C) => T(C.target.value),
                onInput: F,
                placeholder: u,
                disabled: c,
                onFocus: () => N(!0),
                onBlur: () => N(!1),
                ref: k
              }
            ),
            /* @__PURE__ */ l("i", { className: "wx-12Wj21 wx-icon wxi-angle-down" })
          ] })
        ] }),
        !c && /* @__PURE__ */ l(
          ge,
          {
            items: $,
            multiselect: !0,
            onReady: V,
            onSelect: j,
            checkboxes: o,
            value: h,
            ...p,
            children: ({ option: C }) => /* @__PURE__ */ l(_, { children: d ? ae(d, { option: C }) : C[r] })
          }
        )
      ]
    }
  );
}
const rn = ({
  total: e = 0,
  pageSize: n = 20,
  value: t = 1,
  onChange: s
}) => {
  const [r, u] = J(n), [i, c] = J(t), a = (K(q) || X()).getGroup("core"), o = Math.ceil(e / r), f = (i - 1) * r, d = Math.min(i * r, e), p = (w, h) => {
    Number.isNaN(w) || (h || c(w), w = Math.max(1, Math.min(w, o)), h && c(w), w !== i && setTimeout(() => {
      s && s({
        value: w,
        from: (w - 1) * r,
        to: Math.min(w * r, e),
        pageSize: r
      });
    }));
  };
  function v(w) {
    switch (w) {
      case "first":
        p(1);
        break;
      case "prev":
        p(i - 1);
        break;
      case "next":
        p(i + 1);
        break;
      case "last":
        p(o);
        break;
    }
  }
  const x = (w) => {
    const h = w.target.value * 1;
    Number.isNaN(h) || (u(h), s && s({ value: i, from: f, to: d, pageSize: h }));
  };
  return /* @__PURE__ */ R("div", { className: "wx-35Np0p wx-pager", children: [
    /* @__PURE__ */ R("div", { className: "wx-35Np0p wx-left", children: [
      /* @__PURE__ */ R("span", { children: [
        a("Rows per page"),
        ": "
      ] }),
      /* @__PURE__ */ l(
        "input",
        {
          type: "number",
          min: 1,
          className: "wx-35Np0p wx-input",
          value: r,
          onChange: x
        }
      )
    ] }),
    /* @__PURE__ */ R("div", { className: "wx-35Np0p wx-center", children: [
      /* @__PURE__ */ l(
        "i",
        {
          className: "wx-35Np0p wx-icon wxi-angle-dbl-left" + (i === 1 ? " wx-disabled" : ""),
          onClick: () => v("first")
        }
      ),
      /* @__PURE__ */ l(
        "i",
        {
          className: "wx-35Np0p wx-icon wxi-angle-left" + (i === 1 ? " wx-disabled" : ""),
          onClick: () => v("prev")
        }
      ),
      /* @__PURE__ */ l(
        "input",
        {
          type: "text",
          value: i,
          className: "wx-35Np0p wx-input",
          onChange: (w) => p(w.target.value * 1),
          onBlur: (w) => p(w.target.value * 1, !0)
        }
      ),
      /* @__PURE__ */ l(
        "i",
        {
          className: "wx-35Np0p wx-icon wxi-angle-right" + (i === o ? " wx-disabled" : ""),
          onClick: () => v("next")
        }
      ),
      /* @__PURE__ */ l(
        "i",
        {
          className: "wx-35Np0p wx-icon wxi-angle-dbl-right" + (i === o ? " wx-disabled" : ""),
          onClick: () => v("last")
        }
      )
    ] }),
    /* @__PURE__ */ R("div", { className: "wx-35Np0p wx-right", children: [
      a("Total pages"),
      ": ",
      o
    ] })
  ] });
};
function It({
  id: e,
  label: n = "",
  value: t,
  name: s = "",
  inputValue: r = "",
  disabled: u = !1,
  onChange: i
}) {
  const c = _e(), a = z(e) || c, [o, f] = B(!1), d = t !== void 0, p = d ? t : o;
  function v(x) {
    const w = x.target.checked;
    d || f(w), w && i && i({ value: !0, inputValue: r });
  }
  return /* @__PURE__ */ R("div", { className: "wx-3lDF1s wx-radio", children: [
    /* @__PURE__ */ l(
      "input",
      {
        type: "radio",
        className: "wx-3lDF1s wx-input",
        id: a,
        disabled: u,
        name: s,
        value: r,
        checked: p,
        onChange: v
      }
    ),
    /* @__PURE__ */ R("label", { htmlFor: a, className: "wx-3lDF1s wx-label", children: [
      /* @__PURE__ */ l("span", { className: "wx-3lDF1s wx-circle" }),
      n && /* @__PURE__ */ l("span", { className: "wx-3lDF1s wx-radio-text", children: n })
    ] })
  ] });
}
function ln({
  options: e = [{}],
  value: n = "",
  onChange: t,
  type: s = ""
}) {
  const r = oe(), [u, i] = B(n);
  W(() => {
    i(n);
  }, [n]);
  function c(a) {
    const o = a.inputValue;
    i(o), t && t({ value: o });
  }
  return /* @__PURE__ */ l(ue.Provider, { value: null, children: /* @__PURE__ */ l("div", { className: `wx-38w70j wx-radiogroup ${s && `wx-${s}`}`, children: e.map((a) => /* @__PURE__ */ l("div", { className: "wx-38w70j wx-item", children: /* @__PURE__ */ l(
    It,
    {
      label: a.label,
      inputValue: a.id,
      value: u === a.id,
      name: r,
      onChange: c
    }
  ) }, a.id)) }) });
}
function cn({
  value: e = "",
  options: n = [],
  textOptions: t = null,
  placeholder: s = "",
  disabled: r = !1,
  error: u = !1,
  title: i = "",
  textField: c = "label",
  clear: a = !1,
  children: o,
  onChange: f,
  dropdown: d = {}
}) {
  const p = S(null), v = S(null);
  let [x, w] = J(e);
  function h(N) {
    p.current = N.navigate, v.current = N.keydown;
  }
  const m = A(() => x || x === 0 ? (t || n).find((N) => N.id === x) : null, [x, t, n]), g = D(
    ({ id: N }) => {
      (N || N === 0) && (w(N), p.current(null), f && f({ value: N }));
    },
    [w, f]
  ), T = D(
    (N) => {
      N.stopPropagation(), w(""), f && f({ value: "" });
    },
    [w, f]
  ), b = D(() => n.findIndex((N) => N.id === x), [n, x]);
  return /* @__PURE__ */ R(
    "div",
    {
      className: `wx-2YgblL wx-richselect ${u ? "wx-2YgblL wx-error" : ""} ${r ? "wx-2YgblL wx-disabled" : ""} ${o ? "" : "wx-2YgblL wx-nowrap"}`,
      title: i,
      onClick: () => p.current(b()),
      onKeyDown: (N) => v.current(N, b()),
      tabIndex: 0,
      children: [
        /* @__PURE__ */ l("div", { className: "wx-2YgblL wx-label", children: m ? o ? o(m) : m[c] : s ? /* @__PURE__ */ l("span", { className: "wx-2YgblL wx-placeholder", children: s }) : " " }),
        a && !r && x ? /* @__PURE__ */ l("i", { className: "wx-2YgblL wx-icon wxi-close", onClick: T }) : /* @__PURE__ */ l("i", { className: "wx-2YgblL wx-icon wxi-angle-down" }),
        !r && /* @__PURE__ */ l(ge, { items: n, onReady: h, onSelect: g, ...d, children: ({ option: N }) => o ? o(N) : N[c] })
      ]
    }
  );
}
const sn = ({
  options: e = [],
  value: n = "",
  css: t = "",
  children: s,
  onChange: r
}) => {
  const [u, i] = B(n), c = (a) => {
    i(a), r && r({ value: a });
  };
  return W(() => {
    i(n);
  }, [n]), /* @__PURE__ */ l("div", { className: `wx-0hMO7P wx-segmented ${t}`, children: e.map((a) => /* @__PURE__ */ l(
    "button",
    {
      className: `wx-0hMO7P wx-segment ${a.css} ${a.id === u ? "wx-selected" : ""}`,
      title: a.title,
      onClick: () => c(a.id),
      children: s ? ae(s, { option: a }) : /* @__PURE__ */ R(_, { children: [
        a.icon && /* @__PURE__ */ l(
          "i",
          {
            className: `wx-0hMO7P wx-icon ${a.icon} ${a.label ? "" : "wx-only"}`
          }
        ),
        a.label && /* @__PURE__ */ l("span", { className: "wx-0hMO7P wx-label", children: a.label })
      ] })
    },
    a.id
  )) });
};
function an({
  value: e,
  options: n = [],
  placeholder: t = "",
  title: s = "",
  disabled: r = !1,
  error: u = !1,
  textField: i = "label",
  clear: c = !1,
  id: a,
  onChange: o
}) {
  const f = z(a), [d, p] = B(e || ""), v = D(
    (w) => {
      const h = w.target.value;
      p(h), o && o({ value: h });
    },
    [o]
  ), x = D(() => {
    p(""), o && o({ value: "" });
  }, [o]);
  return /* @__PURE__ */ R("div", { className: "wx-2yx1Fo wx-select-box", children: [
    /* @__PURE__ */ R(
      "select",
      {
        id: f,
        value: d,
        disabled: r,
        className: `wx-2yx1Fo wx-select ${u ? "wx-2yx1Fo wx-error" : ""}`,
        title: s,
        onChange: v,
        children: [
          /* @__PURE__ */ l("option", { value: "", children: t }),
          n.map((w) => /* @__PURE__ */ l("option", { value: w.id, children: w[i] }, w.id))
        ]
      }
    ),
    c && !r && d ? /* @__PURE__ */ l("i", { className: "wx-2yx1Fo wx-icon wxi-close", onClick: x }) : /* @__PURE__ */ l("i", { className: "wx-2yx1Fo wx-icon wxi-angle-down" })
  ] });
}
function Ce({
  id: e,
  label: n = "",
  css: t = "",
  min: s = 0,
  max: r = 100,
  value: u = 0,
  step: i = 1,
  title: c = "",
  disabled: a = !1,
  onChange: o,
  width: f = ""
}) {
  const d = z(e), [p, v] = J(u), x = S({ value: p, input: p }), w = A(
    () => (p - s) / (r - s) * 100 + "%",
    [p, s, r]
  ), h = A(() => a ? "" : `linear-gradient(90deg, var(--wx-slider-primary) 0% ${w}, var(--wx-slider-background) ${w} 100%)`, [a, w]);
  function m({ target: N }) {
    const k = N.value * 1;
    v(k), o && o({
      value: k,
      previous: x.current.input,
      input: !0
    }), x.current.input = k;
  }
  function g({ target: N }) {
    const k = N.value * 1;
    v(k), o && o({ value: k, previous: x.current.value }), x.current.value = k;
  }
  W(() => {
    v(u);
  }, [u]);
  const T = S(null);
  W(() => {
    if (T.current)
      return T.current.addEventListener("change", g), () => {
        T.current && T.current.removeEventListener("change", g);
      };
  }, [T, g]);
  const b = { ...f && { width: f } };
  return /* @__PURE__ */ R("div", { className: `wx-2EDJ8G wx-slider ${t}`, style: b, title: c, children: [
    n && /* @__PURE__ */ l("label", { className: "wx-2EDJ8G wx-label", htmlFor: d, children: n }),
    /* @__PURE__ */ l("div", { className: "wx-2EDJ8G wx-inner", children: /* @__PURE__ */ l(
      "input",
      {
        id: d,
        className: "wx-2EDJ8G wx-input",
        type: "range",
        min: s,
        max: r,
        step: i,
        disabled: a,
        value: p,
        onInput: m,
        style: { background: h },
        ref: T
      }
    ) })
  ] });
}
function on({
  id: e,
  value: n = !1,
  disabled: t = !1,
  onChange: s
}) {
  const r = z(e), [u, i] = J(n), c = D(
    (a) => {
      const o = a.target.checked;
      i(o), s && s({ value: o });
    },
    [s]
  );
  return /* @__PURE__ */ R("label", { className: "wx-2dAR5c wx-switch", children: [
    /* @__PURE__ */ l(
      "input",
      {
        className: "wx-2dAR5c wx-input",
        type: "checkbox",
        checked: u,
        onChange: c,
        disabled: t,
        id: r
      }
    ),
    /* @__PURE__ */ l("span", { className: "wx-2dAR5c wx-box" })
  ] });
}
const un = ({
  options: e = [],
  value: n = "",
  type: t = "top",
  css: s = "",
  onChange: r
}) => {
  const [u, i] = J(n), c = (a) => {
    i(a), r && r({ value: a });
  };
  return /* @__PURE__ */ l("div", { className: `wx-138fWJ wx-tabs wx-${t} ${s}`, children: e.map((a) => /* @__PURE__ */ R(
    "button",
    {
      className: `wx-138fWJ wx-tab ${a.css} ${a.id === u ? "wx-active" : ""}`,
      title: a.title,
      onClick: () => c(a.id),
      children: [
        a.icon && /* @__PURE__ */ l(
          "i",
          {
            className: `wx-138fWJ wx-icon ${a.icon} ${a.label ? "" : "wx-only"}`
          }
        ),
        a.label && /* @__PURE__ */ l("span", { className: "wx-138fWJ wx-label", children: a.label })
      ]
    },
    a.id
  )) });
}, dn = ({
  id: e,
  value: n = 0,
  step: t = 1,
  min: s = 0,
  max: r = 1 / 0,
  error: u = !1,
  disabled: i = !1,
  readonly: c = !1,
  onChange: a
}) => {
  const o = z(e), [f, d] = J(n), p = D(() => {
    if (c || f <= s) return;
    const h = f - t;
    d(h), a && a({ value: h });
  }, [f, c, s, t, a]), v = D(() => {
    if (c || f >= r) return;
    const h = f + t;
    d(h), a && a({ value: h });
  }, [f, c, r, t, a]), x = D(() => {
    if (!c) {
      const h = Math.round(Math.min(r, Math.max(f, s)) / t) * t, m = isNaN(h) ? Math.max(s, 0) : h;
      d(m), a && a({ value: m });
    }
  }, [c, f, r, s, t, a]), w = D(
    (h) => {
      const m = h.target.value * 1;
      d(m), a && a({ value: m, input: !0 });
    },
    [a]
  );
  return /* @__PURE__ */ R(
    "div",
    {
      className: `wx-22t21n wx-counter ${i ? "wx-disabled" : ""} ${c ? "wx-readonly" : ""} ${u ? "wx-error" : ""}`,
      children: [
        /* @__PURE__ */ l(
          "button",
          {
            "aria-label": "-",
            className: "wx-22t21n wx-btn wx-btn-dec",
            disabled: i,
            onClick: p,
            children: /* @__PURE__ */ l(
              "svg",
              {
                className: "wx-22t21n wx-dec",
                width: "12",
                height: "2",
                viewBox: "0 0 12 2",
                fill: "none",
                xmlns: "http://www.w3.org/2000/svg",
                children: /* @__PURE__ */ l("path", { d: "M11.2501 1.74994H0.750092V0.249939H11.2501V1.74994Z" })
              }
            )
          }
        ),
        /* @__PURE__ */ l(
          "input",
          {
            id: o,
            type: "text",
            className: "wx-22t21n wx-input",
            disabled: i,
            readOnly: c,
            required: !0,
            value: f,
            onBlur: x,
            onInput: w
          }
        ),
        /* @__PURE__ */ l(
          "button",
          {
            "aria-label": "-",
            className: "wx-22t21n wx-btn wx-btn-inc",
            disabled: i,
            onClick: v,
            children: /* @__PURE__ */ l(
              "svg",
              {
                className: "wx-22t21n wx-inc",
                width: "12",
                height: "12",
                viewBox: "0 0 12 12",
                fill: "none",
                xmlns: "http://www.w3.org/2000/svg",
                children: /* @__PURE__ */ l(
                  "path",
                  {
                    d: `M11.2501
								6.74994H6.75009V11.2499H5.25009V6.74994H0.750092V5.24994H5.25009V0.749939H6.75009V5.24994H11.2501V6.74994Z`
                  }
                )
              }
            )
          }
        )
      ]
    }
  );
};
function Ct({ notice: e = {} }) {
  function n() {
    e.remove && e.remove();
  }
  return /* @__PURE__ */ R(
    "div",
    {
      className: `wx-11sNg5 wx-notice wx-${e.type ? e.type : ""}`,
      role: "status",
      "aria-live": "polite",
      children: [
        /* @__PURE__ */ l("div", { className: "wx-11sNg5 wx-text", children: e.text }),
        /* @__PURE__ */ l("div", { className: "wx-11sNg5 wx-button", children: /* @__PURE__ */ l("i", { className: "wx-11sNg5 wxi-close", onClick: n }) })
      ]
    }
  );
}
function St({ data: e = [] }) {
  return /* @__PURE__ */ l("div", { className: "wx-3nwoO9 wx-notices", children: e.map((n) => /* @__PURE__ */ l(Ct, { notice: n }, n.id)) });
}
function Lt({
  title: e = "",
  buttons: n = ["cancel", "ok"],
  header: t,
  children: s,
  footer: r,
  onConfirm: u,
  onCancel: i
}) {
  const c = (K(q) || X()).getGroup("core"), a = S(null);
  W(() => {
    a.current?.focus();
  }, []);
  function o(d) {
    switch (d.code) {
      case "Enter": {
        const p = d.target.tagName;
        if (p === "TEXTAREA" || p === "BUTTON") return;
        u && u({ event: d });
        break;
      }
      case "Escape":
        i && i({ event: d });
        break;
    }
  }
  function f(d, p) {
    const v = { event: d, button: p };
    p === "cancel" ? i && i(v) : u && u(v);
  }
  return /* @__PURE__ */ l(
    "div",
    {
      className: "wx-1FxkZa wx-modal",
      ref: a,
      tabIndex: 0,
      onKeyDown: o,
      children: /* @__PURE__ */ R("div", { className: "wx-1FxkZa wx-window", children: [
        t || (e ? /* @__PURE__ */ l("div", { className: "wx-1FxkZa wx-header", children: e }) : null),
        /* @__PURE__ */ l("div", { children: s }),
        r || n && /* @__PURE__ */ l("div", { className: "wx-1FxkZa wx-buttons", children: n.map((d) => /* @__PURE__ */ l("div", { className: "wx-1FxkZa wx-button", children: /* @__PURE__ */ l(
          re,
          {
            type: `block ${d === "ok" ? "primary" : "secondary"}`,
            onClick: (p) => f(p, d),
            children: c(d)
          }
        ) }, d)) })
      ] })
    }
  );
}
function Ft({ children: e }, n) {
  const [t, s] = B(null), [r, u] = B([]);
  return qe(
    n,
    () => ({
      showModal: (i) => {
        const c = { ...i };
        return s(c), new Promise((a, o) => {
          c.resolve = (f) => {
            s(null), a(f);
          }, c.reject = (f) => {
            s(null), o(f);
          };
        });
      },
      showNotice: (i) => {
        i = { ...i }, i.id = i.id || oe(), i.remove = () => u((c) => c.filter((a) => a.id !== i.id)), i.expire != -1 && setTimeout(i.remove, i.expire || 5100), u((c) => [...c, i]);
      }
    }),
    []
  ), /* @__PURE__ */ R(_, { children: [
    e,
    t && /* @__PURE__ */ l(
      Lt,
      {
        title: t.title,
        buttons: t.buttons,
        onConfirm: t.resolve,
        onCancel: t.reject,
        children: t.message
      }
    ),
    /* @__PURE__ */ l(St, { data: r })
  ] });
}
const Ht = Qe(Ft);
function fn(e) {
  let { children: n } = e, t = S();
  const s = A(() => ({
    showNotice: (r) => t.current.showNotice(r),
    showModal: (r) => t.current.showModal(r)
  }), []);
  return /* @__PURE__ */ l(Oe.Provider, { value: s, children: /* @__PURE__ */ l(Ht, { ref: t, children: n }) });
}
function Se({
  label: e = "",
  position: n = "",
  error: t = !1,
  type: s = "",
  required: r = !1,
  id: u,
  children: i,
  width: c = ""
}) {
  const a = A(() => u === void 0 ? oe() : u, [u]), o = { ...c && { width: c } };
  return /* @__PURE__ */ l(ue.Provider, { value: a, children: /* @__PURE__ */ R(
    "div",
    {
      style: o,
      className: `wx-2oVUvC wx-field wx-${n} ${t ? "wx-error" : ""} ${r ? "wx-required" : ""}`.trim(),
      children: [
        e && (a ? /* @__PURE__ */ l("label", { className: "wx-2oVUvC wx-label", htmlFor: a, children: e }) : /* @__PURE__ */ l("div", { className: "wx-2oVUvC wx-label", children: e })),
        /* @__PURE__ */ l("div", { className: `wx-2oVUvC wx-field-control wx-${s}`, children: i })
      ]
    }
  ) });
}
const Pt = ({
  value: e = !1,
  type: n = "",
  icon: t = "",
  disabled: s = !1,
  iconActive: r = "",
  onClick: u,
  title: i = "",
  css: c = "",
  text: a = "",
  textActive: o = "",
  children: f,
  active: d,
  onChange: p
}) => {
  const [v, x] = J(e), w = A(() => (v ? "pressed" : "") + (n ? " " + n : ""), [v, n]), h = D(
    (m) => {
      let g = !v;
      u && u(m), m.defaultPrevented || (x(g), p && p({ value: g }));
    },
    [v, u, p]
  );
  return v && d ? /* @__PURE__ */ l(
    re,
    {
      title: i,
      text: v && o || a,
      css: c,
      type: w,
      icon: v && r || t,
      onClick: h,
      disabled: s,
      children: ae(d, { value: v })
    }
  ) : f ? /* @__PURE__ */ l(
    re,
    {
      title: i,
      text: v && o || a,
      css: c,
      type: w,
      icon: v && r || t,
      onClick: h,
      disabled: s,
      children: f
    }
  ) : /* @__PURE__ */ l(
    re,
    {
      title: i,
      text: v && o || a,
      css: c,
      type: w,
      icon: v && r || t,
      onClick: h,
      disabled: s
    }
  );
}, Le = new Date(0, 0, 0, 0, 0);
function wn({
  value: e = Le,
  id: n,
  title: t = "",
  css: s = "",
  disabled: r = !1,
  error: u = !1,
  format: i = "",
  onChange: c,
  dropdown: a
}) {
  let [o, f] = J(e);
  const { calendar: d, formats: p } = (K(q) || X()).getRaw(), v = d.clockFormat == 12, x = 23, w = 59, h = A(() => {
    const y = i || p?.timeFormat;
    return typeof y == "function" ? y : ne(y, d);
  }, [i, p, d]), m = A(() => h(new Date(0, 0, 0, 1)).indexOf("01") != -1, [h]), g = (y, E) => (y < 10 && E ? `0${y}` : `${y}`).slice(-2), T = (y) => g(y, !0), b = (y) => `${y}`.replace(/[^\d]/g, "") || 0, N = (y) => v && (y = y % 12, y === 0) ? "12" : g(y, m), k = D((y, E) => (y = b(y), Math.min(y, E)), []), [M, $] = B(null), V = o || Le, F = k(V.getHours(), x), j = k(V.getMinutes(), w), H = F > 12, L = N(F), P = T(j), C = A(
    () => h(new Date(0, 0, 0, F, j)),
    [F, j, h]
  ), O = D(() => {
    $(!0);
  }, []), U = D(() => {
    const y = new Date(V);
    y.setHours(y.getHours() + (H ? -12 : 12)), f(y), c && c({ value: y });
  }, [V, H, c]), Y = D(
    ({ value: y }) => {
      if (V.getHours() === y) return;
      const E = new Date(V);
      E.setHours(y), f(E), c && c({ value: E });
    },
    [V, c]
  ), I = D(
    ({ value: y }) => {
      if (V.getMinutes() === y) return;
      const E = new Date(V);
      E.setMinutes(y), f(E), c && c({ value: E });
    },
    [V, c]
  ), G = D(
    (y) => (y = k(y, x), v && (y = y * 1, y === 12 && (y = 0), H && (y += 12)), y),
    [k, v, H]
  ), Q = D(() => {
    $(null);
  }, []);
  return /* @__PURE__ */ R(
    "div",
    {
      className: `wx-7f497i wx-timepicker ${u ? "wx-7f497i wx-error" : ""} ${r ? "wx-7f497i wx-disabled" : ""}`,
      onClick: r ? void 0 : O,
      style: { cursor: r ? "default" : "pointer" },
      children: [
        /* @__PURE__ */ l(
          Ne,
          {
            id: n,
            css: s,
            title: t,
            value: C,
            readonly: !0,
            disabled: r,
            error: u,
            icon: "wxi-clock",
            inputStyle: {
              cursor: "pointer",
              width: "100%",
              paddingRight: "calc(var(--wx-input-icon-size) + var(--wx-input-icon-indent) * 2)"
            }
          }
        ),
        M && !r && /* @__PURE__ */ l(le, { onCancel: Q, ...ye(a), children: /* @__PURE__ */ R("div", { className: "wx-7f497i wx-wrapper", children: [
          /* @__PURE__ */ R("div", { className: "wx-7f497i wx-timer", children: [
            /* @__PURE__ */ l(
              "input",
              {
                className: "wx-7f497i wx-digit",
                value: L,
                onChange: (y) => {
                  const E = G(y.target.value);
                  Y({ value: E });
                }
              }
            ),
            /* @__PURE__ */ l("div", { className: "wx-7f497i wx-separator", children: ":" }),
            /* @__PURE__ */ l(
              "input",
              {
                className: "wx-7f497i wx-digit",
                value: P,
                onChange: (y) => {
                  const E = k(y.target.value, w);
                  I({ value: E });
                }
              }
            ),
            v && /* @__PURE__ */ l(
              Pt,
              {
                value: H,
                onClick: U,
                active: () => /* @__PURE__ */ l("span", { children: "pm" }),
                children: /* @__PURE__ */ l("span", { children: "am" })
              }
            )
          ] }),
          /* @__PURE__ */ l(Se, { width: "unset", children: /* @__PURE__ */ l(
            Ce,
            {
              label: d.hours,
              value: F,
              width: "unset",
              onChange: Y,
              max: x
            }
          ) }),
          /* @__PURE__ */ l(Se, { width: "unset", children: /* @__PURE__ */ l(
            Ce,
            {
              label: d.minutes,
              value: j,
              width: "unset",
              onChange: I,
              max: w
            }
          ) })
        ] }) })
      ]
    }
  );
}
function xn({ children: e }) {
  return /* @__PURE__ */ l("div", { className: "wx-KgpO9N wx-modal", children: /* @__PURE__ */ l("div", { className: "wx-KgpO9N wx-window", children: e }) });
}
function mn({ position: e = "right", children: n, onCancel: t }) {
  const s = S(null);
  return W(() => ve(s.current, t).destroy, []), /* @__PURE__ */ l("div", { ref: s, className: `wx-2L733M wx-sidearea wx-pos-${e}`, children: n });
}
const Fe = "#dfe2e6", He = "#2c2f3c", At = 0.75;
function Pe(e) {
  if (e = e?.trim() || "", !e) return "";
  const n = e.split(/\s+/);
  return (n[0][0] + (n[1]?.[0] || "")).toUpperCase().slice(0, 2);
}
function Bt(e) {
  let n = e.replace("#", "");
  if (n.length === 3 && (n = n[0] + n[0] + n[1] + n[1] + n[2] + n[2]), n.length !== 6) return He;
  const t = parseInt(n.slice(0, 2), 16) / 255, s = parseInt(n.slice(2, 4), 16) / 255, r = parseInt(n.slice(4, 6), 16) / 255;
  return 0.299 * t + 0.587 * s + 0.114 * r > 0.5 ? He : "#ffffff";
}
function hn(e) {
  const { value: n, size: t = 32, limit: s } = e, r = S(null), [u, i] = B(null), c = A(() => n ? Array.isArray(n) ? n : [n] : [], [n]), a = A(() => {
    if (u == null || u <= 0)
      return null;
    const w = 1 + (u / t - 1) / At;
    return Math.max(1, Math.floor(w));
  }, [u, t]), o = A(() => {
    const w = s != null ? Math.min(c.length, s) : c.length;
    return a != null ? Math.min(w, a) : w;
  }, [c, s, a]), f = A(() => c.slice(0, o), [c, o]), d = A(() => Math.max(0, c.length - o), [c, o]);
  W(() => {
    const w = r.current;
    if (!w) return;
    const h = new ResizeObserver((m) => {
      const g = m[0];
      g && i(g.contentRect.width);
    });
    return h.observe(w), () => h.disconnect();
  }, []);
  const p = Math.round(t * 0.4), v = {
    width: t + "px",
    height: t + "px",
    minWidth: t + "px",
    minHeight: t + "px",
    fontSize: p + "px"
  };
  function x(w, h) {
    const m = h === 0 ? "0" : `${t * -0.25}px`, g = w.avatar ? "transparent" : w.color || Fe, T = w.avatar ? "transparent" : Bt(w.color || Fe);
    return { marginLeft: m, backgroundColor: g, color: T };
  }
  return /* @__PURE__ */ l("div", { className: "wx-avatar-root wx-aadkRiRf", ref: r, children: f.length > 0 && /* @__PURE__ */ l("div", { className: "wx-avatar-stack wx-aadkRiRf", children: f.map((w, h) => /* @__PURE__ */ R(
    "div",
    {
      className: [
        "wx-avatar",
        "wx-avatar-item",
        h === f.length - 1 && d > 0 ? "wx-avatar-overflow" : "",
        "wx-aadkRiRf"
      ].filter(Boolean).join(" "),
      style: { ...v, ...x(w, h) },
      children: [
        w.avatar ? /* @__PURE__ */ l("img", { src: w.avatar, alt: "", loading: "lazy" }) : Pe(w.name) ? /* @__PURE__ */ l("span", { className: "wx-aadkRiRf", children: Pe(w.name) }) : null,
        h === f.length - 1 && d > 0 && /* @__PURE__ */ R("span", { className: "wx-avatar-overflow-badge wx-aadkRiRf", children: [
          "+",
          d
        ] })
      ]
    },
    w.id
  )) }) });
}
function Et() {
  return /* @__PURE__ */ l(_, {});
}
function pn(e) {
  const { fonts: n = !0, children: t } = e;
  return /* @__PURE__ */ l(ie.Provider, { value: "material", children: /* @__PURE__ */ R(_, { children: [
    t && /* @__PURE__ */ l("div", { className: "wx-theme wx-material-theme", children: t }),
    n && /* @__PURE__ */ R(_, { children: [
      /* @__PURE__ */ l(
        "link",
        {
          rel: "preconnect",
          href: "https://cdn.svar.dev",
          crossOrigin: "true"
        }
      ),
      /* @__PURE__ */ l(Et, {}),
      /* @__PURE__ */ l(
        "link",
        {
          rel: "stylesheet",
          href: "https://cdn.svar.dev/fonts/wxi/wx-icons.css"
        }
      )
    ] })
  ] }) });
}
function Je() {
  return /* @__PURE__ */ l(_, {});
}
function vn(e) {
  const { fonts: n = !0, children: t } = e;
  return /* @__PURE__ */ l(ie.Provider, { value: "willow", children: /* @__PURE__ */ R(_, { children: [
    t && t && /* @__PURE__ */ l("div", { className: "wx-theme wx-willow-theme", children: t }),
    n && /* @__PURE__ */ R(_, { children: [
      /* @__PURE__ */ l(
        "link",
        {
          rel: "preconnect",
          href: "https://cdn.svar.dev",
          crossOrigin: "true"
        }
      ),
      /* @__PURE__ */ l(Je, {}),
      /* @__PURE__ */ l(
        "link",
        {
          rel: "stylesheet",
          href: "https://cdn.svar.dev/fonts/wxi/wx-icons.css"
        }
      )
    ] })
  ] }) });
}
function gn(e) {
  const { fonts: n = !0, children: t } = e;
  return /* @__PURE__ */ l(ie.Provider, { value: "willow-dark", children: /* @__PURE__ */ R(_, { children: [
    t && t && /* @__PURE__ */ l("div", { className: "wx-theme wx-willow-dark-theme", children: t }),
    n && /* @__PURE__ */ R(_, { children: [
      /* @__PURE__ */ l(
        "link",
        {
          rel: "preconnect",
          href: "https://cdn.svar.dev",
          crossOrigin: "true"
        }
      ),
      /* @__PURE__ */ l(Je, {}),
      /* @__PURE__ */ l(
        "link",
        {
          rel: "stylesheet",
          href: "https://cdn.svar.dev/fonts/wxi/wx-icons.css"
        }
      )
    ] })
  ] }) });
}
function Nn(e) {
  const n = oe();
  e.setAttribute("data-wx-portal-root", n);
}
et(pe);
export {
  hn as Avatar,
  re as Button,
  Dt as Calendar,
  Ye as Checkbox,
  Qt as CheckboxGroup,
  ot as ColorBoard,
  Kt as ColorPicker,
  qt as ColorSelect,
  Zt as Combo,
  dn as Counter,
  zt as DatePicker,
  Xt as DateRangePicker,
  le as Dropdown,
  Se as Field,
  en as Fullscreen,
  fn as Globals,
  tn as Icon,
  he as Locale,
  pn as Material,
  Lt as Modal,
  xn as ModalArea,
  ft as Month,
  nn as MultiCombo,
  rn as Pager,
  lt as Popup,
  nt as Portal,
  It as RadioButton,
  ln as RadioButtonGroup,
  Rt as RangeCalendar,
  cn as RichSelect,
  sn as Segmented,
  an as Select,
  mn as SideArea,
  Ce as Slider,
  ge as SuggestDropdown,
  on as Switch,
  un as Tabs,
  Ne as Text,
  _t as TextArea,
  wn as TimePicker,
  Pt as TwoState,
  vn as Willow,
  gn as WillowDark,
  Ut as context,
  Vn as en,
  bn as locale,
  Nn as popupContainer
};
