// @svar-ui/react-filter/dist/index.es.js
import { jsx as e, jsxs as O, Fragment as se } from "react/jsx-runtime";
import gt, { createContext as xt, useContext as ee, useState as _, useRef as x, useEffect as L, useMemo as T, useCallback as Z, Fragment as Xe, forwardRef as Ye, useImperativeHandle as Je } from "react";
import { context as ae, RichSelect as _e, DateRangePicker as Ze, DatePicker as et, Text as Ie, Combo as mt, Button as xe, Checkbox as wt, Locale as Oe, Portal as yt, Popup as vt, Dropdown as bt, Material as Me, Willow as qe, WillowDark as Be } from "@svar-ui/react-core";
import { DropDownMenu as Ct, ContextMenu as Nt } from "@svar-ui/react-menu";
import { dateToString as Ee, delegateClick as Rt, locateID as ze, locale as St } from "@svar-ui/lib-dom";
import { useStore as Ne, useStoreWithCounter as tt, writable as Ft, useWritableProp as At } from "@svar-ui/lib-react";
import { en as Te } from "@svar-ui/filter-locales";
import { getFilters as je, getFilter as nt, DataStore as $t, createFilterRule as _t, getQueryHtml as Tt, prepareFields as kt, buildFieldMaps as Dt, parseSet as He, getAutocompleteContext as Pt, getSuggestions as Vt, idsToLabels as Qe, labelsToIds as Ue } from "@svar-ui/filter-store";
import { createArrayFilter as an, createFilter as un, createFilterRule as dn, getFilter as fn, getFilters as pn, getOptions as hn, getOptionsMap as gn, getQueryHtml as xn, serialize as mn } from "@svar-ui/filter-store";
import { EventBusRouter as It } from "@svar-ui/lib-state";
const Re = xt(null);
function Et(c) {
  const {
    fields: l = null,
    fieldsSelector: v = !0,
    field: a = null,
    buttons: w = !0,
    options: h = null,
    includes: y = null,
    type: R = "text",
    filter: o = "",
    value: u = "",
    format: b = null,
    predicate: f = null
  } = c, F = c.onApply || c.onapply, N = c.onCancel || c.oncancel, A = c.onChange || c.onchange, m = ee(ae.i18n), r = m.getRaw(), p = r.filter?.dateFormat || r.formats?.dateFormat, D = Ee(p, r.calendar), E = m.getGroup("filter"), S = 1, G = 2, U = 4, K = 8, j = 16, X = 32, t = 64, [g, d] = _(), $ = x(), [I, le] = _(), H = x(), [ue, q] = _(), Q = x(), [te, de] = _(), P = x(), [me, Se] = _([]), M = x([]), [we, ye] = _(), J = x(), [, fe] = _(), Y = x(), [, pe] = _(), ce = x(), [W, ke] = _(!1), [Fe, De] = _(null);
  async function Pe(n) {
    if (!h || typeof h != "function") return;
    const i = await h(n);
    De(i || []);
  }
  const he = x(0), ve = x(null);
  function Ae(n) {
    return n ? n.toString() : "";
  }
  function s(n, i, V) {
    n(V), i.current = V;
  }
  function C(n, i) {
    switch (n) {
      case S:
        if (H.current !== i)
          return s(le, H, i), n;
        break;
      case j:
        if (Q.current !== i)
          return s(q, Q, i), n;
        break;
      case G:
        if ($.current !== i)
          return s(d, $, i), n;
        break;
      case U:
        if (P.current !== i)
          return s(de, P, i), n;
        break;
      case K:
        if (Ae(M.current) != Ae(i))
          return s(Se, M, i || []), n;
        break;
      case t:
        if (J.current !== i)
          return s(ye, J, i || []), n;
        break;
    }
    return 0;
  }
  function k(n) {
    he.current = he.current | n, ve.current || (ve.current = setTimeout(() => {
      const i = he.current;
      ve.current = null, he.current = 0, ot(i);
    }));
  }
  const [B, ne] = _(), z = x(), [ge, oe] = _([]), be = x([]), [, rt] = _(), Ge = x(), [lt, ct] = _(), it = x();
  function ot(n) {
    if (n & S && l) {
      s(ye, J, null);
      const i = l.find((re) => re.id === H.current), V = P.current instanceof Date || P.current && typeof P.current == "object", ie = i.type === "tuple" || Q.current == "tuple";
      if (V && i.type !== "date" || !V && i.type == "date" || isNaN(P.current) && i.type == "number" || ie)
        s(de, P, null);
      else if (i.type === "text" && typeof P.current != "string") {
        const re = P.current || P.current === 0 ? P.current.toString() : "";
        s(de, P, re);
      }
      s(fe, Y, i.format), s(pe, ce, i.predicate), n = n | C(j, i.type || "text"), Pe(H.current);
    }
    if (n & j) {
      const i = je(Q.current).map((V) => ({
        id: V.id,
        label: E(V.label || V.id)
      }));
      s(ct, it, i), i.some((V) => V?.id === $.current) || s(d, $, null);
    }
    if (n & j || n & G) {
      let i = nt($.current, Q.current);
      i || (i = je(Q.current).find((V) => V.default), n = n | C(G, i.id)), s(rt, Ge, i), typeof P.current == "object" && !(P.current instanceof Date) && !i.range && (n = n | C(
        U,
        Q.current === "date" ? null : ""
      ));
    }
    if (n & G || n & U || n & t)
      if (J.current) {
        const i = P.current || P.current === 0 ? (re) => Ge.current.handler(re, P.current || "") : null, V = $.current && i ? J.current.filter(i) : J.current;
        s(oe, be, V);
        let ie = M.current;
        M.current.length && (ie = M.current.filter((re) => V.includes(re))), ie.length ? ie.length !== M.current.length && (n = n | C(K, ie)) : n = n | C(K, [...V]);
      } else
        s(oe, be, []);
    if (n & K) {
      const i = M.current.length === be.current.length;
      s(ne, z, i);
    }
    if (n & X) {
      const i = Le();
      A && A({ value: i });
    }
  }
  function st() {
    const n = Le();
    F && F({ value: n });
  }
  function at() {
    N && N();
  }
  function ut() {
    const n = !z.current;
    s(ne, z, n), k(
      C(
        K,
        n ? [...be.current] : []
      ) | X
    );
  }
  function dt(n) {
    const { inputValue: i, value: V } = n, ie = V ? [...M.current, i] : M.current.filter((re) => re != i);
    k(C(K, ie) | X);
  }
  function Ce({ value: n }) {
    n === "$empty" && (n = ""), k(C(U, n) | X);
  }
  function ft({ value: n }) {
    k(C(S, n) | X);
  }
  function pt({ value: n }) {
    k(C(G, n) | X);
  }
  function Le() {
    const n = {
      filter: $.current,
      value: P.current,
      type: Q.current
    };
    return ce.current && (n.predicate = ce.current), H.current && (n.field = H.current), M.current && M.current.length && M.current.length !== be.current.length ? n.includes = [...M.current] : n.includes = [], n;
  }
  function We(n) {
    return Y.current && typeof Y.current == "function" ? Y.current(n) : n instanceof Date ? Y.current ? Ee(Y.current, r.calendar)(n) : D(n) : typeof n == "string" ? n : n.toString();
  }
  function ht(n) {
    let i = n ? n.map((V) => ({ id: V, label: We(V) })) : [];
    return [{ id: "$empty", label: "", emptyLabel: E("None") }].concat(i);
  }
  const Ve = x(null);
  return L(() => {
    H.current && Ve.current && setTimeout(() => {
      const n = Ve.current, i = n && n.querySelector("input");
      i && i.focus();
    }, 1);
  }, [I]), L(() => {
    W && k(C(S, a));
  }, [a, W]), L(() => {
    W && k(C(j, R));
  }, [R, W]), L(() => {
    W && k(C(G, o));
  }, [o, W]), L(() => {
    W && k(C(U, u));
  }, [u, W]), L(() => {
    W && k(C(K, y));
  }, [y, W]), L(() => {
    const n = Array.isArray(h) ? h : Fe;
    W && k(C(t, n));
  }, [h, Fe, W]), L(() => {
    b && !l && s(fe, Y, b), f && !l && s(pe, ce, f);
  }, [b, f, l]), L(() => {
    ke(!0);
  }, []), /* @__PURE__ */ e(ae.fieldId.Provider, { value: null, children: /* @__PURE__ */ O("div", { className: "wx-3z8r9Oys wx-filter-editor", children: [
    l && v ? /* @__PURE__ */ e(_e, { onChange: ft, options: l, value: I }) : null,
    /* @__PURE__ */ O("div", { className: "wx-3z8r9Oys wx-wrapper", children: [
      /* @__PURE__ */ e("div", { className: "wx-3z8r9Oys wx-cell", children: /* @__PURE__ */ e(
        _e,
        {
          onChange: pt,
          options: lt,
          value: g,
          placeholder: E("Click to select")
        }
      ) }),
      /* @__PURE__ */ e("div", { className: "wx-3z8r9Oys wx-cell", ref: Ve, children: ue === "date" ? g == "between" || g == "notBetween" ? /* @__PURE__ */ e(
        Ze,
        {
          format: p,
          value: te,
          buttons: ["done", "clear", "today"],
          onChange: Ce
        }
      ) : /* @__PURE__ */ e(et, { format: p, value: te, onChange: Ce }) : ue === "number" ? /* @__PURE__ */ e(Ie, { value: te, onChange: Ce, type: "number" }) : ue === "tuple" ? /* @__PURE__ */ e(
        mt,
        {
          value: te,
          options: ht(we),
          onChange: Ce,
          children: ({ option: n }) => n.emptyLabel || n.label
        }
      ) : /* @__PURE__ */ e(Ie, { value: te, onChange: Ce }) })
    ] }),
    /* @__PURE__ */ e(xe, { onClick: ut, children: E(B ? "Unselect all" : "Select all") }),
    /* @__PURE__ */ e("div", { className: "wx-3z8r9Oys wx-list", role: "listbox", children: ge.map((n, i) => /* @__PURE__ */ e(
      "div",
      {
        className: "wx-3z8r9Oys wx-item",
        tabIndex: i ? -1 : 0,
        role: "option",
        children: /* @__PURE__ */ e(
          wt,
          {
            label: We(n),
            inputValue: n,
            value: me && me.includes(n),
            onChange: dt
          }
        )
      },
      i
    )) }),
    w ? /* @__PURE__ */ O("div", { className: "wx-3z8r9Oys wx-wrapper", children: [
      /* @__PURE__ */ e(xe, { type: "secondary", onClick: at, children: E("Cancel") }),
      /* @__PURE__ */ e(xe, { type: "primary", onClick: st, children: E("Apply") })
    ] }) : null
  ] }) });
}
function Ot(c) {
  return /* @__PURE__ */ e(Oe, { words: Te, optional: !0, children: /* @__PURE__ */ e(Et, { ...c }) });
}
function Ke(c) {
  const { rule: l, type: v } = c, a = T(
    () => l.includes && l.includes.length ? l.includes : null,
    [l]
  ), w = T(() => l.filter || "contains", [l]), h = T(() => l.value || "", [l]), y = T(() => l.field, [l]), R = T(() => l.id, [l]), o = ee(Re), u = Ne(o, "fields"), b = Ne(o, "options"), f = Z(() => {
    o.exec("edit-rule", {});
  }, [o]), F = Z(
    ({ value: A }) => {
      o.exec("update-rule", { id: R, rule: A }), f();
    },
    [o, R]
  ), N = Z(
    ({ value: A }) => {
      o.exec("change-rule", { id: R, rule: A });
    },
    [o, R]
  );
  return /* @__PURE__ */ e("div", { className: "wx-uOIGm7Ce wx-panel", children: /* @__PURE__ */ e(
    Ot,
    {
      fieldsSelector: v !== "simple",
      fields: u,
      field: y,
      options: b,
      includes: a,
      filter: w,
      value: h,
      onApply: F,
      onCancel: f,
      onChange: N
    }
  ) });
}
function Gt({
  filter: c,
  field: l,
  type: v = "list",
  action: a = "menu"
}) {
  const w = (p) => p, h = ee(ae.i18n), y = x();
  y.current === void 0 && (y.current = h.getRaw());
  const R = y.current, o = x();
  o.current === void 0 && (o.current = l?.format || R.filter?.dateFormat || R.formats.dateFormat);
  const u = o.current, b = T(
    () => typeof u == "function" ? u : l?.type === "date" ? Ee(u, R.calendar) : w,
    [l]
  ), f = x();
  f.current === void 0 && (f.current = h.getGroup("filter"));
  const F = f.current, [N, A] = _(), [m, r] = _();
  return L(() => {
    r(c.value), A(c.filter ? nt(c.filter, c.type) : null);
  }, [c]), /* @__PURE__ */ O("div", { className: `wx-5GeVl88l wx-rule wx-${v}`, "data-id": c.id, children: [
    /* @__PURE__ */ O("span", { className: "wx-5GeVl88l wx-filter", children: [
      /* @__PURE__ */ e("span", { className: "wx-5GeVl88l wx-field", children: l.label }),
      c.includes && c.includes.length ? /* @__PURE__ */ O(se, { children: [
        " " + F("in") + " ",
        c.includes.map((p, D) => /* @__PURE__ */ O(gt.Fragment, { children: [
          /* @__PURE__ */ e("span", { className: "wx-5GeVl88l wx-value", children: b(p) }),
          D < c.includes.length - 1 ? ", " : null
        ] }, p))
      ] }) : c.filter && (m || m === 0) ? /* @__PURE__ */ O(se, { children: [
        " " + (F(N.short || N.label) || N.id) + " ",
        l.type === "date" ? m.start ? /* @__PURE__ */ O(se, { children: [
          /* @__PURE__ */ e("span", { className: "wx-5GeVl88l wx-value", children: b(m.start) }),
          " " + F("and") + " ",
          /* @__PURE__ */ e("span", { className: "wx-5GeVl88l wx-value", children: b(m.end) })
        ] }) : /* @__PURE__ */ e("span", { className: "wx-5GeVl88l wx-value", children: b(m) }) : /* @__PURE__ */ e("span", { className: "wx-5GeVl88l wx-value", children: b(m) })
      ] }) : null
    ] }),
    /* @__PURE__ */ e(
      "i",
      {
        className: `wx-5GeVl88l wxi-${a === "menu" ? "dots-v" : a} wx-menu-icon`,
        role: "button",
        "data-action": a
      }
    )
  ] });
}
function Lt({ mode: c = "and", id: l }) {
  const a = ee(ae.i18n).getGroup("filter"), w = ee(Re);
  function h() {
    w.exec("toggle-glue", { id: l });
  }
  return /* @__PURE__ */ e(
    "div",
    {
      className: `wx-XRdDajFX wx-glue wx-${c}`,
      role: "button",
      onClick: h,
      children: a(c)
    }
  );
}
function $e(c) {
  const { group: l, type: v, onShowMenu: a } = c, w = ee(Re), [h, y] = tt(w, "value"), R = Ne(w, "fields"), o = Ne(w, "_editor"), u = T(() => h && h.getBranch(l.id) || [], [y, l.id]), b = T(() => v == "simple" ? "line" : v, [v]), f = T(
    () => l.$level == 1 ? "top" : "inner",
    [l.$level]
  ), F = T(
    () => ({
      dblclick: (S) => w.exec("edit-rule", { id: S }),
      menu: (S, G) => a && a({ id: S, ev: G }),
      delete: (S) => {
        w.exec("edit-rule", {}), w.exec("delete-rule", { id: S });
      }
    }),
    [w, a]
  ), N = Z(
    (S) => R.find((G) => G.id == S),
    [R]
  ), A = x(null), [m, r] = _(), [p, D] = _();
  L(() => {
    const S = A.current;
    return S ? Rt(S, F) : void 0;
  }, [F]), L(() => {
    if (!(o && p === o.id)) {
      if (o && v !== "list" && u.find((S) => S.id === o.id)) {
        const S = A.current && A.current.querySelector(`[data-id='${o.id}']`);
        if (S) {
          r(S.getBoundingClientRect()), D(o.id);
          return;
        }
      }
      r(null), D(null);
    }
  }, [o, v, u, p]);
  function E() {
    r(null), D(null), w.exec("edit-rule", {});
  }
  return /* @__PURE__ */ O(se, { children: [
    /* @__PURE__ */ e(
      "div",
      {
        ref: A,
        className: `wx-3udP2tTA wx-group wx-${f} wx-${b}`,
        children: u.map((S, G) => /* @__PURE__ */ O(Xe, { children: [
          v === "list" && o && o.id === S.id ? /* @__PURE__ */ e(Ke, { rule: S }) : S.data ? /* @__PURE__ */ e($e, { type: v, group: S, onShowMenu: a }) : /* @__PURE__ */ e("div", { className: "wx-3udP2tTA wx-rule-wrapper", children: /* @__PURE__ */ e(
            Gt,
            {
              action: v == "simple" ? "delete" : "menu",
              type: b,
              filter: S,
              field: N(S.field)
            }
          ) }),
          v !== "simple" && G < u.length - 1 ? /* @__PURE__ */ e("div", { className: "wx-3udP2tTA wx-glue-wrapper", children: /* @__PURE__ */ e(Lt, { mode: l.glue, id: l.id }) }) : null
        ] }, S.id))
      }
    ),
    o && m ? /* @__PURE__ */ e(yt, { children: /* @__PURE__ */ e(
      vt,
      {
        top: m.top + m.height,
        left: m.left + m.width / 2,
        onCancel: E,
        children: /* @__PURE__ */ e("div", { className: "wx-3udP2tTA wx-editor-wrapper", children: /* @__PURE__ */ e(Ke, { type: v, rule: h.byId(o.id) }) })
      }
    ) }) : null
  ] });
}
function Wt({ type: c }) {
  const l = ee(Re), v = ee(ae.i18n), a = T(() => v.getGroup("filter"), [v]), w = Ne(l, "fields"), [h, y] = tt(l, "value"), R = h.getBranch(0)[0];
  function o(m) {
    l.exec("add-rule", { rule: { $temp: !0, ...m }, edit: !0 });
  }
  const u = T(() => {
    const m = {};
    return h.forEach((r) => m[r.field] = !0), w.filter((r) => !m[r.id]).map((r) => ({ id: r.id, text: r.label }));
  }, [y, w]), b = Z(
    (m) => {
      const r = m.action;
      r && o({ field: r.id });
    },
    [o]
  ), f = T(
    () => [
      {
        id: "edit-rule",
        text: a("Edit"),
        icon: "wxi-edit-outline"
      },
      {
        id: "add-rule",
        text: a("Add filter"),
        icon: "wxi-filter-plus-outline",
        resolver: (m) => ({
          after: m,
          rule: { $temp: !0 },
          edit: !0
        })
      },
      {
        id: "add-group",
        text: a("Add group"),
        icon: "wxi-filter-multiple-outline",
        resolver: (m) => ({
          after: m,
          rule: { $temp: !0 },
          edit: !0
        })
      },
      { type: "separator" },
      {
        id: "delete-rule",
        text: a("Delete"),
        icon: "wxi-delete-outline"
      }
    ],
    [a]
  ), F = x(null), N = Z(
    (m) => {
      const { context: r, action: p } = m;
      if (p) {
        const D = p.resolver ? p.resolver(r) : { id: r };
        l.exec(p.id, D);
      }
    },
    [l]
  ), A = Z(({ ev: m, id: r }) => {
    F.current && typeof F.current.show == "function" && F.current.show(m, r);
  }, []);
  return /* @__PURE__ */ O(se, { children: [
    /* @__PURE__ */ e("div", { className: `wx-VDDi7d7g wx-filter-builder wx-${c}`, children: c === "list" ? /* @__PURE__ */ O(se, { children: [
      /* @__PURE__ */ e("div", { className: `wx-VDDi7d7g wx-toolbar wx-${c}`, children: /* @__PURE__ */ e(xe, { type: "primary", onClick: o, children: a("Add filter") }) }),
      /* @__PURE__ */ e($e, { type: c, group: R, onShowMenu: A })
    ] }) : c === "line" ? /* @__PURE__ */ O("div", { className: `wx-VDDi7d7g wx-toolbar wx-${c}`, children: [
      /* @__PURE__ */ e("div", { className: "wx-VDDi7d7g wx-filters", children: /* @__PURE__ */ e($e, { type: c, group: R, onShowMenu: A }) }),
      /* @__PURE__ */ e("div", { className: "wx-VDDi7d7g wx-button", children: /* @__PURE__ */ e(xe, { type: "primary", onClick: o, children: a("Add filter") }) })
    ] }) : c === "simple" ? /* @__PURE__ */ O("div", { className: `wx-VDDi7d7g wx-toolbar wx-${c}`, children: [
      /* @__PURE__ */ e("div", { className: "wx-VDDi7d7g wx-button", children: /* @__PURE__ */ e(
        Ct,
        {
          options: u,
          onClick: b,
          children: /* @__PURE__ */ e(xe, { disabled: !u.length, type: "primary", children: a("Add filter") })
        }
      ) }),
      /* @__PURE__ */ e("div", { className: "wx-VDDi7d7g wx-filters", children: /* @__PURE__ */ e($e, { type: c, group: R, onShowMenu: A }) })
    ] }) : null }),
    c !== "simple" ? /* @__PURE__ */ e(Nt, { options: f, onClick: N, ref: F }) : null
  ] });
}
const Mt = (c) => c.split("-").map((l) => l ? l.charAt(0).toUpperCase() + l.slice(1) : "").join(""), en = Ye(function({
  value: l = { glue: "and", rules: [] },
  fields: v = [],
  options: a = null,
  type: w = "list",
  init: h = null,
  ...y
}, R) {
  const o = x();
  o.current = y;
  const u = T(() => new $t(Ft), []), b = T(() => u.in, [u]), f = x(null);
  f.current === null && (f.current = new It((A, m) => {
    const r = "on" + Mt(A);
    o.current && o.current[r] && o.current[r](m);
  }), b.setNext(f.current));
  const F = T(
    () => ({
      getState: u.getState.bind(u),
      getReactiveState: u.getReactive.bind(u),
      getStores: () => ({ data: u }),
      exec: b.exec,
      setNext: (A) => (f.current = f.current.setNext(A), f.current),
      intercept: b.intercept.bind(b),
      on: b.on.bind(b),
      detach: b.detach.bind(b),
      getValue: u.getValue.bind(u)
    }),
    [u, b]
  ), N = x(!0);
  return L(() => {
    u.init({
      value: l,
      fields: v,
      options: a
    }), N.current && h && (h(F), N.current = !1);
  }, [u, l, v, a, h, F]), N.current && (u.init({
    value: l,
    fields: v,
    options: a
  }), h && h(F), N.current = !1), Je(
    R,
    () => ({
      ...F
    }),
    [F]
  ), /* @__PURE__ */ e(Oe, { words: Te, optional: !0, children: /* @__PURE__ */ e(Re.Provider, { value: F, children: /* @__PURE__ */ e(Wt, { type: w }) }) });
});
function qt(c) {
  const { fields: l = [], debounce: v = 300, onChange: a } = c, [w, h] = _(), [y, R] = _({}), o = ee(ae.i18n), u = o.getRaw(), b = u.filter && u.filter.dateFormat || u.formats && u.formats.dateFormat, f = o.getGroup("filter");
  function F(t) {
    const g = t.options.map((d) => typeof d == "string" || typeof d == "number" ? { id: d, label: d } : d);
    return g.unshift({ id: "$empty", label: f("None") }), g;
  }
  function N(t) {
    return typeof t == "string" ? { id: t, type: "text", filter: "contains" } : {
      ...t,
      filter: t.filter || (t.options ? "equal" : "contains"),
      options: t.type !== "date" && t.options && F(t)
    };
  }
  function A(t) {
    return t.map((g) => g.type === "all" || g.type === "dynamic" ? {
      ...g,
      by: g.by.map((d) => N(d))
    } : N(g));
  }
  function m(t) {
    return t.map((g) => ({ id: g.id, label: g.id }));
  }
  const r = T(() => A(l), [l]), p = x(null);
  L(() => {
    const t = {};
    let g = w;
    r.forEach((d) => {
      d.type === "dynamic" ? (d.by.forEach(($) => {
        typeof $.value < "u" && (t[$.id] = $.value);
      }), g = d.by[0].id) : d.type === "all" ? t[d.type] = d.value ?? "" : typeof d.value < "u" && (t[d.id] = d.value);
    }), p.current = t, R(t), h(g);
  }, [r]);
  function D(t, g, d) {
    return t.map(($) => {
      let I = {
        field: $.type,
        label: $.label,
        placeholder: $.placeholder
      };
      if ($.type === "dynamic") {
        I.dynamicField = !0, I.field = g;
        const le = $.by.find((H) => H.id === I.field);
        le && (I = {
          ...I,
          ...le
        });
      } else $.type !== "all" && (I = { ...I, ...$, field: $.id });
      return $.by && (I.optionsBy = m($.by)), I.value = d[I.field], { type: "text", filter: "contains", ...I };
    });
  }
  const E = T(
    () => D(r, w, y),
    [r, w, y]
  );
  function S(t, g) {
    if (t.field === "all")
      return _t(
        t.optionsBy,
        "contains",
        g[t.field]
      );
    const d = g[t.field];
    return d ? {
      field: t.field,
      filter: t.filter,
      type: t.type,
      value: d
    } : null;
  }
  function G(t) {
    const g = E.map((d) => S(d, t)).filter((d) => !!d);
    return g.length === 1 && g[0]?.glue === "or" ? g[0] : {
      glue: "and",
      rules: g
    };
  }
  const U = x(null);
  function K(t, g) {
    U.current && clearTimeout(U.current), U.current = setTimeout(() => {
      a && a({ value: G(g) });
    }, t);
  }
  L(() => {
    p.current && p.current !== y && K(1, y);
  }, [y]);
  const j = (t, g) => {
    R((d) => ({
      ...d,
      [g]: t === "$empty" ? "" : t
    }));
  }, X = (t) => {
    R((g) => (h(t), w ? { ...g, [w]: "" } : g));
  };
  return /* @__PURE__ */ e("div", { className: "wx-gWqAfosQ wx-filter-bar", children: E.map((t, g) => /* @__PURE__ */ O(Xe, { children: [
    t.label ? /* @__PURE__ */ e("div", { className: "wx-gWqAfosQ wx-label", children: t.label }) : null,
    t.dynamicField ? /* @__PURE__ */ e("div", { className: "wx-gWqAfosQ wx-select", children: /* @__PURE__ */ e(
      _e,
      {
        value: t.field,
        options: t.optionsBy,
        onChange: ({ value: d }) => X(d)
      }
    ) }) : null,
    t.options ? /* @__PURE__ */ e("div", { className: "wx-gWqAfosQ wx-select", children: /* @__PURE__ */ e(
      _e,
      {
        value: t.value,
        placeholder: t.placeholder ?? `${f("filter by")} ${t.field}`,
        options: t.options,
        onChange: ({ value: d }) => j(d, t.field)
      }
    ) }) : t.type === "text" || t.type === "number" ? /* @__PURE__ */ e("div", { className: "wx-gWqAfosQ wx-text", children: /* @__PURE__ */ e(
      Ie,
      {
        value: t.value,
        placeholder: t.placeholder ?? `${f("filter by")} ${t.field} (${f(t.filter)})`,
        onChange: ({ value: d, input: $ }) => {
          $ && j(d, t.field);
        },
        type: t.type
      }
    ) }) : t.type === "date" ? /* @__PURE__ */ e("div", { className: "wx-gWqAfosQ wx-date", children: t.filter === "between" || t.filter === "notBetween" ? /* @__PURE__ */ e(
      Ze,
      {
        value: t.value,
        format: b,
        buttons: ["done", "clear", "today"],
        placeholder: t.placeholder ?? `${f("filter by")} ${t.field} (${f(t.filter)})`,
        onChange: ({ value: d }) => j(d, t.field)
      }
    ) : /* @__PURE__ */ e(
      et,
      {
        value: t.value,
        format: b,
        placeholder: t.placeholder ?? `${f("filter by")} ${t.field} (${f(t.filter)})`,
        onChange: ({ value: d }) => j(d, t.field)
      }
    ) }) : null
  ] }, g)) });
}
function tn(c = {}) {
  return /* @__PURE__ */ e(Oe, { words: Te, optional: !0, children: /* @__PURE__ */ e(qt, { ...c }) });
}
const Bt = Ye(function(l, v) {
  const { items: a = [], children: w, onSelect: h, onClose: y } = l, R = x(null), [o, u] = _(-1);
  function b(r) {
    u(r);
  }
  function f(r) {
    let p;
    o === -1 ? p = r > 0 ? 0 : a.length - 1 : p = Math.max(0, Math.min(o + r, a.length - 1)), p !== o && (b(p), F(p));
  }
  function F(r) {
    if (r >= 0 && R.current) {
      const p = R.current.querySelectorAll(".wx-item")[r];
      p && p.scrollIntoView({ block: "nearest" });
    }
  }
  function N(r) {
    const p = ze(r), D = a.findIndex((E) => E.id == p);
    D !== -1 && D !== o && b(D);
  }
  function A(r) {
    if (a.length)
      switch (r.code) {
        case "Enter":
          o >= 0 ? (r.preventDefault(), h && h({ id: a[o].id })) : y && y();
          break;
        case "Escape":
        case "Tab":
          y && y();
          break;
        case "ArrowDown":
          r.preventDefault(), f(1);
          break;
        case "ArrowUp":
          r.preventDefault(), f(-1);
          break;
      }
  }
  Je(v, () => ({
    keydown: A
  }));
  function m(r) {
    r.stopPropagation();
    const p = ze(r), D = a.find((E) => E.id == p);
    D && h && h({ id: D.id });
  }
  return L(() => {
    (!R.current || !a.length) && u(-1);
  }, [a]), a.length <= 0 ? null : /* @__PURE__ */ e(bt, { onCancel: y, children: /* @__PURE__ */ e(
    "div",
    {
      className: "wx-list wx-aac97azI",
      ref: R,
      onClick: m,
      onMouseMove: N,
      children: a.map((r, p) => /* @__PURE__ */ e(
        "div",
        {
          className: `wx-item wx-aac97azI${p === o ? " wx-focus" : ""}`,
          "data-id": r.id,
          children: w ? w({ option: r }) : r.label
        },
        r.id
      ))
    }
  ) });
});
function zt(c) {
  const {
    query: l = "",
    fields: v = [],
    options: a = {},
    inline: w = !1,
    showErrors: h = !0,
    cursorPos: y = -1,
    parse: R = "allowFreeText"
  } = c, o = T(
    () => Tt(l, {
      fields: v,
      options: a,
      showErrors: h,
      cursorPos: y,
      allowFreeText: R === "allowFreeText"
    }),
    [l, v, a, h, y, R]
  );
  return w ? /* @__PURE__ */ e("span", { dangerouslySetInnerHTML: { __html: o } }) : l ? /* @__PURE__ */ e(
    "div",
    {
      className: "wx-query-highlight wx-aacwWF55",
      dangerouslySetInnerHTML: { __html: o }
    }
  ) : null;
}
function nn(c) {
  const {
    value: l = "",
    placeholder: v = "",
    onChange: a = null,
    parse: w = "allowFreeText",
    fields: h = [],
    options: y = {}
  } = c, o = (ee(ae.i18n) || St(Te)).getGroup("filter"), [u, b] = At(l || ""), f = T(() => w !== "none", [w]), F = T(() => w === "strict", [w]), N = T(() => kt(h), [h]), A = T(() => Dt(N), [N]);
  function m(s) {
    if (!s || typeof s == "boolean") return null;
    const C = o(s.code.toString().toLowerCase());
    return typeof C == "function" ? C(s.field, s.value) : s.message || C;
  }
  const [r, p] = _(""), D = x(!1), E = x(!1), [S, G] = _(!1), [U, K] = _(0), [j, X] = _(!0), [t, g] = _(0), [d, $] = _(!1), I = x(null), le = x(null), H = x(null), ue = T(
    () => f ? He(r, N, y, {
      allowFreeText: !F
    }) : null,
    [f, r, N, y, F]
  ), q = T(() => f ? Pt(r, t, N) : null, [f, r, t, N]), Q = T(
    () => q ? r.slice(q.start, q.end).toLowerCase() : "",
    [q, r]
  ), te = T(() => {
    if (!d) return [];
    if (q && r[q.start - 1] === "#")
      return [];
    const s = Vt(q, N, y);
    return !s || !Q ? s || [] : s.filter((C) => C.id.toLowerCase() !== Q);
  }, [d, q, r, N, y, Q]), de = x(a);
  de.current = a;
  const P = x(f);
  P.current = f;
  const me = x(F);
  me.current = F;
  const Se = x(N);
  Se.current = N;
  const M = x(y);
  M.current = y;
  const we = x(ue);
  we.current = ue;
  const ye = x(u);
  ye.current = u;
  const J = x(A);
  J.current = A;
  const fe = x(r);
  fe.current = r;
  const Y = Z(() => {
    K((s) => s + 1), G(!0);
  }, []), pe = Z(() => {
    G(!1);
  }, []);
  function ce() {
    $(!1);
  }
  const W = Z(() => {
    $(!1);
    const s = J.current;
    p((k) => Qe(k, s)), X(!0);
    const C = de.current;
    if (C) {
      const k = P.current, B = we.current, ne = ye.current, z = fe.current;
      if (k && B) {
        const ge = B.isInvalid ? {
          ...B.isInvalid,
          message: m(B.isInvalid)
        } : null, oe = !ge || ge.code === "NO_DATA" ? B.naturalText || "" : z;
        C({
          parsed: B,
          value: B.config,
          text: oe,
          error: ge,
          startProgress: Y,
          endProgress: pe
        });
      } else
        C({
          value: ne,
          text: z,
          startProgress: Y,
          endProgress: pe
        });
    }
  }, [Y, pe]);
  L(() => () => {
    E.current = !1;
  }, []), L(() => {
    if (D.current) {
      D.current = !1;
      return;
    }
    if (!E.current && (E.current = !0, !u))
      return;
    const s = Qe(u, J.current);
    p(s), fe.current = s, P.current && (we.current = He(s, Se.current, M.current, {
      allowFreeText: !me.current
    })), W();
  }, [u, W]);
  function ke(s) {
    p(s), D.current = !0, b(Ue(s, A));
  }
  function Fe(s) {
    te && te.length && H.current.keydown(s), s.key === "Enter" && !s.defaultPrevented && W();
  }
  function De(s) {
    const C = s.id;
    if (!q) return;
    const k = r.slice(0, q.start), B = r.slice(q.end);
    let ne, z;
    if (q.type === "field")
      ne = `${k}${C}: ${B}`, z = k.length + C.length + 2;
    else if (q.type === "predicate")
      ne = `${k}${C}: ${B}`, z = k.length + C.length + 2;
    else {
      const oe = k.endsWith(":") ? " " : "";
      ne = `${k}${oe}${C}${B}`, z = k.length + oe.length + C.length;
    }
    ke(ne), g(z), ce(), requestAnimationFrame(() => {
      I.current && (I.current.setSelectionRange(z, z), I.current.focus());
    });
  }
  function Pe(s) {
    const C = s.target.value;
    p(C), D.current = !0, b(Ue(C, A));
    const k = s.target.selectionStart;
    g(k), X(k), $(!0);
  }
  function he() {
    le.current && I.current && (le.current.scrollLeft = I.current.scrollLeft);
  }
  function ve() {
    ce();
  }
  function Ae() {
    p(""), D.current = !0, b("");
  }
  return /* @__PURE__ */ O(
    "div",
    {
      className: "wx-filter-query wx-aaccSW8j",
      onKeyDown: Fe,
      children: [
        /* @__PURE__ */ e(
          "div",
          {
            className: `wx-progress-bar wx-aaccSW8j${S ? " active" : ""}`,
            children: S && /* @__PURE__ */ e(
              "div",
              {
                className: "wx-progress-fill wx-aaccSW8j"
              },
              U
            )
          }
        ),
        /* @__PURE__ */ O("div", { className: "wx-filter-query-row wx-aaccSW8j", children: [
          /* @__PURE__ */ O("div", { className: "wx-filter-query-input-wrapper wx-aaccSW8j", children: [
            f && /* @__PURE__ */ e(
              "div",
              {
                className: `wx-filter-query-highlight wx-aaccSW8j${r ? "" : " wx-placeholder"}`,
                ref: le,
                "aria-hidden": "true",
                children: r ? /* @__PURE__ */ O(se, { children: [
                  /* @__PURE__ */ e(
                    zt,
                    {
                      parse: w,
                      query: r,
                      fields: N,
                      options: y,
                      inline: !0,
                      showErrors: j,
                      cursorPos: t
                    }
                  ),
                  " "
                ] }) : v
              }
            ),
            /* @__PURE__ */ e(
              "input",
              {
                type: "text",
                className: `wx-filter-query-input wx-aaccSW8j${f ? " wx-parse-mode" : ""}`,
                ref: I,
                value: r,
                placeholder: f ? "" : v,
                onInput: Pe,
                onScroll: he,
                onClick: ve,
                onChange: () => {
                }
              }
            )
          ] }),
          r && /* @__PURE__ */ e(
            "button",
            {
              type: "button",
              className: "wx-filter-query-clear wx-aaccSW8j",
              onClick: Ae,
              "aria-label": "Clear",
              children: /* @__PURE__ */ e("svg", { viewBox: "0 0 24 24", width: "16", height: "16", children: /* @__PURE__ */ e(
                "path",
                {
                  fill: "currentColor",
                  d: "M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                }
              ) })
            }
          ),
          /* @__PURE__ */ e(
            "button",
            {
              type: "button",
              className: "wx-filter-query-search wx-aaccSW8j",
              onClick: W,
              "aria-label": "Search",
              children: /* @__PURE__ */ e("svg", { viewBox: "0 0 24 24", width: "18", height: "18", children: /* @__PURE__ */ e(
                "path",
                {
                  fill: "currentColor",
                  d: "M15.5 14h-.79l-.28-.27A6.471 6.471 0 0016 9.5 6.5 6.5 0 109.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                }
              ) })
            }
          )
        ] }),
        /* @__PURE__ */ e(
          Bt,
          {
            ref: H,
            items: te,
            onSelect: De,
            onClose: ce,
            children: ({ option: s }) => s.label
          }
        )
      ]
    }
  );
}
function rn({ fonts: c = !0, children: l }) {
  return l ? /* @__PURE__ */ e(Me, { fonts: c, children: l }) : /* @__PURE__ */ e(Me, { fonts: c });
}
function ln({ fonts: c = !0, children: l }) {
  return l ? /* @__PURE__ */ e(qe, { fonts: c, children: l }) : /* @__PURE__ */ e(qe, { fonts: c });
}
function cn(c) {
  const { fonts: l = !0, children: v } = c;
  return v ? /* @__PURE__ */ e(Be, { fonts: l, children: v }) : /* @__PURE__ */ e(Be, { fonts: l });
}
export {
  tn as FilterBar,
  en as FilterBuilder,
  Ot as FilterEditor,
  nn as FilterQuery,
  rn as Material,
  ln as Willow,
  cn as WillowDark,
  an as createArrayFilter,
  un as createFilter,
  dn as createFilterRule,
  fn as getFilter,
  pn as getFilters,
  hn as getOptions,
  gn as getOptionsMap,
  xn as getQueryHtml,
  mn as getQueryString
};
