// @svar-ui/react-tasklist/dist/index.es.js
import { jsxs as y, jsx as s } from "react/jsx-runtime";
import { useContext as j, useRef as b, useCallback as m, useEffect as W, useState as g, useMemo as P } from "react";
import { context as Q, Checkbox as A, Button as G, Locale as H } from "@svar-ui/react-core";
import { tempID as L } from "@svar-ui/lib-state";
import { useWritableProp as S } from "@svar-ui/lib-react";
import { clickOutside as U } from "@svar-ui/lib-dom";
import { en as _ } from "@svar-ui/tasklist-locales";
function R(p) {
  const { task: e, edit: a, readonly: i, onUpdate: c, onRemove: f, onEdit: r } = p, k = j(Q.i18n).getGroup("tasklist"), d = b(null), x = b(""), v = m(() => {
    const n = d.current;
    n && (n.style.height = n.scrollHeight + 2 + "px");
  }, []), C = m(() => {
    const n = d.current;
    n && (x.current = n.value, v());
  }, [v]), K = m(
    ({ value: n }) => {
      c({
        task: e,
        status: n ? 1 : 0,
        content: x.current || e.content
      });
    },
    [c, e]
  ), w = m(
    (n) => {
      c({
        task: e,
        content: x.current,
        status: e.status,
        next: n
      });
    },
    [c, e]
  ), D = m(
    (n) => {
      n.key === "Escape" ? c({ task: e, status: e.status, content: "" }) : n.key === "Enter" && !n.shiftKey && (n.preventDefault(), w(!0));
    },
    [c, e, w]
  );
  W(() => {
    const n = d.current;
    n && (x.current = n.value = e.content, n.focus(), v());
  }, [e, a, v]), W(() => {
    const n = d.current;
    if (n)
      return U(n, w).destroy;
  }, [a, w]);
  const O = m(() => {
    r(e.id);
  }, [r, e.id]);
  return /* @__PURE__ */ y("div", { className: `wx-OQDwWK17 wx-task${e.status ? " wx-done" : ""}`, children: [
    /* @__PURE__ */ s("div", { className: "wx-OQDwWK17 wx-checkbox-wrapper", children: a === e.id && e.id === -1 ? /* @__PURE__ */ s("div", { className: "wx-OQDwWK17 wx-icon-add", children: /* @__PURE__ */ s("i", { className: "wx-OQDwWK17 wxi-plus" }) }) : /* @__PURE__ */ s(A, { onChange: K, value: e.status }) }),
    /* @__PURE__ */ s("div", { className: "wx-OQDwWK17 wx-wrapper", children: a === e.id ? /* @__PURE__ */ s(
      "textarea",
      {
        ref: d,
        className: "wx-OQDwWK17 wx-texarea",
        placeholder: k("Enter the task..."),
        onKeyDown: D,
        onInput: C
      }
    ) : /* @__PURE__ */ s("div", { className: "wx-OQDwWK17 wx-text-wrapper", onDoubleClick: () => r(e.id), children: /* @__PURE__ */ s("span", { className: "wx-OQDwWK17 wx-text", children: e.content }) }) }),
    /* @__PURE__ */ s("div", { className: "wx-OQDwWK17 wx-icon-edit", children: !i && a !== e.id ? /* @__PURE__ */ s(
      "i",
      {
        className: "wx-OQDwWK17 wxi-edit",
        onClick: O
      }
    ) : null }),
    /* @__PURE__ */ s("div", { className: "wx-OQDwWK17 wx-icon-close", children: !i && a !== e.id ? /* @__PURE__ */ s(
      "i",
      {
        className: "wx-OQDwWK17 wxi-close",
        onClick: () => f(e.id)
      }
    ) : null })
  ] });
}
function T(p) {
  const { readonly: e = !1, onChange: a } = p, [i, c] = S(p.data || []), [f, r] = g(null), [k, d] = g(null), x = j(Q.i18n), v = P(() => x.getGroup("tasklist"), [x]);
  function C() {
    r(-1), d({ id: -1, content: "", status: 0 });
  }
  function K(t) {
    const o = {
      id: L(),
      content: t,
      status: 0
    }, u = [...i, o];
    if (c(u), a) {
      const l = a({ value: u, task: o, action: "add" });
      l && typeof l == "object" && (l.then ? l.then((h) => {
        w(o.id, h);
      }) : w(o.id, l));
    }
  }
  function w(t, o) {
    c((u) => {
      const l = u.findIndex((N) => N.id === t);
      if (l === -1) return u;
      const h = [...u];
      return h[l] = { ...u[l], ...o }, h;
    });
  }
  function D(t) {
    f === t && r(null);
    const o = i.filter((u) => u.id !== t);
    c(o), t !== -1 && a && a({ value: o, id: t, action: "delete" });
  }
  function O(t, o, u) {
    let l;
    const h = i.map((N) => N.id === t ? (l = { ...N, content: o, status: u }, l) : N);
    c(h), r(null), a && a({ value: h, id: t, action: "update", task: l });
  }
  function n(t) {
    D(t);
  }
  function E({ task: t, content: o, status: u, next: l }) {
    t.id === -1 ? (r(null), o ? (K(o), l && C()) : D(t.id)) : (o ? (o !== t.content || u !== t.status) && O(t.id, o, u) : D(t.id), r(null));
  }
  function I(t) {
    e || r(t);
  }
  return /* @__PURE__ */ s(Q.fieldId.Provider, { value: null, children: /* @__PURE__ */ y("div", { className: "wx-kro6Nsfl wx-tasks-list", children: [
    /* @__PURE__ */ y("div", { className: "wx-kro6Nsfl wx-list", children: [
      i.map((t) => /* @__PURE__ */ s(
        R,
        {
          task: t,
          edit: f,
          onUpdate: E,
          onRemove: n,
          onEdit: I,
          readonly: e
        },
        t.id
      )),
      f === -1 ? /* @__PURE__ */ s(
        R,
        {
          task: k,
          edit: f,
          onUpdate: E
        },
        f
      ) : null
    ] }),
    !e && f !== -1 ? /* @__PURE__ */ s("div", { className: "wx-kro6Nsfl wx-button", children: /* @__PURE__ */ s(G, { onClick: C, children: v("Add task") }) }) : null
  ] }) });
}
function J(p) {
  const { onData: e, onChange: a, value: i, ...c } = p;
  let [f, r] = g(null);
  W(() => {
    e && i ? Promise.resolve(e(i)).then((d) => r(d)) : i && i.then ? i.then((d) => r(d)) : r(i || []);
  }, [e, i]);
  const k = m(
    (d) => (d.originalValue = i, a && a(d)),
    [a, i]
  );
  return /* @__PURE__ */ s(H, { words: _, optional: !0, children: f === null ? /* @__PURE__ */ s(
    T,
    {
      data: [],
      readonly: !0,
      ...c,
      onChange: k
    }
  ) : /* @__PURE__ */ s(T, { data: f, ...c, onChange: k }) });
}
export {
  J as Tasklist
};
